{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "943bbc45-ec7b-4318-a263-46e326a5f3b1",
   "metadata": {},
   "source": [
    "# Project Description"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aece54cb-aa8b-4406-8b38-b0986de20647",
   "metadata": {},
   "source": [
    "This project involves heterochromation profiling of plasma from lymphoma patients, vs non-lymphoma controls. Primary goal of this project is to identify a signature of lymphoma features from plasma H3K27me3 profiles, which significantly differ from non-lymphoma controls. Additional data includes clinical metadata from each of the lymphoma patients, and other parallel plasma measurements including CAPP-Seq (mutations using a lymphoma-specific panel), and shallow WGS. Metabolic Tumor Volume quantified from PET-CT imaging is also included for these patients."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94e0a7b8-6f4c-4368-aa4b-282ed4dad1e2",
   "metadata": {},
   "source": [
    "# Import libraries and set plot dimensions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28a3a839-d1d1-4a38-b079-ea7cdbe9471a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# libraries\n",
    "library(tidyverse)\n",
    "library(edgeR)\n",
    "library(sva)\n",
    "library(reshape2)\n",
    "library(pheatmap)\n",
    "library(RColorBrewer)\n",
    "library(MetBrewer)\n",
    "library(caret)\n",
    "library(glmnet)\n",
    "library(e1071)\n",
    "library(edgeR)\n",
    "library(umap)\n",
    "library(DMwR)\n",
    "library(pROC)\n",
    "library(gbm)\n",
    "library(survival)\n",
    "library(survminer)\n",
    "library(GenomicRanges)\n",
    "library(recipes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a1931bc-27a9-4eda-8459-8ea055539054",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot dimensions\n",
    "options(repr.plot.width = 10, repr.plot.height = 10) # specific to the IRkernel"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc9e81c3-bea3-4b03-b410-86f10eb4815b",
   "metadata": {},
   "source": [
    "# Data pre-processing + differential analysis (edgeR)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c1d0447-23f1-4961-97f0-91cb6b9b6815",
   "metadata": {},
   "source": [
    "## Import counts data and compile into a numeric matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41f30751-eb49-4360-bff2-b326d752d437",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import 10kb reference, w regulatory features and bin IDs. Reference was created using annotatR. Note; reference generation requires an internet connection.\n",
    "setwd(\"/path/to/reference/\")\n",
    "annotated_ref_no_X_Y_M_blacklist <- read.delim(\"10kb_bin_genome_wide_annotated_reference_with_regulatory_features_no_X_Y_M_ENCODEBlacklist.txt\", header = TRUE, sep = \"\\t\")\n",
    "\n",
    "# import sample list, formatted beforehand.\n",
    "setwd(\"/path/to/list/\")\n",
    "sample_list <- read.delim(\"sample_list_for_lymphoma_cohort.txt\", header = TRUE, sep = \"\\t\")\n",
    "\n",
    "gc() #clear memory if memory is a bottleneck; the data frames in this script are large\n",
    "\n",
    "# import non-normalized counts, output from MEDIPS (in 10kb bins)\n",
    "# BAM files prior to generation of non-normalized counts were generated using MEDIPIPE pre-processing pipeline (10.1093/bioinformatics/btad423, originally intended for cfMeDIP-Seq data, but adapted for cfChIP-Seq). \n",
    "setwd(\"/path/to/samples/\")\n",
    "import_counts <- list.files(path = \".\", pattern = \"counts_10kb.txt$\", recursive = TRUE) # list counts with directory paths\n",
    "\n",
    "# exclude samples with these names, as they are not required for the following analyses. Only T1 (baseline) samples should be included.\n",
    "exclude <- c(\"0018_T3\", \"0025_T4\", \"0027_T3\", \"0027_T4\", \"0101_T3\", \"0030_T1\", \"HUCON_37\", \"HUCON_43\", \"HUCON_44\", \"0036_C\", \"0247_T1_ABC\", \"T3\", \"T5\", \"pbmc_k27\") #character vector of sample strings to be excluded.\n",
    "\n",
    "# convert character vector to data frame\n",
    "import_counts <- as.data.frame(import_counts)\n",
    "\n",
    "# Remove all excluded samples from the original list of samples. Use \"|\" as the seperator between vector elements.\n",
    "import_counts <- import_counts[!grepl(paste(exclude, collapse = \"|\"), import_counts$import_counts),] # \n",
    "\n",
    "# loop over files in import_counts, and read them into memory\n",
    "counts <- mapply(read.delim, import_counts) \n",
    "\n",
    "# bind individual datasets across columns\n",
    "counts_matrix <- do.call(cbind, counts)\n",
    "\n",
    "# change column names to file names from list, and remove NAs\n",
    "colnames(counts_matrix) <- c(paste0(sample_list$sample_name))\n",
    "counts_matrix <- na.omit(counts_matrix)\n",
    "\n",
    "# index rows in counts_matrix\n",
    "rownames(counts_matrix) <- 1:nrow(counts_matrix)\n",
    "\n",
    "gc()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61b47011-947b-43c5-9d38-06563097dbbb",
   "metadata": {},
   "source": [
    "## Data cleaning, filter out low variance and low count features from the matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88601602-a443-4d59-965b-cbcd6cbf092c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#filter out chrX, Y, and ENCODE Blacklist regions. Resulting data frame is your ground truth; refer back to this after feature selection.\n",
    "all_filtered_features <- counts_matrix[as.integer(rownames(counts_matrix)) %in% annotated_ref_no_X_Y_M_blacklist$bin_id,]\n",
    "\n",
    "# calculate the coefficient of variation (CV, ratio of SD to the mean) for all features using the var function, row-wise. Independent of sample origin.\n",
    "cv_features <- apply(all_filtered_features, 1, var)\n",
    "\n",
    "# visualize histogram of feature variances\n",
    "var_hist <- hist(cv_features[cv_features > 0 & cv_features < 40000], breaks = 500) # bimodal distribution, left skewed\n",
    "\n",
    "# remove variance below the valley on the histogram. Valley is at around 2400.\n",
    "all_filtered_features <- all_filtered_features[(which(apply(all_filtered_features, 1, var)>=2400)),]\n",
    "\n",
    "# remove low count features\n",
    "counts_mean <- apply(all_filtered_features, 1, mean) # left skewed, slight bimodal distribution of counts after filtering low variance features.\n",
    "counts_hist <- hist(counts_mean[counts_mean > 0 & counts_mean < 1000], breaks = 500) \n",
    "\n",
    "# Remove lowest expressed features\n",
    "all_filtered_features <- all_filtered_features[(which(apply(all_filtered_features, 1, mean)>=50)),]\n",
    "\n",
    "gc() #clear memory"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f449791-cc91-4f80-ad6a-1a6d6619ce16",
   "metadata": {},
   "source": [
    "## Split dataset into train and test sets for feature selection / validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fe6c785-58fc-42cc-8331-3269d01e2fa0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# split train and test sets. Perform differential analysis on only training set.\n",
    "\n",
    "# Split the data into features (X) and labels (y)\n",
    "X <- all_filtered_features  # Features (numeric matrix plus sample name attributes)\n",
    "y <- sample_list$timepoint # Data labels for lymphoma vs non-lymphoma classification\n",
    "\n",
    "# Training: 70%; Test: 30%\n",
    "# Split the data into training and testing sets\n",
    "set.seed(23)  # For reproducibility\n",
    "\n",
    "train_indices <- createDataPartition(y, p = 0.7, list = FALSE)\n",
    "\n",
    "# save train and test set numeric matrices and labels as independent variables. Lock test set away for later validation.\n",
    "X_train <- X[, train_indices]\n",
    "y_train <- y[train_indices]\n",
    "X_test <- X[, -train_indices]\n",
    "y_test <- y[-train_indices]\n",
    "\n",
    "#check balance of healthy and baseline samples in train/test splits. Ensure it is close to 70/30.\n",
    "table(y_train)\n",
    "table(y_test)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2d3a06f-2266-416d-8251-dec80be03619",
   "metadata": {},
   "source": [
    "## Perform differential analysis using edgeR, using SVA to address unknown patterns in the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef2c43a3-b667-4714-ac64-98aae5b17aab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# this chunk takes time and computational power, so schedule script as a job if working on a virtual machine.\n",
    "# script for differential analysis was adapted from edgeR user guide (https://bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf)\n",
    "gc()\n",
    "\n",
    "# set up DGEList object for differential analysis\n",
    "design <- model.matrix(~y_train) #design model around differences between lymphoma and non-lymphoma profiles. Can change depending on intended comparison.\n",
    "list <- DGEList(counts = X_train, samples = y_train, remove.zeros = FALSE) # no need to remove zeros, since they were removed in matrix pre-processing steps.\n",
    "\n",
    "# apply TMM normalization to raw counts\n",
    "tmm_norm <- calcNormFactors(list, method=\"TMM\")\n",
    "\n",
    "# normalize counts before SVA is performed; otherwise library size is the only surrogate variable.\n",
    "cpm_norm <- cpm(tmm_norm, normalized.lib.sizes = TRUE, log = FALSE)\n",
    "\n",
    "# perform SVA analysis to remove technical variation\n",
    "sva_fit <- sva(cpm_norm, design)\n",
    "\n",
    "#IMPORTANT; add SVA surrogate variables as covariates in model design.\n",
    "design <- model.matrix(~y_train + sva_fit$sv)\n",
    "\n",
    "## save model before differential analysis is applied, in case the analysis is not complete within the bounds of the wall time allocated.\n",
    "#save.image()\n",
    "\n",
    "gc()\n",
    "\n",
    "# steps below here are to perform the differential analysis\n",
    "fit <- estimateDisp(tmm_norm, design = design, trend.method = \"locfit\", robust = TRUE)\n",
    "\n",
    "gc()\n",
    "\n",
    "# fit model, incorporating surrogate variables\n",
    "QL <- glmQLFit(fit, design = design)\n",
    "\n",
    "#Subset to rows with an FDR <0.05\n",
    "fdr <- table(p.adjust(QL$table$PValue, method=\"BH\")<0.05)\n",
    "\n",
    "lrt <- glmQLFTest(QL, coef = 2)\n",
    "\n",
    "# extract all features, sorted by p-val\n",
    "res <- topTags(lrt, n = 246275, adjust.method = \"BH\", sort.by = \"PValue\")\n",
    "\n",
    "# save dataset\n",
    "setwd(\"path/to/working/directory/\")\n",
    "save.image(\"train_test_split_for_differential_analysis.RData\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac5b1fec-82b5-4cab-a144-f3bb359cb7c6",
   "metadata": {},
   "source": [
    "## Subset to significant features from differential analysis, generate volcano plot of features as sanity check"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88daf8c5-5a3a-4086-b3cc-e02bf5016475",
   "metadata": {},
   "outputs": [],
   "source": [
    "setwd(\"path/to/working/directory/\")\n",
    "load(\"train_test_split_for_differential_analysis.RData\")\n",
    "\n",
    "#keep only the features that meet a p-value threshold.\n",
    "#de_features_subset <- res[(res$table$logFC >= 1 | res$table$logFC <= -1) & res$table$PValue <= 0.01,] # logFC >= 1 is 2x change; only 19 features.\n",
    "#de_features_subset <- res[(res$table$logFC >= 0.25 | res$table$logFC <= -0.25) & res$table$PValue <= 0.05,] # 341 features.\n",
    "#de_features_subset <- res[res$table$PValue <= 0.01,] # 1742 features\n",
    "de_features_subset <- res[res$table$PValue <= 0.05,] # 9370 features; stick with this one as it is less stringent. Let the Random Forest decide which features are important.\n",
    "\n",
    "# store feature names (aka bin IDs) for differential features and all features as separate variables\n",
    "de_features <- rownames(de_features_subset)\n",
    "de_features_all <- rownames(res)\n",
    "\n",
    "# calculate logFC, pval, and -log10(pval) for all features\n",
    "logFC <- res$table$logFC\n",
    "p_value <- res$table$PValue\n",
    "neg_log_p_value <- -log10(p_value)\n",
    "\n",
    "# compile data for volcano plot.\n",
    "df_feature_stats <- data.frame(logFC,neg_log_p_value,de_features_all)\n",
    "\n",
    "# calculate logFC, pval, and -log10(pval) for all significant features\n",
    "logFC_signif <- de_features_subset$table$logFC\n",
    "p_value_signif <- de_features_subset$table$PValue\n",
    "neg_log_p_value_signif <- -log10(p_value_signif)\n",
    "\n",
    "# compile data for volcano plot.\n",
    "df_feature_stats_signif <- data.frame(logFC_signif,neg_log_p_value_signif,de_features)\n",
    "\n",
    "v <- ggplot() +\n",
    "  geom_point(data = df_feature_stats, aes(x = logFC, y = neg_log_p_value), colour = \"black\", alpha = 0.3) +\n",
    "  geom_point(data = df_feature_stats_signif, aes(x = logFC_signif, y = neg_log_p_value_signif), colour = \"red\", alpha = 0.3) +\n",
    "  geom_hline(yintercept = -log10(0.05), linetype = \"dashed\", color = \"red\") +\n",
    "  geom_vline(xintercept = c(-1, 1), linetype = \"dashed\", color = \"blue\") +\n",
    "  labs(x = \"Log Fold Change\", y = \"-log10(p-value)\") +\n",
    "  ggtitle(\"Volcano Plot\") +\n",
    "  theme_minimal()\n",
    "print(v)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2d90e2d-07dd-4d00-8a87-dfa015059a32",
   "metadata": {},
   "source": [
    "## Perform Principal Component Analysis (PCA) on all significant differential features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59d48ad0-37a2-4eaf-a847-31c4a517bd78",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot PCA of significant features, to see whether the features differentiate between lymphoma profiles and non-lymphoma controls\n",
    "# compare to PCA of all features\n",
    "\n",
    "# ensure you transpose the feature matrix before PCA\n",
    "# pca_data <- t(cpm_norm) # for all features\n",
    "pca_data <- t(cpm_norm[de_features,]) # for feature subset\n",
    "\n",
    "# perform PCA, with z-scaling\n",
    "pca_res <- prcomp(pca_data, scale = TRUE)\n",
    "\n",
    "# Plot the first two principal components. Commented geoms are additional data layers, for additional visualizations.\n",
    "q <- pca_res$x %>% \n",
    "  as.data.frame %>%\n",
    "  ggplot(aes(x=PC1,y=PC2)) +\n",
    "  #geom_point(aes(color=list$samples$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY),size=5) +\n",
    "  #geom_point(aes(color=list$samples$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY.DIAGNOSIS_SPECIFIC_DLBCL),size=5) +\n",
    "  #geom_point(aes(color=list$samples$timepoint),size=5) +\n",
    "  geom_point(aes(color=y_train),size=5) +\n",
    "  #geom_point(aes(color=list$samples$batch),size=5) +\n",
    "  #geom_text(aes(label=list$samples$sample_name)) +\n",
    "  theme_minimal(base_size=20) + \n",
    "  labs(colour = \"Group\") +\n",
    "  xlab(paste0(\"PC1 (\", round(pca_res$sdev[1]^2*100/sum(pca_res$sdev^2), 1), \"%)\")) +\n",
    "  ylab(paste0(\"PC2 (\", round(pca_res$sdev[2]^2*100/sum(pca_res$sdev^2), 1), \"%)\")) +\n",
    "  theme(legend.position=\"right\")\n",
    "\n",
    "print(q)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54afd57b-03cc-4444-9cd0-540e5d9da50d",
   "metadata": {},
   "source": [
    "# Random Forest of significant differential features (additional feature selection step)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e70156a-ca9a-4a29-b657-2fd9f5ed7d2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# chunk requires at least 60GB of memory on a virtual machine, run script as a scheduled job on compute cluster, requesting compute node.\n",
    "\n",
    "# import dataset with significant differential features\n",
    "setwd(\"path/to/working/directory/\")\n",
    "load(\"train_test_split_for_differential_analysis.RData\")\n",
    "\n",
    "# subset differential results to differential features\n",
    "de_features_subset <- res[res$table$PValue <= 0.05,] # 9370 features\n",
    "# de_features_subset <- res[res$table$PValue <= 0.05 & res$table$logFC > 0,] # 5328 features; only hypermethylated features considered. Compare classification performance\n",
    "\n",
    "# save bin IDs (aka rownames) of differential features, depending on which set you're using\n",
    "de_features <- rownames(de_features_subset)\n",
    "\n",
    "# subset CPM normalized matrix to differential features\n",
    "de_features_matrix <- cpm_norm[de_features,]\n",
    "\n",
    "# specify tune grid for hyperparameter tuning; tuning hyperparameters depend on the model used.\n",
    "tune_grid <- expand.grid(\n",
    "  mtry = c(2, 5, 10) # tuning hyperparameter for rf. Can test a range of values here, or do a grid search in trainControl\n",
    ")\n",
    "\n",
    "# set up train control for Random Forest model (R Caret).\n",
    "ctrl <- trainControl(\n",
    "  method = \"repeatedcv\",  # Use cross-validation for evaluation\n",
    "  repeats = 5, # number of model iterations\n",
    "  number = 10,  # Number of folds for cross-validation\n",
    "  search = \"grid\",\n",
    "  sampling = \"smote\",\n",
    "  verbose = FALSE\n",
    ")\n",
    "\n",
    "# transpose feature matrix for machine learning\n",
    "t_de_features_matrix <- t(de_features_matrix)\n",
    "\n",
    "model <- train(t_de_features_matrix, \n",
    "               as.factor(y_train), \n",
    "               method = \"rf\", # for Random Forest\n",
    "               # method = \"nb\", # naive bayes; turn off tuneGrid.\n",
    "               # method = \"glmnet\", # try with specifying the tuneLength parameter next\n",
    "               trControl = ctrl, \n",
    "               tuneGrid = tune_grid, # turn off for nb\n",
    "               metric = \"Accuracy\",\n",
    "               #metric = \"ROC\",\n",
    "               #metric = \"Kappa\", # for unbalanced sets; since we're using SMOTE, accuracy should be okay\n",
    "               ntree = 500 # for rf\n",
    "               #tuneLength = 8 # take random values for default tuning parameters and pick the best; alternative to tuneGrid\n",
    ")\n",
    "\n",
    "# print error log for model output, in case the model fails\n",
    "# sink(rf_hyperparameter_tuning_accuracy_error_log.txt)\n",
    "\n",
    "# see model results for training data\n",
    "print(model)\n",
    "\n",
    "# extract predictors for further analysis (aka feature validation)\n",
    "predictors <- predictors(model)\n",
    "\n",
    "# save dataset w model\n",
    "setwd(\"path/to/working/directory/\")\n",
    "save.image(\"rf_model_from_differential_features.RData\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32631e13-fbf9-4bd4-8c9e-3eb2034f6aff",
   "metadata": {},
   "source": [
    "# REFINE AND OPTIMIZE CODE FROM HERE"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc33ec90-65b7-407a-a564-3ec8aba8ee7f",
   "metadata": {},
   "source": [
    "# Analyses incorporating discovered features"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1489a2f-b976-453b-9a76-dc5ca750de5d",
   "metadata": {},
   "source": [
    "## Comparison of discovered features across prognostic groups in lymphoma"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2ec6e25-b865-4b56-bd99-dde20e82c118",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load dataset\n",
    "setwd(\"/cluster/projects/scottgroup/people/steven/human_plasma_cfChIP-Seq/liberate_plasma_cfchip/analysis/machine_learning_models/02_07_2023_RFE_RF_10kb_windows/differential_analysis_filter/\")\n",
    "load(\"10_09_2023_rf_10kb_with_hyperparameter_tuning_accuracyevaluation_withdifferentialfilter.RData\")\n",
    "\n",
    "# import clinical metadata\n",
    "setwd(\"/cluster/home/sdemichi/steven_scottgroup.txt/human_plasma_cfChIP-Seq/liberate_plasma_cfchip/temp/\")\n",
    "clinical_metadata <- as.data.frame(readxl::read_xlsx(\"25_07_2023_cohort_metadata_labeling_UPDATED_WITH_CLINICAL_ANNOTATIONS_plus_library_size_cfdna_concn.xlsx\"))\n",
    "\n",
    "# import mean VAF information - UPDATED FOR BATCH 2 - using dcs.unique.burden\n",
    "setwd(\"/cluster/projects/scottgroup/people/steven/human_plasma_cfChIP-Seq/liberate_plasma_cfchip/other_profiles/cappseq_data_lymphoma/updated_batch_2_all_cases/\")\n",
    "mean_VAF_burden <- read.delim(\"Bratman.dcs.uniq.burden.txt\", header = TRUE, sep = \"\\t\") # for dcs.unique\n",
    "\n",
    "# replace spaces in \"Donor ID (rep)\" so that you can match directly to mean VAF burden$Sample_id\n",
    "clinical_metadata$`Donor ID (rep)` <- gsub(\" \", \"_\", clinical_metadata$`Donor ID (rep)`)\n",
    "\n",
    "# match mean VAF information to clinical metadata\n",
    "clinical_metadata$mean_VAF <- mean_VAF_burden$mean_VAF[match(clinical_metadata$'Donor ID (rep)', mean_VAF_burden$Sample_id)]\n",
    "\n",
    "# match annotations from X_train to the new clinical annotations. Append cfDNA concentrations.\n",
    "#all_filtered_features_X_train <- all_filtered_features[,colnames(all_filtered_features) %in% colnames(X_train)]\n",
    "\n",
    "# subset sample metadata to training samples.\n",
    "training_sample_metadata <- sample_list[sample_list$sample_name %in% colnames(cpm_norm),]\n",
    "\n",
    "# filter clinical metadata by append cfDNA concentration to test, from clinical metadata.\n",
    "clinical_metadata <- clinical_metadata[clinical_metadata$PATIENT_ID %in% training_sample_metadata$PATIENT_ID,] \n",
    "\n",
    "# remove all T3 samples from the metadata\n",
    "clinical_metadata <- subset(clinical_metadata, !grepl(\"T3|T5\", clinical_metadata$`Donor ID (rep)`))\n",
    "\n",
    "# subset sample metadata to training samples\n",
    "X_train_attributes <- sample_list[sample_list$sample_name %in% colnames(cpm_norm),]\n",
    "\n",
    "# sort clinical_metadata in order of colnames of X_train_attributes$PATIENT_ID\n",
    "clinical_metadata <- clinical_metadata[match(X_train_attributes$PATIENT_ID, clinical_metadata$PATIENT_ID), ]\n",
    "\n",
    "# overwrite X_train_attributes with clinical_metadata\n",
    "X_train_attributes <- clinical_metadata\n",
    "\n",
    "\n",
    "\n",
    "# check model output\n",
    "print(model)\n",
    "predictors_rf <- predictors(model)\n",
    "\n",
    "# side; SUBSET TO PREDICTORS WITH A POSITIVE logFC\n",
    "predictors_rf_hyper <- res[rownames(res) %in% predictors_rf & res$table$logFC > 0,]\n",
    "predictors_rf_hyper <- rownames(predictors_rf_hyper)\n",
    "\n",
    "# side; SUBSET TO ALL FEATURES WITH A POSITIVE logFC\n",
    "all_hyper <- res[res$table$logFC > 0,]\n",
    "all_hyper <- rownames(all_hyper)\n",
    "\n",
    "# side; SUBSET TO ALL FEATURES WITH A POSITIVE logFC\n",
    "all_sig_hyper <- res[res$table$logFC > 0 & res$table$PValue < 0.05,]\n",
    "all_sig_hyper <- rownames(all_sig_hyper)\n",
    "\n",
    "# side; SUBSET TO ALL FEATURES WITH A NEGATIVE logFC\n",
    "all_sig_hypo <- res[res$table$logFC < 0 & res$table$PValue < 0.05,]\n",
    "all_sig_hypo <- rownames(all_sig_hypo)\n",
    "\n",
    "# SCALE FEATURES BEFORE SUBSETTING AND VISUALIZING\n",
    "# cpm_norm_scaled <- scale(cpm_norm, center = TRUE, scale = TRUE)\n",
    "# cpm_norm_scaled <- scale(cpm_norm, center = TRUE, scale = apply(cpm_norm, 2, function(x) max(x) - min(x)))\n",
    "\n",
    "# alternate normalization; log2(CPM+1) normalize before subsetting\n",
    "cpm_norm_log2 <- log2(cpm_norm + 1)\n",
    "X_train_filtered <- cpm_norm_log2[(rownames(cpm_norm_log2)) %in% predictors_rf_hyper,] # for hypermethylated modeled features\n",
    "# X_train_filtered <- cpm_norm_log2[(rownames(cpm_norm_log2)) %in% all_sig_hyper,] # for hypermethylated modeled features\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# ADDITIONAL STEP; multiply the normalized feature matrix by the weights from the model. Matrix multiplication, after sorting the weights.\n",
    "\n",
    "# model_feature_weights <- model$finalModel$importance # weights\n",
    "\n",
    "# # subset weights to features in X_train_filtered\n",
    "# model_feature_weights <- model_feature_weights[rownames(model_feature_weights) %in% rownames(X_train_filtered),]\n",
    "# model_feature_weights <- as.data.frame(model_feature_weights) # convert to data frame\n",
    "\n",
    "# # make rownames of model_feature_weights a column, and sort in ascending order by row\n",
    "# model_feature_weights$bin_id <- as.numeric(rownames(model_feature_weights))\n",
    "\n",
    "# # sort feature weights by feature\n",
    "# model_feature_weights <- model_feature_weights[order(model_feature_weights$bin_id, decreasing = FALSE),]\n",
    "\n",
    "# # multiply weights from model_feature_weights column, with the normalized counts from X_train_filtered\n",
    "# X_train_filtered <- model_feature_weights$model_feature_weights * X_train_filtered\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# X_train_filtered <- cpm_norm_scaled[(rownames(cpm_norm_scaled)) %in% predictors_rf,]\n",
    "# X_train_filtered <- cpm_norm_scaled[(rownames(cpm_norm_scaled)) %in% predictors_rf_hyper,] # for hypermethylated modeled features\n",
    "# X_train_filtered <- cpm_norm_scaled[(rownames(cpm_norm_scaled)) %in% all_hyper,] # for all hypermethylated features\n",
    "# X_train_filtered <- cpm_norm_scaled[(rownames(cpm_norm_scaled)) %in% all_sig_hyper,] # for all significant hypermethylated features\n",
    "# X_train_filtered <- cpm_norm_scaled[(rownames(cpm_norm_scaled)) %in% all_sig_hypo,] # for all significant hypomethylated features\n",
    "\n",
    "# add the lowest scaled count in the matrix to each of the counts to remove zeros\n",
    "# test <- min(X_train_filtered)\n",
    "# X_train_filtered <- X_train_filtered + abs(test)\n",
    "\n",
    "# subset original training matrix to the features in chosen feature set\n",
    "#X_train_filtered <- cpm_norm[(rownames(cpm_norm)) %in% predictors_rf,]\n",
    "\n",
    "# save new matrix as a DGEobject, since this links the sample attributes and the counts\n",
    "X_train_filtered_features_plus_attributes <- DGEList(counts = X_train_filtered, samples = X_train_attributes, remove.zeros = TRUE)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#calculate the median of each column\n",
    "median_X_train_filtered <- apply(X_train_filtered_features_plus_attributes$counts, \n",
    "                   MARGIN = 2, \n",
    "                   FUN = median)\n",
    "\n",
    "#calculate the mean of each column\n",
    "mean_X_train_filtered <- apply(X_train_filtered_features_plus_attributes$counts, \n",
    "                   MARGIN = 2, \n",
    "                   FUN = mean)\n",
    "\n",
    "#include the sums of the features per sample as an additional feature, and visualize\n",
    "sum_X_train_filtered <- colSums(X_train_filtered_features_plus_attributes$counts)\n",
    "\n",
    "#bind the newly calculated medians and sumsto the pre-existing metadata\n",
    "X_train_attributes <- as.data.frame(cbind(X_train_attributes, median_X_train_filtered)) #Subset this data frame for plotting.\n",
    "X_train_attributes <- as.data.frame(cbind(X_train_attributes, sum_X_train_filtered)) #Subset this data frame for plotting.\n",
    "\n",
    "##### METADATA-SPECIFIC CHANGES BEFORE PLOTTING #####\n",
    "X_train_attributes$ANN_ARBOR_STAGE <- ifelse(X_train_attributes$ANN_ARBOR_STAGE == \"NA\", \"non-lymphoma\", X_train_attributes$ANN_ARBOR_STAGE) # convert NA characters to \"non-lymphoma\"\n",
    "X_train_attributes$IPI <- as.character(X_train_attributes$IPI) #convert IPI from int to char\n",
    "\n",
    "X_train_attributes$mean_VAF <- as.numeric(X_train_attributes$mean_VAF) #convert IPI from int to char\n",
    "\n",
    "\n",
    "#For tumor diameter, create bins (<5cm, 5-10cm, 10-15cm, >15cm), then assign samples to those bins based off of tumor diameter; separate column\n",
    "X_train_attributes$TUMOR_MASS_MAX_CM <- as.numeric(X_train_attributes$TUMOR_MASS_MAX_CM)\n",
    "X_train_attributes <- X_train_attributes %>% \n",
    "  mutate(tumor_diameter_binned = cut(TUMOR_MASS_MAX_CM, breaks = c(0, 5, 10, 15, 30), labels = c(\"<5\", \"5-10\", \"10-15\", \">15\")))\n",
    "\n",
    "# combine Stage I/II and III/IV samples into early and late stage, respectively, for visualization\n",
    "X_train_attributes <- X_train_attributes %>%\n",
    "  mutate(Stage_Group = case_when(\n",
    "    ANN_ARBOR_STAGE %in% c(1, 2) ~ \"Early Stage\",\n",
    "    ANN_ARBOR_STAGE %in% c(3, 4) ~ \"Late Stage\",\n",
    "    TRUE ~ as.character(ANN_ARBOR_STAGE)  # Keep other values as is\n",
    "  ))\n",
    "\n",
    "# set factor level for Stage_Group\n",
    "X_train_attributes$Stage_Group <- factor(X_train_attributes$Stage_Group, levels = c(\"non-lymphoma\", \"Early Stage\", \"Late Stage\"))\n",
    "\n",
    "# 24-10-2023; create new variable where IPI is grouped 1+2 and 4+5, as this is clinically relevant\n",
    "X_train_attributes <- X_train_attributes %>%\n",
    "  mutate(IPI_grouped = case_when(\n",
    "    IPI %in% c(0, 1) ~ \"0+1\",\n",
    "    IPI %in% c(4, 5) ~ \"4+5\",\n",
    "    TRUE ~ as.character(IPI)  # Keep other values as is\n",
    "  ))\n",
    "\n",
    "# 24-10-2023; subset X_train_attributes to FL cases, and visualize FLIPI\n",
    "# X_train_attributes_FL <- X_train_attributes[X_train_attributes$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY == \"FL\", ]\n",
    "\n",
    "# for grouped IPI, remove healthy cases from matrix\n",
    "# X_train_attributes <- X_train_attributes[!(X_train_attributes$IPI_grouped == \"NA\"),] # for grouped IPI\n",
    "\n",
    "\n",
    "\n",
    "##### PLOT #####\n",
    "\n",
    "# X_train_attributes_FL %>% \n",
    "X_train_attributes %>% \n",
    "  ggplot(aes(x = Stage_Group, y = median_X_train_filtered)) +#change to sum or median, depending on what you're visualizing\n",
    "  # ggplot(aes(x = Stage_Group, y = mean_X_train_filtered)) +#change to sum or median, depending on what you're visualizing\n",
    "  # ggplot(aes(x = Stage_Group, y = sum_X_train_filtered)) +#change to sum or median, depending on what you're visualizing\n",
    "  # ggplot(aes(x = IPI_grouped, y = median_X_train_filtered)) +#change to sum or median, depending on what you're visualizing\n",
    "  # ggplot(aes(x = Stage_Group, y = sum_X_train_filtered, colour = Stage_Group)) +\n",
    "  #ggplot(aes(x = as.character(NB_EN_SITES), y = sum_X_train_filtered, colour = as.character(NB_EN_SITES))) +\n",
    "  # geom_boxplot(aes(fill = IPI_grouped), alpha = 0.3) +\n",
    "  geom_boxplot(aes(fill = Stage_Group), alpha = 0.3) +\n",
    "  geom_point(aes(colour = mean_VAF), position = position_jitterdodge(0.2), size = 7, alpha = 0.6) + \n",
    "  # geom_label(label = X_train_attributes$PATIENT_ID, alpha = 0.3) + # to label the data points and figure out which samples are the outliers\n",
    "  theme_minimal(base_size = 18) +\n",
    "  #scale_color_gradient(low = \"#185a9d\", high = \"#43cea2\") +\n",
    "  scale_color_gradient(low = \"blue\", high = \"red\") +\n",
    "  labs(#x = \"Stage Group\",\n",
    "       x = \"IPI Group\",\n",
    "       # y = \"Median of Hypermethylated Feature Expression\",\n",
    "       # y = \"Median of ALL Hypermethylated Feature Expression\",\n",
    "        y = \"Median of Modeled Hypermethylated Feature Expression\",\n",
    "        colour = \"mean VAF\", \n",
    "        # fill = \"Stage Group\", \n",
    "        fill = \"IPI Group\", \n",
    "       #colour = \"IPI\"\n",
    "       )\n",
    "  #ggtitle(\"Median Hypermethylated Counts of Differential Features\")\n",
    "  #ggtitle(\"Median of Counts over Features\")\n",
    "  #ggtitle(\"Sum of Counts over Modeled Features\")\n",
    "  ggtitle(\"Median of Counts over Modeled Features\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea13662b-c3b2-4f58-919a-eb99edb0d77d",
   "metadata": {},
   "source": [
    "### 23-09-2023 Boxplot visualizations for DLBCL and FL separate. RUN ABOVE CHUNK"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfe1c16b-e176-444f-866a-7bb0234e90e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset X_train_attributes to only DLBCL/FL cases\n",
    "X_train_attributes_subtype <- X_train_attributes[!(X_train_attributes$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY == \"FL\"), ] # for DLBCL\n",
    "X_train_attributes_subtype <- X_train_attributes[!(X_train_attributes$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY == \"DLBCL\"), ] # for FL\n",
    "\n",
    "\n",
    "p <- X_train_attributes_subtype %>% \n",
    "  ggplot(aes(x = Stage_Group, y = median_X_train_filtered)) +#change to sum or median, depending on what you're visualizing\n",
    "  #ggplot(aes(x = Stage_Group, y = sum_X_train_filtered, colour = Stage_Group)) +\n",
    "  #ggplot(aes(x = as.character(NB_EN_SITES), y = sum_X_train_filtered, colour = as.character(NB_EN_SITES))) +\n",
    "  geom_boxplot(aes(fill = Stage_Group), alpha = 0.3) +\n",
    "  geom_point(aes(colour = mean_VAF), position = position_jitterdodge(0.2), size = 7, alpha = 0.6) + \n",
    "  theme_minimal(base_size = 18) +\n",
    "  #scale_color_gradient(low = \"#185a9d\", high = \"#43cea2\") +\n",
    "  scale_color_gradient(low = \"blue\", high = \"red\") +\n",
    "  labs(x = \"Stage Group\",\n",
    "       #x = \"IPI\",\n",
    "       # y = \"Median of Hypermethylated Feature Expression\",\n",
    "       # y = \"Median of ALL Hypermethylated Feature Expression\",\n",
    "        y = \"Median of MODELED Hypermethylated Feature Expression\",\n",
    "        colour = \"Stage Group\", \n",
    "       #colour = \"IPI\"\n",
    "       )\n",
    "  #ggtitle(\"Median Hypermethylated Counts of Differential Features\")\n",
    "  #ggtitle(\"Median of Counts over Features\")\n",
    "  #ggtitle(\"Sum of Counts over Modeled Features\")\n",
    "  ggtitle(\"Median of Counts over Modeled Features\")\n",
    "print(p)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0804b7f0-bde2-419a-858d-a2b32c06b80e",
   "metadata": {},
   "source": [
    "- Directionality of the trend over modeled and random features changes depending on whether the sum or median of the features is used. This is an unreliable metric.\n",
    "- Think about a better way to visualize these features."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "092048f4-378b-439e-989a-026164f7b16a",
   "metadata": {},
   "source": [
    "### Stats for comparison between Stage Groups vs healthy and IPI vs healthy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9742f0b6-b36e-4cc8-aacc-e38f0f79aed5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# t-test between the different stage groups and healthy, for hypermethylated features\n",
    "#stats; stage\n",
    "stats_healthy <- X_train_attributes[X_train_attributes$Stage_Group == \"non-lymphoma\",]\n",
    "stats_early <- X_train_attributes[X_train_attributes$Stage_Group == \"Early Stage\",]\n",
    "stats_late <- X_train_attributes[X_train_attributes$Stage_Group == \"Late Stage\",]\n",
    "\n",
    "# sum (less significant (due to the size of the values), but cleaner trend)\n",
    "# t.test(stats_healthy$sum_X_train_filtered, stats_early$sum_X_train_filtered, var.equal = FALSE) #significant; p = 4.322e-09\n",
    "# t.test(stats_healthy$sum_X_train_filtered, stats_late$sum_X_train_filtered, var.equal = FALSE) #significant; p = 1.319e-09\n",
    "# t.test(stats_early$sum_X_train_filtered, stats_late$sum_X_train_filtered, var.equal = FALSE) # not significant; p = 0.07981\n",
    "\n",
    "# median (more significant)\n",
    "t.test(stats_healthy$median_X_train_filtered, stats_early$median_X_train_filtered, var.equal = FALSE) #significant; p = 3.958e-09\n",
    "t.test(stats_healthy$median_X_train_filtered, stats_late$median_X_train_filtered, var.equal = FALSE) #significant; p = 8.591e-13\n",
    "t.test(stats_early$median_X_train_filtered, stats_late$median_X_train_filtered, var.equal = FALSE) #significant; p = 0.01535\n",
    "\n",
    "# sum\n",
    "t.test(stats_healthy$sum_X_train_filtered, stats_early$sum_X_train_filtered, var.equal = FALSE) #significant; p = 1.218e-11\n",
    "t.test(stats_healthy$sum_X_train_filtered, stats_late$sum_X_train_filtered, var.equal = FALSE) #significant; p = < 2.2e-16\n",
    "t.test(stats_early$sum_X_train_filtered, stats_late$sum_X_train_filtered, var.equal = FALSE) #significant; p = 0.4273\n",
    "\n",
    "# median, ALL HYPERMETHYLATED FEATURES, n = 129380\n",
    "t.test(stats_healthy$median_X_train_filtered, stats_early$median_X_train_filtered, var.equal = FALSE) #significant; p = 0.001867\n",
    "t.test(stats_healthy$median_X_train_filtered, stats_late$median_X_train_filtered, var.equal = FALSE) #significant; p = 1.226e-06\n",
    "\n",
    "# median, ALL SIGNIFICANT HYPERMETHYLATED FEATURES, n = 5328\n",
    "t.test(stats_healthy$median_X_train_filtered, stats_early$median_X_train_filtered, var.equal = FALSE) #significant; p = 5.592e-10\n",
    "t.test(stats_healthy$median_X_train_filtered, stats_late$median_X_train_filtered, var.equal = FALSE) #significant; p = 1.493e-13\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#stats; IPI\n",
    "stats_IPI_0 <- X_train_attributes[X_train_attributes$IPI == \"0\",]\n",
    "stats_IPI_1 <- X_train_attributes[X_train_attributes$IPI == \"1\",]\n",
    "stats_IPI_2 <- X_train_attributes[X_train_attributes$IPI == \"2\",]\n",
    "stats_IPI_3 <- X_train_attributes[X_train_attributes$IPI == \"3\",]\n",
    "stats_IPI_4 <- X_train_attributes[X_train_attributes$IPI == \"4\",]\n",
    "stats_IPI_5 <- X_train_attributes[X_train_attributes$IPI == \"5\",]\n",
    "stats_IPI_healthy <- X_train_attributes[X_train_attributes$Stage_Group == \"non-lymphoma\",]\n",
    "\n",
    "# sum\n",
    "t.test(stats_IPI_healthy$sum_X_train_filtered, stats_IPI_1$sum_X_train_filtered, var.equal = FALSE) #significant; p = 5.253e-08\n",
    "t.test(stats_IPI_healthy$sum_X_train_filtered, stats_IPI_2$sum_X_train_filtered, var.equal = FALSE) #significant; p = 2.533e-08\n",
    "t.test(stats_IPI_healthy$sum_X_train_filtered, stats_IPI_3$sum_X_train_filtered, var.equal = FALSE) #significant; p = 1.114e-08\n",
    "t.test(stats_IPI_healthy$sum_X_train_filtered, stats_IPI_4$sum_X_train_filtered, var.equal = FALSE) #significant; p = 3.851e-08\n",
    "t.test(stats_IPI_healthy$sum_X_train_filtered, stats_IPI_5$sum_X_train_filtered, var.equal = FALSE) #not significant; p = 0.1411\n",
    "\n",
    "# median; MODELED FEATURES\n",
    "t.test(stats_IPI_healthy$median_X_train_filtered, stats_IPI_0$median_X_train_filtered, var.equal = FALSE) # significant; p = 2.27e-07\n",
    "t.test(stats_IPI_healthy$median_X_train_filtered, stats_IPI_1$median_X_train_filtered, var.equal = FALSE) # significant; p = 0.0006568\n",
    "t.test(stats_IPI_healthy$median_X_train_filtered, stats_IPI_2$median_X_train_filtered, var.equal = FALSE) # significant; p = 5.456e-07\n",
    "t.test(stats_IPI_healthy$median_X_train_filtered, stats_IPI_3$median_X_train_filtered, var.equal = FALSE) # significant; p = 3.037e-06\n",
    "t.test(stats_IPI_healthy$median_X_train_filtered, stats_IPI_4$median_X_train_filtered, var.equal = FALSE) # significant; p = 0.0002235\n",
    "t.test(stats_IPI_healthy$median_X_train_filtered, stats_IPI_5$median_X_train_filtered, var.equal = FALSE) # not significant; p = 0.1051\n",
    "\n",
    "# median\n",
    "t.test(stats_IPI_healthy$median_X_train_filtered, stats_IPI_0$median_X_train_filtered, var.equal = FALSE) # significant; p = 4.951e-08\n",
    "t.test(stats_IPI_healthy$median_X_train_filtered, stats_IPI_1$median_X_train_filtered, var.equal = FALSE) # significant; p = 0.001438\n",
    "t.test(stats_IPI_healthy$median_X_train_filtered, stats_IPI_2$median_X_train_filtered, var.equal = FALSE) # significant; p = 3.861e-08\n",
    "t.test(stats_IPI_healthy$median_X_train_filtered, stats_IPI_3$median_X_train_filtered, var.equal = FALSE) # significant; p = 5.936e-07\n",
    "t.test(stats_IPI_healthy$median_X_train_filtered, stats_IPI_4$median_X_train_filtered, var.equal = FALSE) # significant; p = 0.0001509\n",
    "t.test(stats_IPI_healthy$median_X_train_filtered, stats_IPI_5$median_X_train_filtered, var.equal = FALSE) # not significant; p = 0.0865\n",
    "\n",
    "\n",
    "# FOR RANDOM SAMPLE\n",
    "#stats; IPI\n",
    "stats_IPI_0 <- X_train_attributes_subset[X_train_attributes_subset$IPI == \"0\",]\n",
    "stats_IPI_1 <- X_train_attributes_subset[X_train_attributes_subset$IPI == \"1\",]\n",
    "stats_IPI_2 <- X_train_attributes_subset[X_train_attributes_subset$IPI == \"2\",]\n",
    "stats_IPI_3 <- X_train_attributes_subset[X_train_attributes_subset$IPI == \"3\",]\n",
    "stats_IPI_4 <- X_train_attributes_subset[X_train_attributes_subset$IPI == \"4\",]\n",
    "stats_IPI_5 <- X_train_attributes_subset[X_train_attributes_subset$IPI == \"5\",]\n",
    "stats_IPI_healthy <- X_train_attributes_subset[X_train_attributes_subset$Stage_Group == \"healthy\",]\n",
    "\n",
    "# median\n",
    "t.test(stats_IPI_healthy$Mean_Median, stats_IPI_0$Mean_Median, var.equal = FALSE) # significant; p = 4.951e-08\n",
    "t.test(stats_IPI_healthy$Mean_Median, stats_IPI_1$Mean_Median, var.equal = FALSE) # significant; p = 0.001438\n",
    "t.test(stats_IPI_healthy$Mean_Median, stats_IPI_2$Mean_Median, var.equal = FALSE) # significant; p = 3.861e-08\n",
    "t.test(stats_IPI_healthy$Mean_Median, stats_IPI_3$Mean_Median, var.equal = FALSE) # significant; p = 5.936e-07\n",
    "t.test(stats_IPI_healthy$Mean_Median, stats_IPI_4$Mean_Median, var.equal = FALSE) # significant; p = 0.0001509\n",
    "t.test(stats_IPI_healthy$Mean_Median, stats_IPI_5$Mean_Median, var.equal = FALSE) # not significant; p = 0.0865"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "243fcbd9-c009-48f8-b988-47fbb130601c",
   "metadata": {},
   "source": [
    "- Makes sense that the trends are not significant, as the directionality of the scaled values is in both directions; need a better way to look at this.\n",
    "- Consider doing this analysis again but for only modeled features which have a positive logFC in the volcano plot.\n",
    "- If this doesn't work, stick with the median because there is a significant difference between an IPI of 1 and 5 across the modeled features."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccb0439b-a4e5-4879-bdf5-756b6a8ede73",
   "metadata": {},
   "source": [
    "### ANOVA across correlation coefficients for the different IPI groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3bb52f3-d6c5-40a3-b62c-77120db60f5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create matrix of samples from each IPI group\n",
    "matrix_data <- X_train_attributes[,c(\"IPI_grouped\",\"median_X_train_filtered\")] # for grouped IPI\n",
    "# matrix_data <- X_train_attributes[,c(\"IPI\",\"median_X_train_filtered\")]\n",
    "\n",
    "# matrix_data_lymphoma <- matrix_data[!(matrix_data$IPI == \"NA\"),] # for individual IPI\n",
    "matrix_data_lymphoma <- matrix_data[!(matrix_data$IPI_grouped == \"NA\"),] # for grouped IPI\n",
    "# matrix_data_lymphoma_no_IPI5 <- matrix_data[!(matrix_data$IPI == \"NA\" | matrix_data$IPI == \"5\"),]\n",
    "\n",
    "# ANOVA\n",
    "result_anova <- aov(median_X_train_filtered ~ IPI, data = matrix_data)\n",
    "result_anova <- aov(median_X_train_filtered ~ IPI, data = matrix_data_lymphoma)\n",
    "result_anova <- aov(median_X_train_filtered ~ IPI, data = matrix_data_lymphoma_no_IPI5)\n",
    "summary(result_anova)\n",
    "\n",
    "# alternative; kruskal.test (non-parametric)\n",
    "# kruskal.test(median_X_train_filtered ~ IPI, data = matrix_data_lymphoma)\n",
    "kruskal.test(median_X_train_filtered ~ IPI_grouped, data = matrix_data_lymphoma)\n",
    "# kruskal.test(median_X_train_filtered ~ IPI, data = matrix_data_lymphoma_no_IPI5)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# EXTRA\n",
    "# # Kruskal-Wallis (doesn't work)\n",
    "# matrix_data_lymphoma$IPI <- as_factor(matrix_data_lymphoma$IPI)\n",
    "# result_kw <- kruskal.test(median_X_train_filtered ~ IPI, data = matrix_data_lymphoma)\n",
    "# summary(result_kw)\n",
    "\n",
    "# Fisher's Exact Test\n",
    "# Create a table for the contingency data\n",
    "contingency_table <- table(df$IPI, df$median_X_train_filtered)\n",
    "\n",
    "# Perform Fisher's Exact Test for each IPI group against all others\n",
    "results <- lapply(unique(df$IPI), function(ipi_group) {\n",
    "  # Create a 2x2 contingency table for the current IPI group vs. all others\n",
    "  current_group <- contingency_table[ipi_group, ]\n",
    "  other_groups <- rowSums(contingency_table[-ipi_group, ])\n",
    "  contingency <- matrix(c(current_group, other_groups), nrow = 2)\n",
    "  \n",
    "  # Perform Fisher's Exact Test\n",
    "  fisher_result <- fisher.test(contingency)\n",
    "  \n",
    "  # Store the result along with the IPI group name\n",
    "  result <- list(IPI_Group = ipi_group, Fisher_Test_Result = fisher_result)\n",
    "  \n",
    "  return(result)\n",
    "})\n",
    "\n",
    "# Combine results into a data frame\n",
    "results_df <- do.call(rbind, results)\n",
    "\n",
    "# Print the results\n",
    "print(results_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86653ec0-f037-4a51-8e52-188cb50de43d",
   "metadata": {},
   "source": [
    "### Subset to features where a a positive logFC was observed for differential analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6510fdae-cb16-4d9d-9851-c9e0482ff2bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# subset to hypomethylated features from the volcano plot, measure sum and median of these features and see whether the \n",
    "# trends improve\n",
    "\n",
    "# SUBSET TO PREDICTORS WITH A POSITIVE logFC\n",
    "predictors_rf_hyper <- res[rownames(res) %in% predictors_rf & res$table$logFC > 0,]\n",
    "\n",
    "# advance to first chunk in this heading\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f5ffcf9-3c89-4bb2-b62a-70dcfcffba4e",
   "metadata": {},
   "source": [
    "### Random sampling of X number of features over 1000 iterations, and visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1548ce9-3b15-4dfe-920c-1301d72c29b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# sub chunk of previous chunk, take 1000 random samples of the values from the matrix and visualize the sum of these. Compare to the modeled features.\n",
    "\n",
    "# Initialize an empty list to store the results\n",
    "results_list <- list()\n",
    "\n",
    "for (iteration in 1:1000) {\n",
    "  # Randomly select X number of features; adjust according to the number of desired features\n",
    "  #selected <- sample(rownames(X_train), 5056, replace = FALSE)\n",
    "  selected <- sample(rownames(X_train), 2850, replace = FALSE)\n",
    "  # selected <- sample(rownames(X_train), 5328, replace = FALSE)\n",
    "\n",
    "  # Subset the matrix\n",
    "  subset_matrix <- cpm_norm_log2[rownames(cpm_norm_log2) %in% selected,]\n",
    "\n",
    "  # Add the lowest scaled count in the matrix to each of the counts to remove zeros\n",
    "  # test <- min(subset_matrix)\n",
    "  # subset_matrix <- subset_matrix + abs(test) + 0.001\n",
    "\n",
    "  # Calculate the median of each column of the random subset of features\n",
    "  median_X_train_filtered_subset <- apply(subset_matrix, \n",
    "                                          MARGIN = 2, \n",
    "                                          FUN = median)\n",
    "\n",
    "  # Include the sums of the features per sample as an additional feature\n",
    "  sum_X_train_filtered_subset <- colSums(subset_matrix)\n",
    "\n",
    "  # Create a data frame for the results of this iteration\n",
    "  iteration_results <- data.frame(\n",
    "    Median = median_X_train_filtered_subset,\n",
    "    Sum = sum_X_train_filtered_subset\n",
    "  )\n",
    "\n",
    "  # Append the results to the list\n",
    "  results_list[[iteration]] <- iteration_results\n",
    "}\n",
    "\n",
    "# Combine the results from all iterations into a single data frame\n",
    "combined_results <- bind_rows(results_list, .id = \"Iteration\")\n",
    "\n",
    "# make rowname a column in combined_results\n",
    "combined_results <- combined_results %>%\n",
    "  mutate(SampleName = str_extract(rownames(combined_results), \".*K27me3\"))\n",
    "\n",
    "# Calculate the mean of each observation in the \"Sum\" column across all iterations\n",
    "mean_values <- combined_results %>%\n",
    "  group_by(SampleName) %>%\n",
    "  summarise(Mean_Median = mean(Median),\n",
    "            Mean_Sum = mean(Sum))\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## append new values to metadata for visualization\n",
    "# subset sample metadata to training samples, then visualize expression over the random features\n",
    "X_train_attributes_subset <- sample_list[sample_list$sample_name %in% colnames(cpm_norm),]\n",
    "\n",
    "#bind the newly calculated median / sum to the pre-existing metadata\n",
    "X_train_attributes_subset <- as.data.frame(cbind(X_train_attributes_subset, mean_values)) #Subset this data frame for plotting.\n",
    "\n",
    "# normalize data by cfDNA concentrations. Once data is normalized visualize variance, distribution of expression levels, etc.\n",
    "X_train_attributes_subset <- X_train_attributes_subset %>%\n",
    "mutate(median_X_train_filtered_subset_cfDNA_adj = median_X_train_filtered_subset / clinical_metadata$`Plasma cfDNA concentration (ng/uL)`)\n",
    "\n",
    "X_train_attributes_subset$ANN_ARBOR_STAGE[is.na(X_train_attributes_subset$ANN_ARBOR_STAGE)] <- \"healthy\" #make NA = healthy for Ann Arbor stage.\n",
    "X_train_attributes_subset$IPI <- as.character(X_train_attributes_subset$IPI) #convert IPI from int to char\n",
    "\n",
    "#For tumor diameter, create bins (<5cm, 5-10cm, 10-15cm, >15cm), then assign samples to those bins based off of tumor diameter; separate column\n",
    "X_train_attributes_subset <- X_train_attributes_subset %>% \n",
    "  mutate(tumor_diameter_binned = cut(TUMOR_MASS_MAX_CM, breaks = c(0, 5, 10, 15, 30), labels = c(\"<5\", \"5-10\", \"10-15\", \">15\")))\n",
    "\n",
    "# combine Stage I/II and III/IV samples into early and late stage, respectively, for visualization\n",
    "X_train_attributes_subset <- X_train_attributes_subset %>%\n",
    "  mutate(Stage_Group = case_when(\n",
    "    ANN_ARBOR_STAGE %in% c(1, 2) ~ \"Early Stage\",\n",
    "    ANN_ARBOR_STAGE %in% c(3, 4) ~ \"Late Stage\",\n",
    "    TRUE ~ as.character(ANN_ARBOR_STAGE)  # Keep other values as is\n",
    "  ))\n",
    "\n",
    "# set factor level for visualization\n",
    "X_train_attributes_subset$Stage_Group <- factor(X_train_attributes_subset$Stage_Group, levels = c(\"healthy\", \"Early Stage\", \"Late Stage\"))\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## RUN IF RANDOM SUBSET WAS ALREADY GENERATED, import\n",
    "setwd(\"/cluster/projects/scottgroup/people/steven/human_plasma_cfChIP-Seq/liberate_plasma_cfchip/analysis/machine_learning_models/02_07_2023_RFE_RF_10kb_windows/differential_analysis_filter/r_objects\")\n",
    "load(\"26_09_2023_1000_iterations_random_subset_2850_features_after_scaling.RData\")\n",
    "\n",
    "# add mean VAF to this data frame from clinical_metadata. Sample order is the same, so cbind is sufficient.\n",
    "mean_VAF <- clinical_metadata$mean_VAF\n",
    "X_train_attributes_subset <- cbind(X_train_attributes_subset, mean_VAF)\n",
    "\n",
    "p <- X_train_attributes_subset %>% \n",
    "  ggplot(aes(x = Stage_Group, y = Mean_Median)) +#change to sum or median, depending on what you're visualizing\n",
    "  #ggplot(aes(x = Stage_Group, y = sum_X_train_filtered, colour = Stage_Group)) +\n",
    "  #ggplot(aes(x = as.character(NB_EN_SITES), y = sum_X_train_filtered, colour = as.character(NB_EN_SITES))) +\n",
    "  geom_boxplot(aes(fill = Stage_Group), alpha = 0.3) +\n",
    "  geom_point(aes(colour = mean_VAF), position = position_jitterdodge(0.2), size = 7, alpha = 0.6) + \n",
    "  theme_minimal(base_size = 18) +\n",
    "  #scale_color_gradient(low = \"#185a9d\", high = \"#43cea2\") +\n",
    "  scale_color_gradient(low = \"blue\", high = \"red\") +\n",
    "  labs(x = \"Stage Group\",\n",
    "       # x = \"IPI\",\n",
    "       y = \"mean of 1000 iterations(sum(Feature Counts of 2850 Random Features))\",\n",
    "       # y = \"mean of 1000 iterations(median(Feature Counts of 5328 Random Features))\",\n",
    "       colour = \"mean VAF\", \n",
    "       fill = \"Stage Group\", \n",
    "       # colour = \"IPI\"\n",
    "       )\n",
    "  #ggtitle(\"Median Hypermethylated Counts of Differential Features\")\n",
    "  #ggtitle(\"Median of Counts over Features\")\n",
    "  #ggtitle(\"Sum of Counts over Modeled Features\")\n",
    "  ggtitle(\"Median of Counts over Modeled Features\")\n",
    "print(p)\n",
    "\n",
    "# save this iteration as an R Object for visualization later, along with the individual values for mean and median.\n",
    "setwd(\"/cluster/projects/scottgroup/people/steven/human_plasma_cfChIP-Seq/liberate_plasma_cfchip/analysis/machine_learning_models/02_07_2023_RFE_RF_10kb_windows/differential_analysis_filter/r_objects\")\n",
    "save(X_train_attributes_subset, file = \"26_09_2023_1000_iterations_random_subset_2850_features_after_scaling.RData\")\n",
    "save(combined_results, file = \"26_09_2023_1000_iterations_random_subset_2850_features_after_scaling_rawdata.RData\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aef32562-1253-4ed1-8d66-225d4e072f2d",
   "metadata": {},
   "source": [
    "### Incorporate expression of K27me3 over modeled features in T3 (post-treatment) samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "381a0976-90c9-4daf-8f01-41f1d984978d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# incorporate T3 samples into the numeric matrix\n",
    "# see introductory chunk in this document for details\n",
    "# for this to work, you need to normalize all the data in the same way, and perform the same train/test split for visualization.\n",
    "# CPM normalization and scaling to be performed on all values as well.\n",
    "# scaling is applied by column, where the mean and SD is calculated per column.\n",
    "# Same with CPM normalization. This is good, so you can apply the two primary normalizations this way.\n",
    "\n",
    "# what you can do is 1) create the matrix w all samples, 2) subset to all features in original data matrix \n",
    "# after pre-processing steps, and 3) apply the two normalizations before subsetting to the hypermethylated \n",
    "# modeled features (since the normalizations are applied column-wise).\n",
    "\n",
    "\n",
    "# load original dataset from modeling, for reference and subsetting to relevant samples\n",
    "# setwd(\"/cluster/projects/scottgroup/people/steven/human_plasma_cfChIP-Seq/liberate_plasma_cfchip/analysis/machine_learning_models/02_07_2023_RFE_RF_10kb_windows/differential_analysis_filter/\")\n",
    "# load(\"10_09_2023_rf_10kb_with_hyperparameter_tuning_accuracyevaluation_withdifferentialfilter.RData\")\n",
    "\n",
    "\n",
    "#Novaigate to directory with new annotated list. Import list to R session with differential features.\n",
    "setwd(\"/cluster/projects/scottgroup/people/steven/reference_sets/other_references\")\n",
    "annotated_ref_no_X_Y_M_blacklist <- read.delim(\"10kb_bin_genome_wide_annotated_reference_with_regulatory_features_no_X_Y_M_ENCODEBlacklist.txt\", header = TRUE, sep = \"\\t\")\n",
    "#annotated_ref_no_X_Y_M_blacklist <- read.delim(\"100kb_bin_genome_wide_annotated_reference_with_regulatory_features_no_X_Y_M_ENCODEBlacklist.txt\", header = TRUE, sep = \"\\t\") # for 100kb\n",
    "\n",
    "#CREATE SAMPLE MATRIX IN THE FORM OF A DGEList froM edgeR#\n",
    "setwd(\"/cluster/home/sdemichi/steven_scottgroup.txt/human_plasma_cfChIP-Seq/liberate_plasma_cfchip/temp/\")\n",
    "sample_list <- read.delim(\"15_05_2023_sample_list_edgeR_without_basic_exclusions_including_T3_outliers_T5_no_pbmc.txt\", header = TRUE, sep = \"\\t\") #without T3\n",
    "\n",
    "gc() #clear memory; these data frames are large\n",
    "setwd(\"/cluster/home/sdemichi/steven_scottgroup.txt/human_plasma_cfChIP-Seq/liberate_plasma_cfchip/\")\n",
    "import_counts <- list.files(path = \".\", pattern = \"counts_10kb.txt$\", recursive = TRUE) #list counts data, 10kb windows\n",
    "exclude <- c(\"0018_T3\", \"0025_T4\", \"0027_T3\", \"0027_T4\", \"0101_T3\", \"0030_T1\", \"HUCON_37\", \"HUCON_43\", \"HUCON_44\", \"0036_C\", \"0247_T1_ABC\", \"pbmc_k27\") #character vector of sample strings to be excluded.\n",
    "\n",
    "# look into why you excluded \"0036_C\" and \"0247_T1_ABC\" from the prior analysis. It works without them which is good.\n",
    "# for now, remove them from the analysis\n",
    "sl_exclude <- c(\"liberate_0036_C_K27me3\", \"liberate_0247_T1_ABC_K27me3\")\n",
    "sample_list <- sample_list[!(sample_list$sample_name) %in% sl_exclude,]\n",
    "\n",
    "#Remove all excluded samples from the original list of samples; these are samples without baseline profiles, and T5. Only T1 and T3 should be included.\n",
    "import_counts <- as.data.frame(import_counts)\n",
    "import_counts <- import_counts[!grepl(paste(exclude, collapse = \"|\"), import_counts$import_counts),] #this will be a helpful line for the future.\n",
    "#import_counts <- as.data.frame(import_counts) #convert back to data frame\n",
    "\n",
    "counts <- mapply(read.delim, import_counts) #loop over files in import_counts, and read them into memory\n",
    "counts_matrix <- do.call(cbind, counts) #bind individual datasets across columns\n",
    "\n",
    "colnames(counts_matrix) <- c(paste0(sample_list$sample_name)) #change column names to file names from list\n",
    "\n",
    "# index rows in counts_matrix\n",
    "rownames(counts_matrix) <- 1:nrow(counts_matrix)\n",
    "\n",
    "#Store data in a DGEList object\n",
    "gc()\n",
    "\n",
    "\n",
    "## code from here\n",
    "# subset to only features present in original dataset after filtering\n",
    "all_filtered_features_including_T3 <- counts_matrix[rownames(counts_matrix) %in% rownames(all_filtered_features),]\n",
    "\n",
    "# in the numeric matrix, keep only samples that are in training set for T1, and only matched samples for T3; no test T3\n",
    "# should be included\n",
    "# subset sample_list by PATIENT_ID, then subset the matrix by the updated sample_list\n",
    "sample_list <- sample_list[sample_list$PATIENT_ID %in% X_train_attributes$PATIENT_ID,]\n",
    "all_filtered_features_including_T3 <- all_filtered_features_including_T3[,colnames(all_filtered_features_including_T3) %in% sample_list$sample_name]\n",
    "\n",
    "# log2(CPM+1) normalize matrix\n",
    "# CPM normalize\n",
    "all_filtered_features_including_T3 <- cpm(all_filtered_features_including_T3, normalized.lib.sizes = TRUE, log = FALSE)\n",
    "\n",
    "# log(CPM+1) transform\n",
    "all_filtered_features_including_T3 <- log2(all_filtered_features_including_T3 + 1)\n",
    "\n",
    "#X_train_filtered <- cpm_norm_scaled[(rownames(cpm_norm_scaled)) %in% predictors_rf,]\n",
    "T3_modeled_features <- all_filtered_features_including_T3[(rownames(all_filtered_features_including_T3)) %in% predictors_rf_hyper,] # for hypermethylated features\n",
    "\n",
    "\n",
    "\n",
    "# ## SANITY CHECK; check that the counts over modeled features in T3_modeled_features, for samples shared with X_train_filtered\n",
    "# X_train_filtered <- cpm_norm_scaled[(rownames(cpm_norm_scaled)) %in% predictors_rf_hyper,]\n",
    "\n",
    "# test <- T3_modeled_features[,colnames(T3_modeled_features) %in% colnames(X_train_filtered)]\n",
    "# identical(test, X_train_filtered) # there are differences. FALSE produced.\n",
    "\n",
    "# # find values where differences occur between the two matrices.\n",
    "# differences <- test - X_train_filtered # almost every value is different. Why? They should be the same since they were processed in the same way. There's a problem somewhere.\n",
    "\n",
    "# identical(rownames(T3_modeled_features), rownames(T3_modeled_features)) # correct feature subset\n",
    "# identical(colnames(T3_modeled_features), colnames(T3_modeled_features)) # correct sample subset\n",
    "\n",
    "# identical(all_filtered_features_including_T3, tmm_norm$counts) # correct set of features, and feature expression before CPM normalization\n",
    "# identical(all_filtered_features_including_T3, cpm_norm_scaled) # not the same, scaling may be different. But the matrices do look the same using head().\n",
    "\n",
    "# identical(T3_modeled_features, X_train_filtered) # FALSE. Changes occuring during the scaling step, I believe.\n",
    "\n",
    "# differences <- T3_modeled_features - X_train_filtered # the differences are at a rounding level, there are essentially no differences between the two matrices.\n",
    "\n",
    "# max(differences) #1.4210854715202e-14, basically negligable.\n",
    "\n",
    "# confirmed, difference has to be the addition of min to the matrix.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# import clinical metadata\n",
    "setwd(\"/cluster/home/sdemichi/steven_scottgroup.txt/human_plasma_cfChIP-Seq/liberate_plasma_cfchip/temp/\")\n",
    "clinical_metadata <- as.data.frame(readxl::read_xlsx(\"25_07_2023_cohort_metadata_labeling_UPDATED_WITH_CLINICAL_ANNOTATIONS_plus_library_size_cfdna_concn.xlsx\"))\n",
    "\n",
    "# import mean VAF information\n",
    "setwd(\"/cluster/projects/scottgroup/people/steven/human_plasma_cfChIP-Seq/liberate_plasma_cfchip/other_profiles/cappseq_data_lymphoma/updated_batch_2_all_cases/\")\n",
    "mean_VAF_burden <- read.delim(\"Bratman.dcs.uniq.burden.txt\", header = TRUE, sep = \"\\t\") # for dcs.unique\n",
    "\n",
    "# replace spaces in \"Donor ID (rep)\" so that you can match directly to mean VAF burden$Sample_id\n",
    "clinical_metadata$`Donor ID (rep)` <- gsub(\" \", \"_\", clinical_metadata$`Donor ID (rep)`)\n",
    "\n",
    "# match mean VAF information to clinical metadata. Hopefully some of the T3 burden data comes back soon from OICR.\n",
    "clinical_metadata$mean_VAF <- mean_VAF_burden$mean_VAF[match(clinical_metadata$'Donor ID (rep)', mean_VAF_burden$Sample_id)]\n",
    "\n",
    "# match annotations from X_train to the new clinical annotations. Append cfDNA concentrations.\n",
    "#all_filtered_features_X_train <- all_filtered_features[,colnames(all_filtered_features) %in% colnames(X_train)]\n",
    "\n",
    "\n",
    "## SUBSET TO ONLY SAMPLES IN TRAINING SET; continue from here\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# add additional column to sample list, which matches the format of clinical_metadata$`Bioinformatic Name`\n",
    "sample_list <- sample_list %>%\n",
    "mutate(`Bioinformatic Name` = sub(\"^liberate_(.*)_K27me3$\", \"\\\\1\", sample_list$sample_name))\n",
    "\n",
    "# subset sample metadata to training samples. #### TROUBLESHOOT; ONE OF THESE STEPS IS CREATING DUPLICATE ENTRIES IN clinical_metadata$`Bioinformatic Name`\n",
    "training_sample_metadata <- sample_list[sample_list$sample_name %in% colnames(T3_modeled_features),]\n",
    "\n",
    "# filter clinical metadata by append cfDNA concentration to test, from clinical metadata.\n",
    "clinical_metadata <- clinical_metadata[clinical_metadata$`Bioinformatic Name` %in% training_sample_metadata$`Bioinformatic Name`,] \n",
    "\n",
    "# subset sample metadata to training samples\n",
    "X_train_attributes_plus_T3 <- sample_list[sample_list$sample_name %in% colnames(all_filtered_features_including_T3),]\n",
    "\n",
    "# sort clinical_metadata in order of colnames of X_train_attributes$PATIENT_ID\n",
    "clinical_metadata <- clinical_metadata[match(X_train_attributes_plus_T3$`Bioinformatic Name`, clinical_metadata$`Bioinformatic Name`), ]\n",
    "\n",
    "# overwrite X_train_attributes with clinical_metadata\n",
    "X_train_attributes_plus_T3 <- clinical_metadata\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#calculate the median of each column (aka counts over top 1500 differential features, for each sample)\n",
    "median_X_train_filtered_plus_T3 <- apply(T3_modeled_features, \n",
    "                   MARGIN = 2, \n",
    "                   FUN = median)\n",
    "\n",
    "#include the sums of the features per sample as an additional feature, and visualize\n",
    "sum_X_train_filtered_plus_T3 <- colSums(T3_modeled_features)\n",
    "\n",
    "#bind the newly calculated medians and sumsto the pre-existing metadata\n",
    "X_train_attributes_plus_T3 <- as.data.frame(cbind(X_train_attributes_plus_T3, median_X_train_filtered_plus_T3)) #Subset this data frame for plotting.\n",
    "X_train_attributes_plus_T3 <- as.data.frame(cbind(X_train_attributes_plus_T3, sum_X_train_filtered_plus_T3)) #Subset this data frame for plotting.\n",
    "\n",
    "##### METADATA-SPECIFIC CHANGES BEFORE PLOTTING #####\n",
    "X_train_attributes_plus_T3$ANN_ARBOR_STAGE <- ifelse(X_train_attributes_plus_T3$ANN_ARBOR_STAGE == \"NA\", \"non-lymphoma\", X_train_attributes_plus_T3$ANN_ARBOR_STAGE) # convert NA characters to \"non-lymphoma\"\n",
    "X_train_attributes_plus_T3$IPI <- as.character(X_train_attributes_plus_T3$IPI) #convert IPI from int to char\n",
    "\n",
    "X_train_attributes_plus_T3$mean_VAF <- as.numeric(X_train_attributes_plus_T3$mean_VAF) #convert IPI from int to char\n",
    "\n",
    "\n",
    "#For tumor diameter, create bins (<5cm, 5-10cm, 10-15cm, >15cm), then assign samples to those bins based off of tumor diameter; separate column\n",
    "X_train_attributes_plus_T3$TUMOR_MASS_MAX_CM <- as.numeric(X_train_attributes_plus_T3$TUMOR_MASS_MAX_CM)\n",
    "X_train_attributes_plus_T3 <- X_train_attributes_plus_T3 %>% \n",
    "  mutate(tumor_diameter_binned = cut(TUMOR_MASS_MAX_CM, breaks = c(0, 5, 10, 15, 30), labels = c(\"<5\", \"5-10\", \"10-15\", \">15\")))\n",
    "\n",
    "# combine Stage I/II and III/IV samples into early and late stage, respectively, for visualization\n",
    "X_train_attributes_plus_T3 <- X_train_attributes_plus_T3 %>%\n",
    "  mutate(Stage_Group = case_when(\n",
    "    ANN_ARBOR_STAGE %in% c(1, 2) ~ \"Early Stage\",\n",
    "    ANN_ARBOR_STAGE %in% c(3, 4) ~ \"Late Stage\",\n",
    "    TRUE ~ as.character(ANN_ARBOR_STAGE)  # Keep other values as is\n",
    "  ))\n",
    "\n",
    "# set factor level for Stage_Group\n",
    "X_train_attributes_plus_T3$Stage_Group <- factor(X_train_attributes_plus_T3$Stage_Group, levels = c(\"non-lymphoma\", \"Early Stage\", \"Late Stage\"))\n",
    "\n",
    "# add another column representing whether the samples are baseline or post-treatment\n",
    "X_train_attributes_plus_T3 <- X_train_attributes_plus_T3 %>%\n",
    "mutate(time_of_collection = case_when(\n",
    "    str_detect(`Donor ID (rep)`, \"HUCON|_C\") ~ \"Non-Lymphoma\",\n",
    "    str_detect(`Donor ID (rep)`, \"T1\") ~ \"Baseline\",\n",
    "    str_detect(`Donor ID (rep)`, \"T3|T5\") ~ \"Post-Treatment\",\n",
    "))\n",
    "\n",
    "# visualize with T3 samples\n",
    "p <- X_train_attributes_plus_T3 %>% \n",
    "  ggplot(aes(x = time_of_collection, y = median_X_train_filtered_plus_T3)) +#change to sum or median, depending on what you're visualizing\n",
    "  #ggplot(aes(x = Stage_Group, y = sum_X_train_filtered, colour = Stage_Group)) +\n",
    "  #ggplot(aes(x = as.character(NB_EN_SITES), y = sum_X_train_filtered, colour = as.character(NB_EN_SITES))) +\n",
    "  geom_boxplot(aes(fill = time_of_collection), alpha = 0.3) +\n",
    "  geom_point(aes(colour = mean_VAF), position = position_jitterdodge(0.2), size = 7, alpha = 0.6) + \n",
    "  theme_minimal(base_size = 18) +\n",
    "  #scale_color_gradient(low = \"#185a9d\", high = \"#43cea2\") +\n",
    "  scale_color_gradient(low = \"blue\", high = \"red\") +\n",
    "  labs(x = \"Stage Group\",\n",
    "       #x = \"IPI\",\n",
    "       #y = \"Sum of Scaled Feature Counts\",\n",
    "       y = \"Median of Hypermethylated Feature Expression\",\n",
    "       colour = \"Stage Group\", \n",
    "       #colour = \"IPI\"\n",
    "       )\n",
    "  #ggtitle(\"Median Hypermethylated Counts of Differential Features\")\n",
    "  #ggtitle(\"Median of Counts over Features\")\n",
    "  #ggtitle(\"Sum of Counts over Modeled Features\")\n",
    "  ggtitle(\"Median of Counts over Modeled Features\")\n",
    "print(p)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "077c8d4b-450e-47ac-a7e3-c0a40a02a0fa",
   "metadata": {},
   "source": [
    "- 29-09-2023 work in progress. Figured out how to subset the matrix to all training T1 and associated T3 samples.\n",
    "- Next step is to perform all the boxplot transformations as performed before, and visualizing non-lymphoma, T1 and T3.\n",
    "- 04-01-2024; worked fine with log2(CPM+1).\n",
    "- NEED TO SUBSET TO ONLY TRAINING SET CASES!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9162231e-6fec-45b9-aac1-e6e1baf7e049",
   "metadata": {},
   "source": [
    "### 12-10-2023 Feature Importance for Modeled Features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23bd79e2-0091-46b6-a220-74feaee64cc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# use the VarImp function to measure the importance of each of the variables on classification performance.\n",
    "# see if there is a subset of features that are particularly important.\n",
    "\n",
    "# import modeling data\n",
    "setwd(\"/cluster/projects/scottgroup/people/steven/human_plasma_cfChIP-Seq/liberate_plasma_cfchip/analysis/machine_learning_models/02_07_2023_RFE_RF_10kb_windows/differential_analysis_filter/\")\n",
    "load(\"10_09_2023_rf_10kb_with_hyperparameter_tuning_accuracyevaluation_withdifferentialfilter.RData\")\n",
    "\n",
    "# in case you're interested in only the hypermethylated features for this; SUBSET TO PREDICTORS WITH A POSITIVE logFC\n",
    "# predictors_rf_hyper <- res[rownames(res) %in% predictors_rf & res$table$logFC > 0,]\n",
    "# predictors_rf_hyper <- rownames(predictors_rf_hyper)\n",
    "\n",
    "# check feature importance. Visualize top 100 features using bar plot.\n",
    "importance <- varImp(model, scale = FALSE) # try with and without scaling\n",
    "# importance <- importance[order(importance$importance$Overall, decreasing = TRUE), ]\n",
    "\n",
    "variable_importance <- data.frame(feature = row.names(importance$importance),\n",
    "                                  importance = importance$importance$Overall)\n",
    "\n",
    "# sort by decreasing variable importance\n",
    "variable_importance <- variable_importance[order(variable_importance$importance, decreasing = TRUE), ]\n",
    "\n",
    "# subset to top 5% of important features (9370*0.05 = 469)\n",
    "variable_importance_subset <- variable_importance[1:469, ]\n",
    "\n",
    "# plot variables\n",
    "p <- variable_importance_subset %>% \n",
    "  ggplot(aes(x = reorder(feature, -importance), y = importance, fill = feature)) +\n",
    "  geom_bar(stat=\"identity\") + \n",
    "  labs(x = \"Features\", y = \"Variable Importance\") + \n",
    "  # geom_text(aes(label = round(importance, 2)), vjust=1.6, color=\"white\", size=4) + \n",
    "  theme_minimal() + \n",
    "  theme(legend.position = \"none\")\n",
    "print(p)\n",
    "\n",
    "\n",
    "\n",
    "# perform visualizations. Subset to top 5% by feature importance, and visualize these on the volcano plot. See if they are the \n",
    "# features with higher effect size.\n",
    "de_features_subset_rf_highimportance <- res[(rownames(res)) %in% variable_importance_subset$feature,]\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67b91f95-721d-4ca6-ac4d-ce44ff32e58c",
   "metadata": {},
   "source": [
    "- Keep in mind, variable importance may not be additive. The sum of output of VarImp does not equal 1."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9254579-70e3-4c43-b52d-a161987baae1",
   "metadata": {},
   "source": [
    "### Enrichment of modeled features in regulatory elements, genome wide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2be12dfc-476d-4991-bf03-1fb5553c22f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# See where modeled features are enriched\n",
    "# load annotation set\n",
    "setwd(\"/cluster/projects/scottgroup/people/steven/reference_sets/other_references\")\n",
    "annotated_ref_no_X_Y_M_blacklist <- read.delim(\"10kb_bin_genome_wide_annotated_reference_with_regulatory_features_no_X_Y_M_ENCODEBlacklist.txt\", header = TRUE, sep = \"\\t\")\n",
    "\n",
    "# subset to modeled features\n",
    "\n",
    "rf_hyper_annotations <- annotated_ref_no_X_Y_M_blacklist[annotated_ref_no_X_Y_M_blacklist$bin_id %in% predictors_rf_hyper,]\n",
    "# test <- table(rf_hyper_annotations)\n",
    "\n",
    "# take the summary of the annotations, and extract the important info for visualization\n",
    "# cbind the summary of the annotations to one another into a single data frame in long format, and mutate a label representing the regulatory set.\n",
    "cpg_annot <- as.data.frame(table(rf_hyper_annotations$cpg_annotations))\n",
    "cpg_annot$annotation <- c(\"cpg_annotations\")\n",
    "cpg_annot$Var1 <- as.character(cpg_annot$Var1)\n",
    "cpg_annot[5,1] <- \"not_cpg\" # change \"other\" annotation to \"not_cpg\"\n",
    "\n",
    "gene_annot <- as.data.frame(table(rf_hyper_annotations$genic_annotations))\n",
    "gene_annot$annotation <- c(\"genic_annotations\")\n",
    "\n",
    "enhancer_annot <- as.data.frame(table(rf_hyper_annotations$enhancer_annotations))\n",
    "enhancer_annot$annotation <- c(\"enhancer_annotations\")\n",
    "enhancer_annot$Var1 <- as.character(enhancer_annot$Var1)\n",
    "enhancer_annot[2,1] <- \"not_enhancer\" # change \"other\" annotation to \"not_enhancer\"\n",
    "\n",
    "# combine into single annotation summary set\n",
    "annotation_summary <- rbind(cpg_annot,gene_annot,enhancer_annot)\n",
    "\n",
    "# set factor level for annotations\n",
    "annotation_summary$annotation <- as.factor(annotation_summary$annotation)\n",
    "# annotation_summary$Var1 <- unique(annotation_summary$Var1)\n",
    "\n",
    "# visualize as bar plot\n",
    "annotation_summary %>%\n",
    "ggplot(aes(x = Var1, y = Freq, colour = annotation)) +\n",
    "geom_point(size = 10) +\n",
    "theme_minimal()\n",
    "\n",
    "\n",
    "\n",
    "# create random subset of 2850 features, 100 times; compile in the same way as above.\n",
    "\n",
    "# Initialize an empty list to store the results\n",
    "results_list <- list()\n",
    "\n",
    "for (iteration in 1:20) {\n",
    "  # Randomly select X number of features; adjust according to the number of desired features\n",
    "  #selected <- sample(rownames(X_train), 5056, replace = FALSE)\n",
    "  selected <- sample(rownames(X_train), 2850, replace = FALSE)\n",
    "  # selected <- sample(rownames(X_train), 5328, replace = FALSE)\n",
    "\n",
    "  # Subset the matrix\n",
    "  subset_matrix <- annotated_ref_no_X_Y_M_blacklist[annotated_ref_no_X_Y_M_blacklist$bin_id %in% selected,]\n",
    "\n",
    "  # take the summary of the annotations, and extract the important info for visualization\n",
    "  # cbind the summary of the annotations to one another into a single data frame in long format, and mutate a label representing the regulatory set.\n",
    "  cpg_annot_random <- as.data.frame(table(subset_matrix$cpg_annotations))\n",
    "  cpg_annot_random$annotation <- c(\"cpg_annotations\")\n",
    "  cpg_annot_random$Var1 <- as.character(cpg_annot_random$Var1)\n",
    "  cpg_annot_random[5,1] <- \"not_cpg\" # change \"other\" annotation to \"not_cpg\"\n",
    "\n",
    "  gene_annot_random <- as.data.frame(table(subset_matrix$genic_annotations))\n",
    "  gene_annot_random$annotation <- c(\"genic_annotations\")\n",
    "\n",
    "  enhancer_annot_random <- as.data.frame(table(subset_matrix$enhancer_annotations))\n",
    "  enhancer_annot_random$annotation <- c(\"enhancer_annotations\")\n",
    "  enhancer_annot_random$Var1 <- as.character(enhancer_annot_random$Var1)\n",
    "  enhancer_annot_random[2,1] <- \"not_enhancer\" # change \"other\" annotation to \"not_enhancer\"\n",
    "\n",
    "  # combine into single annotation summary set\n",
    "  annotation_summary_random <- rbind(cpg_annot_random,\n",
    "                                     gene_annot_random,\n",
    "                                     enhancer_annot_random)\n",
    "\n",
    "  # Append the results to the list\n",
    "  results_list[[iteration]] <- annotation_summary_random[,2]\n",
    "}\n",
    "\n",
    "# Combine the results from all iterations into a single data frame\n",
    "combined_results <- bind_rows(results_list, .id = \"Iteration\")\n",
    "\n",
    "# make rowname a column in combined_results\n",
    "combined_results <- combined_results %>%\n",
    "  mutate(SampleName = str_extract(rownames(combined_results), \".*K27me3\"))\n",
    "\n",
    "# Calculate the mean of each observation in the \"Sum\" column across all iterations\n",
    "mean_values <- combined_results %>%\n",
    "  group_by(SampleName) %>%\n",
    "  summarise(Mean_Median = mean(Median),\n",
    "            Mean_Sum = mean(Sum))\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# visualize\n",
    "hist(rf_hyper_annotations$bin_id, breaks = 22) # genome wide distribution of the modeled features (not scaled for genome size). Quick vis\n",
    "# random iterations\n",
    "\n",
    "# 20-10-2023 additional visualization; look at the genome-wide distribution of features by logFC using geom_segment.\n",
    "# de_features_subset <- res[res$table$PValue <= 0.05,] # 9370 features\n",
    "de_features_subset <- res[rownames(res) %in% predictors_rf,] # all modeled features\n",
    "\n",
    "de_features_subset_rownames <- rownames(de_features_subset)\n",
    "\n",
    "# need to import annotation set without regions removed, so full length segments can be created. Remove X, Y, and M from the annotation set.\n",
    "setwd(\"/cluster/projects/scottgroup/people/steven/reference_sets/other_references\")\n",
    "annotated_ref_no_X_Y_M <- read.delim(\"10kb_bin_genome_wide_annotated_reference_with_regulatory_features_no_X_Y_M.bed\", header = TRUE, sep = \"\\t\")\n",
    "\n",
    "# add chromosomal lengths to the annotation set by taking the sum of the widths over each chromosome.\n",
    "annotated_ref_no_X_Y_M <- annotated_ref_no_X_Y_M %>%\n",
    "  group_by(chrom) %>%\n",
    "  mutate(chrom_lengths = sum(width)) %>%\n",
    "  ungroup()\n",
    "\n",
    "# subset to ranges of these features in the annotation set.\n",
    "all_sig_annotations <- annotated_ref_no_X_Y_M[annotated_ref_no_X_Y_M$bin_id %in% de_features_subset_rownames,]\n",
    "\n",
    "# sort de_features_subset by rowname, or order of bin ID. Then cbind to annotation set subset, for visualization.\n",
    "de_features_subset <- de_features_subset[order(as.numeric(rownames(de_features_subset))), ]\n",
    "\n",
    "combined_feature_expression_annotations <- cbind(de_features_subset, all_sig_annotations)\n",
    "\n",
    "# set factor level for chromosome number.\n",
    "combined_feature_expression_annotations$chrom <- factor(combined_feature_expression_annotations$chrom,\n",
    "                                                        levels = paste0(\"chr\", 1:22))\n",
    "\n",
    "# add term for hypermethylated and hypomethylated features to a new column\n",
    "combined_feature_expression_annotations <- combined_feature_expression_annotations %>%\n",
    "  mutate(status = case_when(\n",
    "      logFC > 0 ~ \"hypermethylated\",\n",
    "      logFC < 0 ~ \"hypomethylated\")\n",
    "        )\n",
    "\n",
    "\n",
    "# TROUBLESHOOT; TRY TO MAKE START A NEW VARIABLE\n",
    "\n",
    "# visualize, while taking into account chromosome size\n",
    "genome_base <- combined_feature_expression_annotations %>%\n",
    "    ggplot +\n",
    "    # geom_bar(aes(x = chrom, y = chrom_lengths), # background; faulty, y-axis way too  big\n",
    "    #         stat='identity', \n",
    "    #         fill='grey80', \n",
    "    #         colour='grey80', \n",
    "    #         width=.2) +\n",
    "    # geom_point(aes(x=chrom, y=start, colour = logFC), # Vis 1 (primary)\n",
    "    #     size=1,\n",
    "    #     alpha = 0.8, \n",
    "    #     position=position_jitter(.4)) +\n",
    "    # geom_segment(data = combined_feature_expression_annotations %>%\n",
    "    #                 group_by(chrom) %>%\n",
    "    #                 summarize(min_length = min(chrom_lengths), max_length = max(chrom_lengths)),\n",
    "    #              aes(x = chrom, xend = chrom, y = min_length, yend = max_length),\n",
    "    #              color = \"black\", linewidth = 10) +    \n",
    "    geom_tile(aes(x=chrom, y=start, colour = logFC), # Vis 1 (primary)\n",
    "        linewidth=1,\n",
    "        alpha = 0.8) +\n",
    "    theme_minimal() +\n",
    "    theme(panel.border = element_blank(),\n",
    "            panel.grid.major = element_blank(),\n",
    "            panel.grid.minor = element_blank()) + \n",
    "    theme(axis.text.x = element_text(angle = 90, hjust = 1),\n",
    "          axis.ticks.x = element_blank()) +\n",
    "    labs(x = \"Chromosome\",\n",
    "         y = \"Chromosome Position\") +\n",
    "    scale_colour_gradient2(low = \"blue\", mid = \"white\", high = \"red\")\n",
    "print(genome_base)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# visualize, without taking into account chromosome size\n",
    "genome_base <- combined_feature_expression_annotations %>%\n",
    "  ggplot() +\n",
    "  geom_point(aes(x = chrom, y = logFC, colour = status),\n",
    "            size=0.2,\n",
    "            alpha = 0.5, \n",
    "            position=position_jitter(.3)) +\n",
    "  theme_minimal() +\n",
    "  theme(panel.border = element_blank(),\n",
    "        panel.grid.major = element_blank(),\n",
    "        panel.grid.minor = element_blank()) +\n",
    "  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +\n",
    "  labs(x = \"Chromosome\", y = \"logFC\") #+\n",
    "  # scale_colour_manual(values = c(\"hypermethylated\" = \"red\", \"hypomethylated\" = \"blue\")) #+\n",
    "  # scale_y_log10() # can't log scale as you have negative logFC\n",
    "print(genome_base)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# extra\n",
    "feature_expression_distribution <- genome_base + \n",
    "    geom_segment(data = combined_feature_expression_annotations, \n",
    "        aes(x=chrom, xend=chrom, y=start, yend=end, colour = logFC),\n",
    "        linewidth = 2)\n",
    "print(feature_expression_distribution)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c404f4be-0cec-43ce-9a87-d4f88bd2f6f5",
   "metadata": {},
   "source": [
    "- Problem with loop; if there is none of a particular annotation in a subset of annotations, then the entry won't be in the final annotation table.\n",
    "- Need a way to include the annotation if it isn't present in the subset, with a zero."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1505628-42f3-4610-833b-5e53df5992f8",
   "metadata": {},
   "source": [
    "### 03-10-2023 picard fragment length, short fragment enrichment in plasma H3K27me3 profiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec0278c5-8c8e-4b6a-becb-33521bcddab7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# use output from MEDIPIPE to calculate the abundance of short fragments\n",
    "#Goal; plot output of picard in a single figure to get an idea of the insert sizes/fragment length and whether there's a shift to the left in cancer.\n",
    "setwd(\"/cluster/projects/scottgroup/people/steven/human_plasma_cfChIP-Seq/liberate_plasma_cfchip/\")\n",
    "\n",
    "#loop over all subdirectories and call all fragment length profiles.\n",
    "k27me3_profiles <- list.files(path = \".\", pattern = \"insert_size_metrics.txt\", recursive = TRUE)\n",
    "\n",
    "#calculate frequency of fragment length for each of the pilot files\n",
    "frag_length_all <- data.frame()\n",
    "\n",
    "all_rows <- data.frame(insert_size = seq(1, 1000)) #to fill all rows to a fixed number\n",
    "\n",
    "#for this loop, if extracting fragment length characteristics for violin plot, exclude row filling; only use for stacked histogram! Otherwise NA = error.\n",
    "for (i in k27me3_profiles) {\n",
    "  temp <- read.table(i, header = TRUE, skip = 10)\n",
    "  temp$sample <- paste0(i)\n",
    "  temp$total_sum <- sum(temp$All_Reads.fr_count)\n",
    "  temp$frequency <- temp$All_Reads.fr_count / temp$total_sum\n",
    "  #temp <- merge(all_rows, temp, all = TRUE) #merge dummy frame and temp to make temp a fixed number of rows\n",
    "  #temp <- temp %>% \n",
    "  #  fill(sample, total_sum, .direction = c(\"updown\")) #fill NAs in repeated character/numerical rows with values above and below\n",
    "  #temp$sample <- str_match(temp$sample, \"liberate_(.*?)_K27me3\")[[1]] #make \"sample\" column characters a substring of the current characters\n",
    "  frag_length_all <- rbind(frag_length_all,temp)\n",
    "} \n",
    "\n",
    "#colour palate for line plot\n",
    "colour_line <- met.brewer(\"Hokusai1\", 9)\n",
    "#plot_line\n",
    "frag_length_all %>% \n",
    "  ggplot(aes(x = insert_size, y = frequency, colour = sample)) +\n",
    "  geom_line() +\n",
    "  theme_minimal(base_size = 16) +\n",
    "  labs(x = \"Insert Size\",\n",
    "       y = \"Frequency\",\n",
    "       colour = \"H3K27me3 Profile\") +\n",
    "  scale_colour_manual(values = colour_line) +\n",
    "  scale_x_continuous(limits = c(100,200))\n",
    "\n",
    "# #plot_alt; stacked histogram\n",
    "# frag_length_all %>% \n",
    "#   ggplot(aes(x = insert_size, y = sample, fill = frequency)) +\n",
    "#   geom_tile() +\n",
    "#   theme_minimal(base_size = 10) +\n",
    "#   labs(x = \"Insert Size\",\n",
    "#        y = \"Sample\",\n",
    "#        fill = \"Frequency\") +\n",
    "#   scale_fill_gradient(low = \"black\", high = \"#E85e19\", na.value = \"black\") +\n",
    "#   scale_x_continuous(limits = c(0,600))\n",
    "\n",
    "#interestingly; there are a few cancer samples with a surplus of <150bp fragments; quantify the fraction of <150bp fragments across the different samples. and bin by cancer status.Do this for the entire cohort.\n",
    "\n",
    "fraction_short_fragments <- data.frame()\n",
    "range <- as.numeric(c(126:135, 240:330)) #range 1\n",
    "# range <- as.numeric(c(170:200, 240:330)) #range 2; determined from median fragment length across T1, T3, and non-lymphoma. Not ideal.\n",
    "# range <- as.numeric(c(170:200)) #range 3; determined from median fragment length across T1, T3, and non-lymphoma. Mononucleosome alone, not ideal.\n",
    "# range <- as.numeric(c(126:135)) #range 4; mononucleosome alone.\n",
    "# range <- as.numeric(c(240:330)) #range 5; dinucleosome alone. Truly the most informative range.\n",
    "# range <- as.numeric(c(250:350)) #subtype\n",
    "\n",
    "for (i in unique(frag_length_all$sample)) {\n",
    "  temp <- sum(frag_length_all[which(frag_length_all$insert_size %in% range\n",
    "                                    & frag_length_all$sample == i),5]) #select all rows where the insert size falls in the specified range\n",
    "  fraction_short_fragments <- rbind(fraction_short_fragments,temp)\n",
    "} #loop over each of the samples and take the sum of the frequency of fragments in the specified insert size range\n",
    "\n",
    "#change depending on range selected\n",
    "fraction_short_fragments <- cbind(fraction_short_fragments, unique(frag_length_all$sample)) #bind names of the samples to the new data frame\n",
    "colnames(fraction_short_fragments) <- c(\"Value\", \"Sample\")\n",
    "\n",
    "# #healthy vs cancer condition\n",
    "# fraction_short_fragments <- fraction_short_fragments %>%\n",
    "#   mutate(cancer_status = case_when(\n",
    "#     str_detect(Sample, \"HUCON|_C\") ~ \"Healthy\",\n",
    "#     str_detect(Sample, \"GCB|ABC|FL|NA\") ~ \"Cancer\"\n",
    "#     )) #conditional statement; if sample name has HUCON or _C, print \"Healthy\" in new column; if else, print \"Cancer\"\n",
    "\n",
    "#T1 vs T3 vs healthy condition\n",
    "fraction_short_fragments <- fraction_short_fragments %>%\n",
    "  mutate(cancer_status = case_when(\n",
    "    str_detect(Sample, \"HUCON|_C\") ~ \"Healthy\",\n",
    "    str_detect(Sample, \"T1\") ~ \"T1 (baseline)\",\n",
    "    str_detect(Sample, \"T3|T4|T5\") ~ \"T3 (post-treatment)\",\n",
    "    #str_detect(Sample, \"T5\") ~ \"T5 (relapse)\"\n",
    "    )) #keep in mind you grouped two of the T4 samples (longitudinal follow up) with the T3 group.\n",
    "\n",
    "#subtypes\n",
    "fraction_short_fragments <- fraction_short_fragments %>%\n",
    "  mutate(subtype = case_when(\n",
    "    str_detect(Sample, \"Non-lymphoma\") ~ \"Healthy\",\n",
    "    str_detect(Sample, \"GCB|ABC|NA\") ~ \"DLBCL\",\n",
    "    str_detect(Sample, \"FL\") ~ \"FL\",\n",
    "    )) %>% \n",
    "  filter(!str_detect(Sample, \"T3|T4|T5\")) #Exclude T3, T4, and T5.\n",
    "\n",
    "# extract the bioinformatic name from the Sample column, mutate as additional column to match to the sample metadata for other visualizations\n",
    "fraction_short_fragments <- fraction_short_fragments %>%\n",
    "mutate(bioinformatic_name = regmatches(Sample, regexpr(\"liberate_(.*?)_K27me3\", Sample)))\n",
    "\n",
    "# additional; draw lines bw T1 and T3 samples. Set up grouping variable\n",
    "\n",
    "# colour palatte\n",
    "colour_fraction_short_fragments <- MetBrewer::met.brewer(name = \"Archambault\", n = 3)\n",
    "\n",
    "#plot frequencies for cancer vs normal\n",
    "fraction_short_fragments %>% \n",
    "  ggplot(aes(x = cancer_status, y = Value, fill = cancer_status)) +\n",
    "  geom_violin() +\n",
    "  geom_point(alpha = 0.7, position = position_jitterdodge(0.05), size = 2.5) +\n",
    "  # geom_line(aes(group = cancer_status)) +\n",
    "  theme_minimal(base_size = 20) +\n",
    "  labs(x = \"Status\",\n",
    "       #y = \"Fraction of <150bp fragments\"\n",
    "       y = \"Fraction of fragments between 126-135bp and 240-330bp\") +\n",
    "  scale_fill_manual(values = colour_fraction_short_fragments)\n",
    "\n",
    "##### metrics #####\n",
    "\n",
    "#mean fraction of <150bp fragments, based off condition.\n",
    "mean_sub150_cancer <- mean(fraction_under_150$Value[fraction_under_150$cancer_status == \"Cancer\"])\n",
    "mean_sub150_healthy <- mean(fraction_under_150$Value[fraction_under_150$cancer_status == \"Healthy\"])\n",
    "\n",
    "#backup code\n",
    "sum(frag_length_all[which(frag_length_all$insert_size <= 150 & frag_length_all$sample == \"liberate_0261_T1_GCB_K27me3_pool6_5_S5_insert_size_metrics.txt\"),5]) #take the sum of the frequency of fragments over a specified range; this is currently taking all fragments less than 150 across all samples; subdivide calculation by character.\n",
    "\n",
    "sum(frag_length_all[1:150,5])\n",
    "\n",
    "#calculate median faction of short fragments for T1 and healthy\n",
    "healthy_fraction_short_fragments <- fraction_short_fragments[fraction_short_fragments$cancer_status == \"Healthy\",]\n",
    "T1_fraction_short_fragments <- fraction_short_fragments[fraction_short_fragments$cancer_status == \"T1 (baseline)\",]\n",
    "T3_fraction_short_fragments <- fraction_short_fragments[fraction_short_fragments$cancer_status == \"T3 (post-treatment)\",]\n",
    "\n",
    "mean(T1_fraction_short_fragments$Value)\n",
    "mean(T3_fraction_short_fragments$Value)\n",
    "mean(healthy_fraction_short_fragments$Value)\n",
    "\n",
    "t.test(x = healthy_fraction_short_fragments$Value, y = T1_fraction_short_fragments$Value, var.equal = FALSE) #p = 0.0022 (significant)\n",
    "t.test(x = T1_fraction_short_fragments$Value, y = T3_fraction_short_fragments$Value, var.equal = FALSE) #p = 0.0014 (significant)\n",
    "t.test(x = healthy_fraction_short_fragments$Value, y = T3_fraction_short_fragments$Value, var.equal = FALSE) #p = 0.555 (not significant)\n",
    "\n",
    "\n",
    "\n",
    "# 03-10-2023 CONTINUED; match fragment lengths to sample metadata. CONTINUE TOMORROW.\n",
    "sample_list_training <- sample_list[sample_list$sample_name %in% colnames(cpm_norm),]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# 26-10-2023; generate average picard fragment length for all cancer and non-lymphoma samples. See if there is a significant difference.\n",
    "#calculate frequency of fragment length for each of the pilot files\n",
    "frag_length_all <- data.frame()\n",
    "\n",
    "all_rows <- data.frame(insert_size = seq(1, 1000)) #to fill all rows to a fixed number\n",
    "\n",
    "#for this loop, if extracting fragment length characteristics for violin plot, exclude row filling; only use for stacked histogram! Otherwise NA = error.\n",
    "for (i in k27me3_profiles) {\n",
    "  temp <- read.table(i, header = TRUE, skip = 10)\n",
    "  temp$sample <- paste0(i)\n",
    "  temp$total_sum <- sum(temp$All_Reads.fr_count)\n",
    "  temp$frequency <- temp$All_Reads.fr_count / temp$total_sum\n",
    "  #temp <- merge(all_rows, temp, all = TRUE) #merge dummy frame and temp to make temp a fixed number of rows\n",
    "  #temp <- temp %>% \n",
    "  #  fill(sample, total_sum, .direction = c(\"updown\")) #fill NAs in repeated character/numerical rows with values above and below\n",
    "  #temp$sample <- str_match(temp$sample, \"liberate_(.*?)_K27me3\")[[1]] #make \"sample\" column characters a substring of the current characters\n",
    "  frag_length_all <- rbind(frag_length_all,temp)\n",
    "} \n",
    "\n",
    "# mutate cancer status (non-lymphoma, T1, T3) to each respective row.\n",
    "#T1 vs T3 vs healthy condition\n",
    "frag_length_all <- frag_length_all %>%\n",
    "  mutate(cancer_status = case_when(\n",
    "    str_detect(sample, \"HUCON|_C\") ~ \"Non-lymphoma\",\n",
    "    str_detect(sample, \"T1\") ~ \"T1 (baseline)\",\n",
    "    str_detect(sample, \"T3|T4|T5\") ~ \"T3 (post-treatment)\",\n",
    "    #str_detect(Sample, \"T5\") ~ \"T5 (relapse)\"\n",
    "    )) #keep in mind you grouped two of the T4 samples (longitudinal follow up) with the T3 group.\n",
    "\n",
    "#subtypes\n",
    "# frag_length_all <- frag_length_all %>%\n",
    "#   mutate(subtype = case_when(\n",
    "#     str_detect(sample, \"Non-lymphoma\") ~ \"Healthy\",\n",
    "#     str_detect(sample, \"GCB|ABC|NA\") ~ \"DLBCL\",\n",
    "#     str_detect(sample, \"FL\") ~ \"FL\",\n",
    "#     )) %>% \n",
    "#   filter(!str_detect(sample, \"T3|T4|T5\")) #Exclude T3, T4, and T5.\n",
    "\n",
    "# SIDE; for non-lymphoma vs lymphoma condition comparison only (ISLB poster)\n",
    "frag_length_all <- frag_length_all[frag_length_all$cancer_status == c(\"Non-lymphoma\", \"T1 (baseline)\"),]\n",
    "\n",
    "# group by status, and calculate mean frequency across groups (non-lymphoma, T1, T3)\n",
    "frag_length_grouped_median <- frag_length_all %>%\n",
    "group_by(cancer_status, insert_size) %>%\n",
    "summarize(median_frequency = median(frequency))\n",
    "\n",
    "# group by lymphoma subtype, see if subtype related differences exist in the fragment lengths\n",
    "# frag_length_grouped_median_subtype <- frag_length_all %>%\n",
    "# group_by(subtype, insert_size) %>%\n",
    "# summarize(median_frequency = median(frequency))\n",
    "\n",
    "\n",
    "\n",
    "# plot\n",
    "frag_length_grouped_median %>%\n",
    "ggplot(aes(x = insert_size, y = median_frequency, colour = cancer_status)) +\n",
    "geom_line() +\n",
    "theme_minimal(base_size = 20) +\n",
    "labs(x = \"Fragment Size\",\n",
    "    y = \"Fragment Frequency\",\n",
    "    colour = \"Cancer Status\") +\n",
    "scale_x_continuous(limits = c(225,400)) +\n",
    "scale_colour_manual(values = colour_fraction_short_fragments)\n",
    "\n",
    "# stats for difference in mononucleosomal range\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1799ba0e-a6af-4e01-a65f-eb644832b1ce",
   "metadata": {},
   "source": [
    "### Heatmap of features identified from modeling (10-09-2023)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ca3dc35-f604-4ffc-975f-1c3bc8d0d899",
   "metadata": {},
   "outputs": [],
   "source": [
    "# use heatmap as before to see the difference in the features between the groups. See if clusters of individual features are different, as we saw bi-directional \n",
    "# changes in feature expression from the differential analysis.\n",
    "\n",
    "# load dataset\n",
    "setwd(\"/cluster/projects/scottgroup/people/steven/human_plasma_cfChIP-Seq/liberate_plasma_cfchip/analysis/machine_learning_models/02_07_2023_RFE_RF_10kb_windows/differential_analysis_filter/\")\n",
    "load(\"10_09_2023_rf_10kb_with_hyperparameter_tuning_accuracyevaluation_withdifferentialfilter.RData\")\n",
    "\n",
    "# subset sample metadata to training samples, then visualize expression over the predictors.\n",
    "X_train_attributes <- sample_list[sample_list$sample_name %in% colnames(cpm_norm),]\n",
    "\n",
    "# subset original training matrix to the features in chosen feature set\n",
    "#X_train_filtered <- cpm_norm[(rownames(cpm_norm)) %in% predictors_rf,]\n",
    "X_train_filtered <- cpm_norm[(rownames(cpm_norm)) %in% predictors_rf_hyper,] # for hypermethylated features\n",
    "\n",
    "#normalize counts over significant features from PCA.\n",
    "heatmap_data <- X_train_filtered\n",
    "\n",
    "heatmap_metadata <- data.frame(X_train_attributes$timepoint, \n",
    "                               X_train_attributes$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY.DIAGNOSIS_SPECIFIC_DLBCL,\n",
    "                               #X_train_attributes$IPI,\n",
    "                               #X_train_attributes$FLIPI,\n",
    "                               X_train_attributes$ANN_ARBOR_STAGE#,\n",
    "                               #X_train_attributes$TUMOR_MASS_MAX_CM\n",
    "                               )\n",
    "\n",
    "#change metadata labels (aka column names); add metadata as required\n",
    "colnames(heatmap_metadata) <- c(\"timepoint\",\n",
    "                                \"subtype\",\n",
    "                                #\"IPI\",\n",
    "                                \"ANN_ARBOR_STAGE\"#,\n",
    "                                #\"z\",\n",
    "                                #\"ab\",\n",
    "                                )\n",
    "\n",
    "rownames(heatmap_metadata) <- colnames(heatmap_data)\n",
    "\n",
    "###Convert integer columns to character, for metadata labeling\n",
    "#heatmap_metadata$sample_list.IPI <- as.character(heatmap_metadata$sample_list.IPI)\n",
    "#heatmap_metadata$sample_list.IPI <- as.character(heatmap_metadata$sample_list.FLIPI)\n",
    "heatmap_metadata$ANN_ARBOR_STAGE <- as.character(heatmap_metadata$ANN_ARBOR_STAGE)\n",
    "\n",
    "###Make all NAs as characters, for metadata labeling\n",
    "heatmap_metadata$subtype[is.na(heatmap_metadata$subtype)] <- \"NA\"\n",
    "#heatmap_metadata$sample_list.IPI[is.na(heatmap_metadata$sample_list.IPI)] <- \"NA\"\n",
    "#heatmap_metadata$sample_list.FLIPI[is.na(heatmap_metadata$sample_list.FLIPI)] <- \"NA\"\n",
    "heatmap_metadata$ANN_ARBOR_STAGE[is.na(heatmap_metadata$ANN_ARBOR_STAGE)] <- \"NA\"\n",
    "#heatmap_metadata$sample_list.TUMOR_MASS_MAX_CM[is.na(heatmap_metadata$sample_list.TUMOR_MASS_MAX_CM)] <- 0\n",
    "\n",
    "#Independent palattes for each piece of metadata\n",
    "palatte_timepoint <- met.brewer(name = \"Homer1\", n = 2, type = \"discrete\")\n",
    "palatte_subtype <- met.brewer(name = \"Homer2\", n = 5, type = \"discrete\")\n",
    "#palatte_ipi <- RColorBrewer::brewer.pal(n = 7, name = \"Spectral\")\n",
    "#palatte_flipi <- met.brewer(name = \"Homer1\", n = 6, type = \"discrete\")\n",
    "palatte_stage <- met.brewer(name = \"Homer2\", n = 5, type = \"discrete\")\n",
    "#palatte_diameter <- met.brewer(name = \"Hokusai3\", n = 92, type = \"continuous\")\n",
    "\n",
    "###assign a colour palatte to each of the metadata labels, in list format for annotation_colours command in pheatmap.\n",
    "heatmap_metadata_colour <- list(timepoint = setNames(palatte_timepoint, unique(heatmap_metadata$timepoint)),\n",
    "                                subtype = setNames(palatte_subtype, unique(heatmap_metadata$subtype)))\n",
    "\n",
    "h <- pheatmap(heatmap_data, \n",
    "         scale=\"row\",\n",
    "         #cluster_rows = FALSE, \n",
    "         clustering_method = \"ward.D2\",\n",
    "         annotation_col = heatmap_metadata,\n",
    "         annotation_colors = heatmap_metadata_colour,\n",
    "         color=colorRampPalette(c(\"red\", \"white\", \"blue\"))(100),\n",
    "         show_rownames = FALSE)\n",
    "print(h)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1884950-2d3f-4ad7-bd07-1c5ba7f96da0",
   "metadata": {},
   "source": [
    "### Volcano plot of modeled features, on top of differential features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "413cdedd-ef87-4ff3-acf3-5f2790a4c990",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot 3 colours; black = all features, red = differential features, green = modeled features from RF+SMOTE\n",
    "predictors_rf <- predictors(model)\n",
    "de_features_subset <- res[res$table$PValue < 0.05,]\n",
    "de_features_subset_rf <- res[(rownames(res)) %in% predictors_rf,]\n",
    "# de_features_subset_rf <- de_features_subset_rf[de_features_subset_rf$table$logFC > 0,] # if you only want the hypermethylated fraction\n",
    "de_features_rf <- rownames(de_features_subset_rf)\n",
    "de_features_all <- rownames(res)\n",
    "# de_features_subset_rf_highimportance <- res[(rownames(res)) %in% variable_importance_subset$feature,]\n",
    "# de_features_rf_highimportance <- rownames(de_features_subset_rf_highimportance)\n",
    "\n",
    "\n",
    "# plot volcano, visualize the differentially epressed features above significance threshold in red.\n",
    "# table for all features\n",
    "logFC <- res$table$logFC\n",
    "p_value <- res$table$PValue\n",
    "neg_log_p_value <- -log10(p_value)\n",
    "df_feature_stats <- data.frame(logFC,neg_log_p_value,de_features_all) #compile data for volcano plot.\n",
    "\n",
    "# table for all significant features\n",
    "logFC_signif <- de_features_subset$table$logFC\n",
    "p_value_signif <- de_features_subset$table$PValue\n",
    "neg_log_p_value_signif <- -log10(p_value_signif)\n",
    "df_feature_stats_signif <- data.frame(logFC_signif,neg_log_p_value_signif,de_features_subset) #compile data for volcano plot.\n",
    "\n",
    "# table for all modeled features (subset of differential features)\n",
    "logFC_signif_rf <- de_features_subset_rf$table$logFC\n",
    "p_value_signif_rf <- de_features_subset_rf$table$PValue\n",
    "neg_log_p_value_signif_rf <- -log10(p_value_signif_rf)\n",
    "df_feature_stats_signif_rf <- data.frame(logFC_signif_rf,neg_log_p_value_signif_rf,de_features_subset_rf) #compile data for volcano plot.\n",
    "\n",
    "# table for top 5% of modeled features, by feature importance using VarImp function (subset of modeled features)\n",
    "# logFC_signif_rf_highimportance <- de_features_subset_rf_highimportance$table$logFC\n",
    "# p_value_signif_rf_highimportance <- de_features_subset_rf_highimportance$table$PValue\n",
    "# neg_log_p_value_signif_rf_highimportance <- -log10(p_value_signif_rf_highimportance)\n",
    "# df_feature_stats_signif_rf_highimportance <- data.frame(logFC_signif_rf_highimportance,\n",
    "                                                        # neg_log_p_value_signif_rf_highimportance,\n",
    "                                                        # de_features_subset_rf_highimportance) #compile data for volcano plot.\n",
    "\n",
    "v <- ggplot() +\n",
    "  geom_point(data = df_feature_stats, aes(x = logFC, y = neg_log_p_value), colour = \"black\", alpha = 0.3) +\n",
    "  geom_point(data = df_feature_stats_signif, aes(x = logFC_signif, y = neg_log_p_value_signif), colour = \"red\", alpha = 0.3) +\n",
    "  geom_point(data = df_feature_stats_signif_rf, aes(x = logFC_signif_rf, y = neg_log_p_value_signif_rf), colour = \"blue\", alpha = 0.3) +\n",
    "  geom_label(data = df_feature_stats_signif, aes(x = logFC_signif, y = neg_log_p_value_signif, label = rownames(df_feature_stats_signif))) +\n",
    "  # geom_point(data = df_feature_stats_signif_rf_highimportance, \n",
    "             # aes(x = logFC_signif_rf_highimportance, y = neg_log_p_value_signif_rf_highimportance), colour = \"green\", alpha = 0.3) +\n",
    "  geom_hline(yintercept = -log10(0.05), linetype = \"dashed\", color = \"red\") +\n",
    "  #geom_vline(xintercept = c(-1, 1), linetype = \"dashed\", color = \"blue\") +\n",
    "  labs(x = \"Log Fold Change\", y = \"-log10(p-value)\") +\n",
    "  ggtitle(\"Volcano Plot\") +\n",
    "  theme_minimal() #+ \n",
    "  # geom_text(data = df_feature_stats_signif, aes(x = logFC_signif, y = neg_log_p_value_signif, label = rownames(df_feature_stats_signif)), vjust = -0.5, hjust = 0.5, color = \"red\", size = 3)\n",
    "\n",
    "print(v)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "827490e2-2936-47a4-86d4-4e41b305f8ec",
   "metadata": {},
   "source": [
    "### 28-09-2023 Linear regression, mean VAF vs median of hypermethylated features, compared to random features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "278bbd85-047f-4e7e-9860-c989758da91e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# linear model, in ggplot. Requires X_train_attributes to be formatted from above chunks.\n",
    "X_train_attributes %>% \n",
    "  ggplot(aes(x = mean_VAF, y = median_X_train_filtered)) +\n",
    "  # ggplot(aes(x = mean_VAF, y = mean_X_train_filtered)) +\n",
    "  # ggplot(aes(x = mean_VAF, y = sum_X_train_filtered)) +\n",
    "  geom_point(size = 5) + \n",
    "  stat_smooth(method = \"lm\", formula = y ~ x, color = \"red\", se = TRUE) + \n",
    "  theme_minimal(base_size = 20) +\n",
    "  labs(x = \"Mean VAF\",\n",
    "       y = \"Median of Hypermethylated Feature Expression\"\n",
    "      )\n",
    "\n",
    "# lm stats\n",
    "summary(lm(median_X_train_filtered ~ mean_VAF, X_train_attributes))\n",
    "# summary(lm(mean_X_train_filtered ~ mean_VAF, X_train_attributes))\n",
    "# summary(lm(sum_X_train_filtered ~ mean_VAF, X_train_attributes))\n",
    "\n",
    "\n",
    "# linear model, but for random features. Requires X_train_attributes_subset to be formatted from above chunks.\n",
    "p <- X_train_attributes_subset %>% \n",
    "  ggplot(aes(x = mean_VAF, y = Mean_Median)) +\n",
    "  geom_point(size = 5) + \n",
    "  stat_smooth(method = \"lm\", formula = y ~ x, color = \"red\", se = TRUE) + \n",
    "  theme_minimal(base_size = 20) +\n",
    "  labs(x = \"Mean VAF\",\n",
    "       y = \"Median Expression of 5328 Random Features, over 1000 Iterations\"\n",
    "      )\n",
    "print(p)\n",
    "\n",
    "# lm stats\n",
    "summary(lm(Mean_Median ~ mean_VAF, X_train_attributes_subset))$r.squared\n",
    "lm_k27_mean_VAF_randomsubset <- summary(lm(Mean_Median ~ mean_VAF, X_train_attributes_subset))\n",
    "\n",
    "\n",
    "\n",
    "# 23-10-2023\n",
    "# split DLBCL and FL cases\n",
    "X_train_attributes_subtype <- X_train_attributes[!(X_train_attributes$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY == \"FL\"), ] # for DLBCL\n",
    "X_train_attributes_subtype <- X_train_attributes[!(X_train_attributes$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY == \"DLBCL\"), ] # for FL\n",
    "\n",
    "p <- X_train_attributes_subtype %>% \n",
    "  ggplot(aes(x = mean_VAF, y = median_X_train_filtered)) +\n",
    "  geom_point(size = 5) + \n",
    "  stat_smooth(method = \"lm\", formula = y ~ x, color = \"red\", se = TRUE) + \n",
    "  theme_minimal(base_size = 20) +\n",
    "  labs(x = \"Mean VAF\",\n",
    "       y = \"Median of Hypermethylated Feature Expression\"\n",
    "      )\n",
    "print(p)\n",
    "\n",
    "# lm stats\n",
    "summary(lm(median_X_train_filtered ~ mean_VAF, X_train_attributes_subtype))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a153b8c-9c5e-4e0f-8486-9b941a2d3b43",
   "metadata": {},
   "source": [
    "## test RF model on test set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8db094f4-ca8d-43b0-b148-481b422b8e8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# test set preparation\n",
    "# apply cpm normalization to test data\n",
    "X_test <- cpm(X_test, normalized.lib.sizes = TRUE, log = FALSE)\n",
    "\n",
    "# subset X_test to features from training set differential analysis\n",
    "X_test <- X_test[de_features,]\n",
    "\n",
    "# project model onto new test data\n",
    "X_test <- t(X_test)\n",
    "predictions_differential_rf <- predict(model, newdata = X_test)\n",
    "\n",
    "#create confusion matrix for the predictions\n",
    "confusion_matrix_differential_rf <- table(Reference = y_test, Prediction = predictions_differential_rf)\n",
    "\n",
    "# poor performance for healthy, great performance for lymphoma. This is just a limitation of the study, continue working with the features, regardless.\n",
    "# More balanced dataset for the subtyping question, so maybe test performance will be better there?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe546cbd-f985-46f5-80f3-a5a18c689ab5",
   "metadata": {},
   "source": [
    "# 11-09-2023 Survival analysis over modeled features from differential + RF"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0751757f-80b5-47ff-b460-0546a6e0d377",
   "metadata": {},
   "source": [
    "## Pre-process data, incorporate MTV and tumor burden from ichorCNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fed122c3-9c3e-4360-8af9-3433942657ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import MTV data; data generated by the nuclear medicine team at UHN\n",
    "setwd(\"/cluster/projects/scottgroup/people/steven/human_plasma_cfChIP-Seq/liberate_plasma_cfchip/other_profiles/tmtv_data_lymphoma_cases\")\n",
    "tmtv_lymphoma <- readxl::read_xlsx(\"17-11-2023_PET_scan_data_lymphoma_cases.xlsx\")\n",
    "\n",
    "# import ichorCNA output and match to X_train_attributes in the same way you incorporated TMTV\n",
    "setwd(\"/cluster/projects/scottgroup/people/steven/human_plasma_cfChIP-Seq/liberate_plasma_cfchip/other_profiles/ichorCNA_data_shallow_WGS_lymphoma\")\n",
    "ichorCNA_lymphoma <- read.delim(\"Bratman.ichorCNA.txt\", header = TRUE, sep = \"\\t\")\n",
    "\n",
    "# merge MTV data frame with X_train_attributes\n",
    "X_train_attributes_all_data_types <- merge(X_train_attributes, tmtv_lymphoma, by.x = \"PATIENT_ID\", by.y = \"LIBERATE ID\", all.x = TRUE)\n",
    "\n",
    "# merge ichorCNA data frame with X_train_attributes; this step is where you're gaining duplicate rows!!\n",
    "X_train_attributes_all_data_types <- merge(X_train_attributes_all_data_types, ichorCNA_lymphoma, by.x = \"PATIENT_ID\", by.y = \"PATIENT_ID\", all.x = TRUE)\n",
    "\n",
    "# remove duplicate ichorCNA entries due to accidental duplicate sequencing of some shallow WGS profiles. Selected single representative entry for duplicates.\n",
    "X_train_attributes_all_data_types <- X_train_attributes_all_data_types %>%\n",
    "  distinct(PATIENT_ID, .keep_all = TRUE)\n",
    "\n",
    "# remove healthy cases from X_train_attributes\n",
    "X_train_attributes_all_data_types <- X_train_attributes_all_data_types[!(X_train_attributes_all_data_types$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY == \"HEALTHY\"),]\n",
    "\n",
    "# convert PFS and OS to numeric\n",
    "X_train_attributes_all_data_types$PFS <- as.numeric(X_train_attributes_all_data_types$PFS)\n",
    "X_train_attributes_all_data_types$OS <- as.numeric(X_train_attributes_all_data_types$OS)\n",
    "X_train_attributes_all_data_types$CODE_PFS <- as.numeric(X_train_attributes_all_data_types$CODE_PFS)\n",
    "X_train_attributes_all_data_types$CODE_OS <- as.numeric(X_train_attributes_all_data_types$CODE_OS)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3898f7c5-d11c-48f5-aa98-06cee5cebc53",
   "metadata": {},
   "source": [
    "## all training cases, survival by median K27me3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de473c1e-9467-4eeb-8c63-a26745fc0579",
   "metadata": {},
   "outputs": [],
   "source": [
    "# survival analysis\n",
    "\n",
    "# calculate median cutoff for median lymphoma thresholds\n",
    "# median_threshold <- median(X_train_attributes_all_data_types$median_X_train_filtered[!(X_train_attributes_all_data_types$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY == \"HEALTHY\")])\n",
    "mean_threshold <- mean(X_train_attributes_all_data_types$median_X_train_filtered[!(X_train_attributes_all_data_types$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY == \"HEALTHY\")])\n",
    "# mean_threshold <- median(X_train_attributes_update$cpm_norm_genes_median[!(X_train_attributes_update$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY == \"HEALTHY\")]) # for individual genes\n",
    "# mean_threshold <- mean(X_train_attributes_update$cpm_norm_genes_median[!(X_train_attributes_update$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY == \"HEALTHY\")]) # for individual genes\n",
    "\n",
    "library(survival)\n",
    "library(survminer)\n",
    "\n",
    "# determine optimal cutoff using survminer surv_cutpoint\n",
    "# cutpoint <- surv_cutpoint(data = X_train_attributes_lymphoma, \n",
    "#                           time = \"PFS\", \n",
    "#                           event = \"CODE_PFS\",\n",
    "#                           variables = \"median_X_train_filtered\")\n",
    "\n",
    "# cutpoint <- cutpoint$cutpoint$cutpoint # this is the stupidest functional code I've ever written\n",
    "\n",
    "# create survival object\n",
    "# survival_object <- Surv(time = X_train_attributes_lymphoma$OS, event = X_train_attributes_lymphoma$CODE_OS) # for overall survival\n",
    "survival_object <- Surv(time = X_train_attributes_all_data_types$PFS, event = X_train_attributes_all_data_types$CODE_PFS) # for progression free survival\n",
    "\n",
    "# might have to remove NAs (aka healthy samples)\n",
    "# high = above threshold, low = below threshold\n",
    "# X_train_attributes_lymphoma$threshold <- ifelse(X_train_attributes_lymphoma$median_X_train_filtered > median_threshold, \"High\", \"Low\")\n",
    "X_train_attributes_all_data_types$threshold <- ifelse(X_train_attributes_all_data_types$median_X_train_filtered > mean_threshold, \"High\", \"Low\")\n",
    "# X_train_attributes_lymphoma$threshold <- ifelse(X_train_attributes_lymphoma$median_X_train_filtered > cutpoint, \"High\", \"Low\")\n",
    "\n",
    "# \n",
    "km_fit <- survfit(survival_object ~ threshold, data = X_train_attributes_all_data_types)\n",
    "\n",
    "# plot Kaplan Meier. Healthy samples were checked for exclusion from this plot.\n",
    "survplot <- ggsurvplot(km_fit, \n",
    "                data = X_train_attributes_all_data_types, \n",
    "                pval = TRUE,\n",
    "                conf.int = TRUE,\n",
    "                ggtheme = theme_minimal(),\n",
    "                risk.table = TRUE,\n",
    "                xlab = \"Time (Years)\",\n",
    "                ylab = \"Overall Survival\",\n",
    "                # ylab = \"Progression Free Survival (PFS)\"\n",
    "                )\n",
    "survplot\n",
    "\n",
    "\n",
    "\n",
    "# check that assumptions of cox proportional hazards model are met for each of the continuous measurements from each patient.\n",
    "# Fit a Cox Proportional Hazards model\n",
    "cox_model <- coxph(survival_object ~ median_X_train_filtered + `STLG (SUV*mL/Kg)` + mean_VAF + Tumor.Fraction, data = X_train_attributes_all_data_types)\n",
    "\n",
    "# Check proportional hazards assumption using Schoenfeld residuals\n",
    "cox_residuals <- cox.zph(cox_model)\n",
    "print(cox_residuals) # individually, the assumptions are met, but combined, the assumption of residuals not correlated with time is violated.\n",
    "\n",
    "# Plot Schoenfeld residuals\n",
    "plot(cox_residuals) # distribution of residuals over time not horizontal; assumption of the model violated\n",
    "\n",
    "# Check for linearity assumption using Martingale residuals\n",
    "# cox_martingale_result <- residuals(cox_model, type = \"martingale\")\n",
    "# plot(cox_martingale_result ~ X_train_attributes_all_data_types$median_X_train_filtered, xlab = \"Variable 1\", ylab = \"Martingale Residuals\")\n",
    "\n",
    "# Check for outliers or influential points\n",
    "# infl <- influence(cox_model)\n",
    "# summary(infl)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f289772d-9652-4a7b-9bdc-edf85a0c77d3",
   "metadata": {},
   "source": [
    "- Mean gives us better results for the survival analysis, with log2(CPM+1) normalized counts. But the groups are unbalanced in that case; do they need to be equal? Look into this.\n",
    "- Distribution of the median values is left skewed. Maybe median is most correct? I think mean is also acceptable. It matches the result of what the optimal cut point is.\n",
    "- Regardless; OS looks awful so we need another approach. Maybe the multivariate cox proportional hazard is the way to do it.\n",
    "- Assumptions violated for the cox model (for univariate and multivariate); leave this on the backburner for now."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c4e39cf-46e2-4d77-971b-69217574dca6",
   "metadata": {},
   "source": [
    "## repeat for independent lymphoma subtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "866dff2e-4d31-4a32-a79e-26f0cb9da286",
   "metadata": {},
   "outputs": [],
   "source": [
    "# repeat this for DLBCL and FL, separately\n",
    "\n",
    "# in case code fails\n",
    "X_train_attributes_lymphoma <- X_train_attributes[!(X_train_attributes$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY == \"HEALTHY\"),]\n",
    "\n",
    "# convert PFS and OS to numeric\n",
    "X_train_attributes_lymphoma$PFS <- as.numeric(X_train_attributes_lymphoma$PFS)\n",
    "X_train_attributes_lymphoma$OS <- as.numeric(X_train_attributes_lymphoma$OS)\n",
    "X_train_attributes_lymphoma$CODE_PFS <- as.numeric(X_train_attributes_lymphoma$CODE_PFS)\n",
    "X_train_attributes_lymphoma$CODE_OS <- as.numeric(X_train_attributes_lymphoma$CODE_OS)\n",
    "\n",
    "median_threshold <- median(X_train_attributes_lymphoma$median_X_train_filtered[!(X_train_attributes_lymphoma$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY %in% c(\"HEALTHY\",\"FL\"))])\n",
    "# median_threshold <- median(X_train_attributes_lymphoma$median_X_train_filtered[!(X_train_attributes_lymphoma$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY %in% c(\"HEALTHY\",\"DLBCL\"))])\n",
    "\n",
    "# subset X_train_attributes to DLBCL and FL samples only\n",
    "X_train_attributes_lymphoma <- X_train_attributes_lymphoma[X_train_attributes_lymphoma$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY == \"DLBCL\",]\n",
    "# X_train_attributes_lymphoma <- X_train_attributes_lymphoma[X_train_attributes_lymphoma$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY == \"FL\",]\n",
    "\n",
    "# create survival object\n",
    "# survival_object <- Surv(time = X_train_attributes_lymphoma$OS, event = X_train_attributes_lymphoma$CODE_OS) # for overall survival\n",
    "survival_object <- Surv(time = X_train_attributes_lymphoma$PFS, event = X_train_attributes_lymphoma$CODE_PFS) # for progression free survival\n",
    "\n",
    "# might have to remove NAs (aka healthy samples)\n",
    "# high = above threshold, low = below threshold\n",
    "X_train_attributes_lymphoma$threshold <- ifelse(X_train_attributes_lymphoma$median_X_train_filtered > median_threshold, \"High\", \"Low\")\n",
    "\n",
    "# \n",
    "km_fit <- survfit(survival_object ~ threshold, data = X_train_attributes_lymphoma)\n",
    "\n",
    "# plot Kaplan Meier. Healthy samples were checked for exclusion from this plot.\n",
    "ggsurvplot(km_fit, \n",
    "                data = X_train_attributes_lymphoma, \n",
    "                pval = TRUE,\n",
    "                conf.int = TRUE,\n",
    "                ggtheme = theme_minimal(),\n",
    "                risk.table = TRUE,\n",
    "                xlab = \"Time (Years)\",\n",
    "                #ylab = \"Overall Survival\",\n",
    "                ylab = \"Progression Free Survival (PFS)\"\n",
    "                )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "904eaa4f-6470-4a9f-a419-942da810b740",
   "metadata": {},
   "source": [
    "- Consider adding the test set to this analysis. Needs more samples.\n",
    "- Think about how to better stratify the cases using CAPP-Seq and MTV. Benchmark against these.\n",
    "- How do I combine the independent clinical variables into a single survival model?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c79a2dc6-782d-44d6-82b6-dd0bd9cd6158",
   "metadata": {},
   "source": [
    "## Benchmark; survival prediction from median of mean VAF and MTV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51143d86-269d-4118-8af5-3f96d7e5104c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# MTV\n",
    "\n",
    "# median_threshold <- median(X_train_attributes_lymphoma$`STLG (SUV*mL/Kg)`[!(X_train_attributes_lymphoma$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY %in% c(\"HEALTHY\",\"FL\"))])\n",
    "median_threshold <- median(X_train_attributes_lymphoma$`STLG (SUV*mL/Kg)`[!(X_train_attributes_lymphoma$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY %in% c(\"HEALTHY\",\"DLBCL\"))])\n",
    "\n",
    "# subset X_train_attributes to DLBCL and FL samples only\n",
    "X_train_attributes_lymphoma <- X_train_attributes_lymphoma[X_train_attributes_lymphoma$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY == \"DLBCL\",]\n",
    "# X_train_attributes_lymphoma <- X_train_attributes_lymphoma[X_train_attributes_lymphoma$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY == \"FL\",]\n",
    "\n",
    "# create survival object\n",
    "# survival_object <- Surv(time = X_train_attributes_lymphoma$OS, event = X_train_attributes_lymphoma$CODE_OS) # for overall survival\n",
    "survival_object <- Surv(time = X_train_attributes_lymphoma$PFS, event = X_train_attributes_lymphoma$CODE_PFS) # for progression free survival\n",
    "\n",
    "# might have to remove NAs (aka healthy samples)\n",
    "# high = above threshold, low = below threshold\n",
    "X_train_attributes_lymphoma$threshold <- ifelse(X_train_attributes_lymphoma$`STLG (SUV*mL/Kg)` > median_threshold, \"High\", \"Low\")\n",
    "\n",
    "# \n",
    "km_fit <- survfit(survival_object ~ threshold, data = X_train_attributes_lymphoma)\n",
    "\n",
    "# plot Kaplan Meier. Healthy samples were checked for exclusion from this plot.\n",
    "ggsurvplot(km_fit, \n",
    "                data = X_train_attributes_lymphoma, \n",
    "                pval = TRUE,\n",
    "                conf.int = TRUE,\n",
    "                ggtheme = theme_minimal(),\n",
    "                risk.table = TRUE,\n",
    "                xlab = \"Time (Years)\",\n",
    "                #ylab = \"Overall Survival\",\n",
    "                ylab = \"Progression Free Survival (PFS)\"\n",
    "                )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a29197e-3181-4ec0-b13e-822fa3f1ded5",
   "metadata": {},
   "source": [
    "- Issue; there are NA MTV values in the FL cases; exclude these from the grouped analyses."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3aca981-4169-4b45-bf82-653f0404f768",
   "metadata": {},
   "source": [
    "## Multivariate Cox Proportional Hazards Analysis; K27me3 + MTV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "555c1541-8548-4fb8-b1ae-f05c0a8f90a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# RUN FIRST CHUNK IN THIS SECTION BEFORE STARTING\n",
    "\n",
    "# remove FL cases from X_train_attributes_all_data_types\n",
    "# X_train_attributes_all_data_types <- X_train_attributes_all_data_types[!(X_train_attributes_all_data_types$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY == \"FL\"),]\n",
    "\n",
    "# remove DLBCL cases from X_train_attributes_all_data_types\n",
    "# X_train_attributes_all_data_types <- X_train_attributes_all_data_types[!(X_train_attributes_all_data_types$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY == \"DLBCL\"),]\n",
    "\n",
    "\n",
    "# for PFS\n",
    "# cox_model <- coxph(Surv(time = X_train_attributes_all_data_types$PFS, \n",
    "                        # event = X_train_attributes_all_data_types$CODE_PFS) ~ X_train_attributes_all_data_types$median_X_train_filtered + X_train_attributes_all_data_types$`STLG (SUV*mL/Kg)` + X_train_attributes_all_data_types$mean_VAF, \n",
    "                   # data = X_train_attributes_all_data_types)\n",
    "\n",
    "# for OS\n",
    "# cox_model <- coxph(Surv(time = X_train_attributes_all_data_types$PFS, \n",
    "#                         event = X_train_attributes_all_data_types$CODE_PFS) ~ X_train_attributes_all_data_types$median_X_train_filtered + X_train_attributes_all_data_types$`STLG (SUV*mL/Kg)` + X_train_attributes_all_data_types$mean_VAF, \n",
    "#                         data = X_train_attributes_all_data_types)\n",
    "\n",
    "# # display this output as a forest plot; incorporate all cases in the cohort near publication, use only training samples for now.\n",
    "# summary(cox_model)\n",
    "# conf_int <- confint(cox_model)\n",
    "\n",
    "\n",
    "# using the finalfit function; good way to look at all the univariate HRs.\n",
    "library(finalfit)\n",
    "\n",
    "# split K27, CAPP-Seq, MTV, and WGS measurements by median, and compare the two sets of groups; high and low.\n",
    "\n",
    "\n",
    "dependent_os <- \"Surv(time = OS, event = CODE_OS)\"\n",
    "\n",
    "dependent_pfs <- \"Surv(time = PFS, event = CODE_PFS)\"\n",
    "\n",
    "explanatory_variables <- c(\"median_X_train_filtered\", \n",
    "                           # \"`STLG (SUV*mL/Kg)`\",\n",
    "                           \"`MTV (mL)`\",\n",
    "                           \"mean_VAF\", \n",
    "                           \"Tumor.Fraction\"#, \n",
    "                           # \"ANN_ARBOR_STAGE\",\n",
    "                           # \"IPI_grouped\"#,\n",
    "                           # \"ECOG\",\n",
    "                           # \"NB_EN_SITES\",\n",
    "                           # \"SEX\"\n",
    "                          )\n",
    "\n",
    "# omit NA values from certain variables, as they may be causing trouble\n",
    "# X_train_attributes_all_data_types$mean_VAF <- na.omit(X_train_attributes_all_data_types$mean_VAF)\n",
    "\n",
    "# coxph.control() # might be worth adjusting if you encounter an interation issue, or no model convergence.\n",
    "\n",
    "# generate model\n",
    "X_train_attributes_all_data_types %>%\n",
    "  # finalfit(dependent_os, explanatory_variables, metrics=TRUE)\n",
    "  finalfit(dependent_pfs, explanatory_variables, metrics=TRUE) #%>%\n",
    "  # setNames(1:5, c(\"Variable\", \"Level\", \"\", \"HR (univariable)\", \"HR (multivariable)\"))\n",
    "\n",
    "# for Odds Ratios; doesn't work for now.\n",
    "X_train_attributes_all_data_types %>%\n",
    "  or_plot(dependent_os, explanatory_variables)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63367ca5-4f86-4c21-8629-fa51a6ad022a",
   "metadata": {},
   "source": [
    "- Continue with this once you've installed the forestplot package into your conda env.\n",
    "- Also, install finalfit package in the meantime.\n",
    "- For the git repository, make the coxph model generation a function to avoid repeat code.\n",
    "- ISSUES; almost nothing is significant, except for ECOG 3; sex (M) is trending towards significance, strangely.\n",
    "- ECOG 3 is still the only significant independent variable when we subset to DLBCL and FL cases only.\n",
    "- Understand this analysis a bit more before showing to Scott or Robert.\n",
    "- Also, HRs for median K27 expression is Inf, and confidence interval is from 0-Inf; that's strange.\n",
    "- Recommendation from Bing AI bot; try a less complex model to start. OUTCOME IS GETTING BETTER.\n",
    "- Figure out what to do with the missing values!\n",
    "- 16-01-2024; Scott made a good point, OS may not be the best metric, as ECOG and IPI should be expected to correlate with these, as age should too.\n",
    "- PFS may be better, and correlates better with our blood and MTV measurements."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64009914-ca52-4668-879f-47437bcc60c7",
   "metadata": {},
   "source": [
    "## 06_12_2023 integration of K27 and MTV for prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b221e08b-2849-4676-b9aa-ec062c5012fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# take the median of both K27 and MTV, to create 4 groups; low k27/low MTV, low K27/high MTV, high K27/low MTV, high K27/high MTV.\n",
    "\n",
    "# remove cases NA for MTV\n",
    "X_train_attributes_lymphoma <- X_train_attributes_lymphoma[complete.cases(X_train_attributes_lymphoma$`STLG (SUV*mL/Kg)`),]\n",
    "\n",
    "# IF DESIRED; subset X_train_attributes to DLBCL and FL samples only\n",
    "# X_train_attributes_lymphoma <- X_train_attributes_lymphoma[X_train_attributes_lymphoma$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY == \"DLBCL\",]\n",
    "# X_train_attributes_lymphoma <- X_train_attributes_lymphoma[X_train_attributes_lymphoma$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY == \"FL\",]\n",
    "\n",
    "# calculate median cutoff for median lymphoma thresholds\n",
    "median_threshold_k27 <- median(X_train_attributes_lymphoma$median_X_train_filtered)\n",
    "median_threshold_mtv <- median(X_train_attributes_lymphoma$`STLG (SUV*mL/Kg)`)\n",
    "\n",
    "library(survival)\n",
    "library(survminer)\n",
    "\n",
    "# create survival object\n",
    "survival_object <- Surv(time = X_train_attributes_lymphoma$OS, event = X_train_attributes_lymphoma$CODE_OS) # for overall survival\n",
    "# survival_object <- Surv(time = X_train_attributes_lymphoma$PFS, event = X_train_attributes_lymphoma$CODE_PFS) # for progression free survival\n",
    "\n",
    "# create the 4 groups listed at the top\n",
    "X_train_attributes_lymphoma <- X_train_attributes_lymphoma %>%\n",
    "mutate(threshold = case_when(\n",
    "    X_train_attributes_lymphoma$median_X_train_filtered <= median_threshold_k27 & X_train_attributes_lymphoma$`STLG (SUV*mL/Kg)` <= median_threshold_mtv ~ \"H3K27me3_low_STLG_low\",\n",
    "    X_train_attributes_lymphoma$median_X_train_filtered <= median_threshold_k27 & X_train_attributes_lymphoma$`STLG (SUV*mL/Kg)` > median_threshold_mtv ~ \"H3K27me3_low_STLG_high\",\n",
    "    X_train_attributes_lymphoma$median_X_train_filtered > median_threshold_k27 & X_train_attributes_lymphoma$`STLG (SUV*mL/Kg)` <= median_threshold_mtv ~ \"H3K27me3_high_STLG_low\",\n",
    "    X_train_attributes_lymphoma$median_X_train_filtered > median_threshold_k27 & X_train_attributes_lymphoma$`STLG (SUV*mL/Kg)` > median_threshold_mtv ~ \"H3K27me3_high_STLG_high\"\n",
    "))\n",
    "\n",
    "# \n",
    "km_fit <- survfit(survival_object ~ threshold, data = X_train_attributes_lymphoma)\n",
    "\n",
    "# plot Kaplan Meier. Healthy samples were checked for exclusion from this plot.\n",
    "ggsurvplot(km_fit, \n",
    "                data = X_train_attributes_lymphoma, \n",
    "                pval = TRUE,\n",
    "                conf.int = TRUE,\n",
    "                ggtheme = theme_minimal(),\n",
    "                risk.table = TRUE,\n",
    "                xlab = \"Time (Years)\",\n",
    "                #ylab = \"Overall Survival\",\n",
    "                ylab = \"Progression Free Survival (PFS)\"\n",
    "                )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecd4728d-851b-4f1e-82a5-a8840307b68f",
   "metadata": {},
   "source": [
    "- Didn’t work as expected, considering some of the low low cases are passing / progressing before the mixed cases. Fortunately, high high definitely predicts the worst prognosis. Think about this some more."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "224ee850-5705-4c6a-8f35-0077ec437f41",
   "metadata": {},
   "source": [
    "## 14-12-2023 Swimmer plot, survival"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef1bfeeb-e47a-455b-b896-b0839125c677",
   "metadata": {},
   "outputs": [],
   "source": [
    "# subset to survival data\n",
    "survival_data <- X_train_attributes_lymphoma[,c(\"PFS\", \"CODE_PFS\")]\n",
    "colnames(survival_data) <- c(\"time\", \"status\")\n",
    "\n",
    "# Generate a swimmer plot\n",
    "survminer::ggsurvplot(\n",
    "  survival_object,\n",
    "  data = survival_data,  # your_data should contain both time and status columns\n",
    "  risk.table = TRUE,\n",
    "  pval = TRUE,\n",
    "  pval.coord = c(0, 0.1),  # Adjust the position of the p-value on the plot\n",
    "  conf.int = TRUE\n",
    ")\n",
    "\n",
    "# Print the swimmer plot\n",
    "print(swimmer_plot)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6bb14e48-b053-413f-9adf-244758376a31",
   "metadata": {},
   "source": [
    "# 11-09-2023 Repeat ML framework (Differential + RF) without SMOTE, see if results are better than previous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed0b486e-9b76-4b47-b3ac-ccf695465cbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# No SMOTE results\n",
    "# load dataset\n",
    "setwd(\"/cluster/projects/scottgroup/people/steven/human_plasma_cfChIP-Seq/liberate_plasma_cfchip/analysis/machine_learning_models/02_07_2023_RFE_RF_10kb_windows/differential_analysis_filter/\")\n",
    "load(\"11_09_2023_rf_10kb_with_hyperparameter_tuning_accuracyevaluation_withdifferentialfilter_NO_SMOTE.RData\")\n",
    "\n",
    "# check model output\n",
    "print(model)\n",
    "predictors_rf <- predictors(model)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c04a43a-3d0d-440e-8143-ac3ed80d7a73",
   "metadata": {},
   "source": [
    "- While SMOTE may reduce generalizability of the model for the minority set, performance is far better with SMOTE so stick with it for now.\n",
    "- Apply the current strategy for subtyping question."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f495712a-388d-4bdc-adb3-a91df02b0b23",
   "metadata": {},
   "source": [
    "# 11-09-2023 Repeat ML framework (Differential + RF) for subtyping question, see if results are better than previous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f3658e8-8f23-4754-8af0-b154558c38ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# results\n",
    "# load dataset\n",
    "setwd(\"/cluster/projects/scottgroup/people/steven/human_plasma_cfChIP-Seq/liberate_plasma_cfchip/analysis/machine_learning_models/02_07_2023_RFE_RF_10kb_windows/differential_analysis_filter/\")\n",
    "load(\"13_09_2023_subtyping_rf_10kb_with_hyperparameter_tuning_accuracyevaluation_withdifferentialfilter.RData\")\n",
    "\n",
    "# model results\n",
    "print(model)\n",
    "predictors_rf_subtype <- predictors(model)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12b2a90a-1dfd-411a-97ff-2a7639a00568",
   "metadata": {},
   "source": [
    "- Okay, not bad, not great. Probably requires the one vs all approach Althaf was discussing. I'll leave this to the future directions for now."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38b0e0e1-7c54-4c2f-a432-726aa4e13191",
   "metadata": {},
   "source": [
    "# 07-11-2023 Figure 1 as proposed - Comparisons to other sequencing types from plasma and PBMC"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2dfc4854-1c23-479a-87ea-8f1ab7be307b",
   "metadata": {},
   "source": [
    "## Difference in fragment length distributions between data types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "050ef768-318c-4aef-babf-b9be7a798c8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# goal here is to, using Picard fragment length frequencies, to figure out whether there is a significant difference in the fragment length distributions between\n",
    "# plasma data types (plasma K27me3, cfMeDIP-Seq, CAPP-Seq (eventually), and WGS (DELFI)).\n",
    "# only healthy profiles are used for this analysis, to reduce variability.\n",
    "\n",
    "#loop over directories and call all fragment length profiles.\n",
    "k27me3_profiles <- list.files(path = \"/cluster/projects/scottgroup/people/steven/human_plasma_cfChIP-Seq/liberate_plasma_cfchip/.\", \n",
    "                              pattern = \"insert_size_metrics.txt\", \n",
    "                              recursive = TRUE,\n",
    "                              full.names = TRUE) # list the full paths of all plasma H3K27me3 profiles\n",
    "k27me3_profiles <- k27me3_profiles[str_detect(k27me3_profiles, \"HUCON\")] # subset these to only non-lymphoma profiles; HUCON.\n",
    "\n",
    "wgs_profiles <- list.files(path = \"/cluster/projects/scottgroup/people/steven/reference_sets/delfi_wgs/picard_fragment_profile/.\", \n",
    "                           pattern = \"IS_metrics.txt\",\n",
    "                           full.names = TRUE)\n",
    "\n",
    "medip_profiles <- list.files(path = \"/cluster/projects/scottgroup/people/steven/reference_sets/justin_hhp_medip/alignment/samtools_dedup/picard_fragment_profile/.\", \n",
    "                             pattern = \"IS_metrics.txt\",\n",
    "                             full.names = TRUE)\n",
    "\n",
    "k4me3_profiles <- list.files(path = \"/cluster/projects/scottgroup/people/steven/human_plasma_cfChIP-Seq/k4me3/alignment/filter_bam/filter_mapq/fragment_length/\", \n",
    "                             pattern = \"IS_metrics.txt\",\n",
    "                             full.names = TRUE)\n",
    "\n",
    "# combine lists into a single list and continue with calculating fragment frequencies\n",
    "all_profiles <- c(k27me3_profiles,wgs_profiles,medip_profiles,k4me3_profiles)\n",
    "\n",
    "#calculate frequency of fragment length for each of the pilot files\n",
    "frag_length_all <- data.frame()\n",
    "\n",
    "all_rows <- data.frame(insert_size = seq(1, 1000)) #to fill all rows to a fixed number\n",
    "\n",
    "# import files and extract fragment length metrics in the form of fragment frequencies\n",
    "for (i in all_profiles) {\n",
    "  # Get the current working directory\n",
    "  current_dir <- getwd()\n",
    "\n",
    "  # Extract the directory path from the file path\n",
    "  dir_path <- dirname(i)\n",
    "\n",
    "  # Set the working directory to the directory of interest\n",
    "  setwd(dir_path)\n",
    "\n",
    "  # Read the file, and calculate fragment frequencies\n",
    "  temp <- read.table(basename(i), header = TRUE, skip = 10)\n",
    "  temp$sample <- paste0(i)\n",
    "  temp$total_sum <- sum(temp$All_Reads.fr_count)\n",
    "  temp$frequency <- temp$All_Reads.fr_count / temp$total_sum\n",
    "\n",
    "  # Set the working directory back to the original directory\n",
    "  setwd(current_dir)\n",
    "\n",
    "  frag_length_all <- rbind(frag_length_all, temp)\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# mutate data type (plasma H3K27me3, cfMeDIP-Seq, WGS) to each respective row.\n",
    "frag_length_all <- frag_length_all %>%\n",
    "  mutate(data_type = case_when(\n",
    "    str_detect(sample, \"K27me3\") ~ \"healthy plasma H3K27me3, n = 11\",\n",
    "    str_detect(sample, \"h3k4me3\") ~ \"healthy plasma H3K4me3, n = 4\",\n",
    "    str_detect(sample, \"WGS\") ~ \"DELFI healthy WGS, n = 6\",\n",
    "    str_detect(sample, \"medip\") ~ \"healthy cfMeDIP-Seq, n = 4\",\n",
    "    ))\n",
    "\n",
    "# group by data type, and calculate mean frequency across groups.\n",
    "frag_length_grouped_median <- frag_length_all %>%\n",
    "group_by(data_type, insert_size) %>%\n",
    "summarize(median_frequency = median(frequency))\n",
    "\n",
    "# plot\n",
    "frag_length_grouped_median %>%\n",
    "ggplot(aes(x = insert_size, y = median_frequency, colour = data_type)) +\n",
    "geom_line() +\n",
    "theme_minimal(base_size = 16) +\n",
    "labs(x = \"Fragment Size\",\n",
    "    y = \"Fragment Frequency\",\n",
    "    colour = \"Data Type\") +\n",
    "scale_x_continuous(limits = c(100,250)) # definitely more dinucleosome enrichment\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e96cc813-f006-4431-b304-bdf9cebe72a4",
   "metadata": {},
   "source": [
    "- Need to fix the for loop, to call the different directories that the files are contained within."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11aa4e5f-b39c-499a-94c2-56b1f56727f6",
   "metadata": {},
   "source": [
    "## Compasrison between healthy plasma H3K27me3 profile (HUCON) and PBMC H3K27me3 profile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e81d46ea-e178-4974-94c3-406edb9fdd83",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import counts from one of the PBMCs and healthy profiles (HUCON). Subset to bins after filtering.\n",
    "setwd(\"/cluster/projects/scottgroup/people/steven/reference_sets/pbmc_k27me3_Yang_et_al_2022/counts_10kb/remove_headers\")\n",
    "import_counts <- list.files(path = \".\", pattern = \"counts_10kb.txt$\") #list counts data, 10kb windows\n",
    "\n",
    "pbmc_counts <- mapply(read.delim, import_counts) #loop over files in import_counts, and read them into memory\n",
    "pbmc_counts_matrix <- do.call(cbind, pbmc_counts) #bind individual datasets across columns\n",
    "rownames(pbmc_counts_matrix) <- 1:nrow(pbmc_counts_matrix) # index rows\n",
    "\n",
    "# subset to bins from above analyses\n",
    "pbmc_counts_matrix_sub <- pbmc_counts_matrix[rownames(pbmc_counts_matrix) %in% de_features_all,]\n",
    "\n",
    "# compare global counts to one another in 10Kb bins\n",
    "# cbind counts from PBMCs to original matrix. Then, CPM normalize and scale as prior.\n",
    "all_filtered_features <- cbind(all_filtered_features, pbmc_counts_matrix_sub)\n",
    "\n",
    "cpm_norm_all <- cpm(all_filtered_features, normalized.lib.sizes = TRUE, log = FALSE)\n",
    "cpm_norm_all <- as.data.frame(cpm_norm_all)\n",
    "\n",
    "cpm_norm_all_scaled <- scale(cpm_norm_all) # try without scaling the values first\n",
    "cpm_norm_all_scaled <- as.data.frame(cpm_norm_all_scaled)\n",
    "\n",
    "\n",
    "cpm_norm_all_scaled %>%\n",
    "ggplot(aes(x = liberate_HUCON_45_K27me3, y = SRR21893770_pbmc_k27_control5_dedup_counts_10kb.txt.X0)) +\n",
    "geom_point(alpha = 0.1, colour = \"#09456b\") +\n",
    "stat_smooth(method = \"lm\", formula = y ~ x, color = \"red\", se = TRUE) + \n",
    "theme_minimal(base_size = 20) +\n",
    "labs(x = \"Healthy Donor Plasma H3K27me3 #1\",\n",
    "    y = \"PBMC H3K27me3\") +\n",
    "scale_x_log10() +\n",
    "scale_y_log10()\n",
    "\n",
    "# linear correlation\n",
    "summary(lm(liberate_HUCON_45_K27me3 ~ SRR21893770_pbmc_k27_control5_dedup_counts_10kb.txt.X0, data = cpm_norm_all_scaled)) # looks good, visualize it\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d66e477-ddf9-4f9e-a74d-52c4d3b17585",
   "metadata": {},
   "source": [
    "- The linear correlation with PBMCs is not great, as they are much higher in expression. Consider using another leukocyte set, for instance the one from Sadeh et al."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "076a09c5-4f2d-4838-bb0f-fa1f7424fd30",
   "metadata": {},
   "source": [
    "## Comparison between healthy plasma profiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54728080-5a08-4ca4-ab2a-dd565da8a94e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# linear correlation between healthy plasma profiles (HUCON)\n",
    "df_cpm_norm_scaled <- as.data.frame(cpm_norm_scaled)\n",
    "\n",
    "summary(lm(liberate_HUCON_42_K27me3 ~ liberate_HUCON_45_K27me3, data = df_cpm_norm_scaled)) # looks good, visualize it\n",
    "\n",
    "df_cpm_norm_scaled %>%\n",
    "# ggplot(aes(x = liberate_HUCON_39_K27me3, y = liberate_HUCON_45_K27me3)) +\n",
    "# ggplot(aes(x = liberate_HUCON_39_K27me3, y = liberate_HUCON_42_K27me3)) +\n",
    "ggplot(aes(x = liberate_HUCON_45_K27me3, y = liberate_HUCON_42_K27me3)) +\n",
    "geom_point(alpha = 0.1, colour = \"#09456b\") +\n",
    "stat_smooth(method = \"lm\", formula = y ~ x, color = \"red\", se = TRUE) + \n",
    "theme_minimal(base_size = 20) +\n",
    "labs(x = \"Healthy Donor Plasma H3K27me3 #1\",\n",
    "    y = \"Healthy Donor Plasma H3K27me3 #2\") +\n",
    "scale_x_log10() +\n",
    "scale_y_log10()\n",
    "\n",
    "\n",
    "\n",
    "# spearman correlation across all healthy profiles in the study, including HUCON and LIBERATE controls\n",
    "# subset matrix to only columns of healthy samples\n",
    "cpm_norm_scaled <- as.data.frame(cpm_norm_scaled)\n",
    "cpm_norm_scaled_healthy <- cpm_norm_scaled %>%\n",
    "select(matches(\"HUCON|_C_\"))\n",
    "\n",
    "# convert back to matrix for Spearman correlation\n",
    "cpm_norm_scaled_healthy <- as.matrix(cpm_norm_scaled_healthy)\n",
    "\n",
    "cor_matrix <- cor(cpm_norm_scaled_healthy, method = \"spearman\")\n",
    "cor_matrix <- as.data.frame(as.table(cor_matrix))\n",
    "\n",
    "# Create a heatmap of the correlation matrix\n",
    "cor_matrix %>%\n",
    "ggplot(aes(x = Var1, y = Var2, fill = Freq)) +\n",
    "geom_tile() +\n",
    "scale_fill_gradient() +\n",
    "theme_minimal(base_size = 20) +\n",
    "theme(axis.text.x = element_text(angle = 45, hjust = 1)) +\n",
    "labs(x = \"Sample\", \n",
    "     y = \"Sample\", \n",
    "     fill = \"Correlation\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f3dc541-4fdd-4ace-8f38-d89ab5dbbacf",
   "metadata": {},
   "source": [
    "## 09-01-2024 heatmap of fragment length profiles for all healthy and lymphoma cases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "772cbdd4-5eda-4cf7-bcce-85bf8cbdcb1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# carry code from R\n",
    "setwd(\"~/Library/CloudStorage/OneDrive-UniversityofToronto/1 MBP/1 Bratman Lab/1 Experiments/11 Sequencing/19-08-2022_k4_k27me3_troubleshooting/bam/k27me3/fragment_length\")\n",
    "setwd(\"~/Library/CloudStorage/OneDrive-UniversityofToronto/1 MBP/1 Bratman Lab/1 Experiments/11 Sequencing/DLBCL_Cohort/medipipe_output/221124-8/fragment_length\")\n",
    "\n",
    "k27me3_profiles <- list.files(path = \".\", pattern = \".txt\")\n",
    "\n",
    "#loop over all subdirectories and call all fragment length profiles.\n",
    "k27me3_profiles <- list.files(path = \"~/OneDrive - University of Toronto/1 MBP/1 Bratman Lab/1 Experiments/11 Sequencing/DLBCL_Cohort/medipipe_output/\", pattern=\"insert_size_metrics.txt\", recursive=TRUE)\n",
    "\n",
    "setwd(\"~/Library/CloudStorage/OneDrive-UniversityofToronto/1 MBP/1 Bratman Lab/1 Experiments/11 Sequencing/DLBCL_Cohort/medipipe_output\")\n",
    "\n",
    "#calculate frequency of fragment length for each of the pilot files\n",
    "frag_length_all <- data.frame()\n",
    "\n",
    "all_rows <- data.frame(insert_size = seq(1, 1000)) #to fill all rows to a fixed number\n",
    "\n",
    "#for this loop, if extracting fragment length characteristics for violin plot, exclude row filling; only use for stacked histogram! Otherwise NA = error.\n",
    "for (i in k27me3_profiles) {\n",
    "  temp <- read.table(i, header = TRUE, skip = 10)\n",
    "  temp$sample <- paste0(i)\n",
    "  temp$total_sum <- sum(temp$All_Reads.fr_count)\n",
    "  temp$frequency <- temp$All_Reads.fr_count / temp$total_sum\n",
    "  #temp <- merge(all_rows, temp, all = TRUE) #merge dummy frame and temp to make temp a fixed number of rows\n",
    "  #temp <- temp %>% \n",
    "  #  fill(sample, total_sum, .direction = c(\"updown\")) #fill NAs in repeated character/numerical rows with values above and below\n",
    "  #temp$sample <- str_match(temp$sample, \"liberate_(.*?)_K27me3\")[[1]] #make \"sample\" column characters a substring of the current characters\n",
    "  frag_length_all <- rbind(frag_length_all,temp)\n",
    "} \n",
    "\n",
    "#colour palate for line plot\n",
    "colour_line <- met.brewer(\"Hokusai1\", 9)\n",
    "#plot_line\n",
    "frag_length_all %>% \n",
    "  ggplot(aes(x = insert_size, y = frequency, colour = sample)) +\n",
    "  geom_line() +\n",
    "  theme_minimal(base_size = 16) +\n",
    "  labs(x = \"Insert Size\",\n",
    "       y = \"Frequency\",\n",
    "       colour = \"H3K27me3 Profile\") +\n",
    "  scale_colour_manual(values = colour_line) +\n",
    "  scale_x_continuous(limits = c(100,200))\n",
    "\n",
    "#plot_alt; stacked histogram\n",
    "frag_length_all %>% \n",
    "  ggplot(aes(x = insert_size, y = sample, fill = frequency)) +\n",
    "  geom_tile() +\n",
    "  theme_minimal(base_size = 10) +\n",
    "  labs(x = \"Insert Size\",\n",
    "       y = \"Sample\",\n",
    "       fill = \"Frequency\") +\n",
    "  scale_fill_gradient(low = \"black\", high = \"#E85e19\", na.value = \"black\") +\n",
    "  scale_x_continuous(limits = c(0,600))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98a9759f-3aab-43ff-bd3f-caddbcc7f622",
   "metadata": {},
   "source": [
    "- Fill missing values with 0s."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "387e64ad-da8c-4607-851d-8670c0fece6d",
   "metadata": {},
   "source": [
    "# 14-11-2023 NMF on lymphoma features: Identification of subgroups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e3200e6-0fe0-4fe3-88d7-e02e46cfa51f",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(NMF)\n",
    "\n",
    "# subset feature matrix to just lymphoma samples\n",
    "X_train_filtered <- as.data.frame(X_train_filtered)\n",
    "X_train_filtered_lymphoma <- X_train_filtered %>% select(-matches(\"HUCON|_C_\"))\n",
    "X_train_filtered_lymphoma <- as.matrix(X_train_filtered_lymphoma)\n",
    "\n",
    "# Assuming your data is in a matrix called 'lymphoma_data'\n",
    "# Rows are features, and columns are samples\n",
    "\n",
    "# Create an NMF model with 3 components (subgroups)\n",
    "nmf_model <- nmf(X_train_filtered_lymphoma, rank = 2, method = \"brunet\")\n",
    "\n",
    "# Extract the basis matrix (features associated with each subgroup)\n",
    "basis_matrix <- basis(nmf_model)\n",
    "\n",
    "# Extract the coefficient matrix (sample memberships for each subgroup)\n",
    "coef_matrix <- coef(nmf_model)\n",
    "\n",
    "# Visualize the coefficient matrix to assign samples to subgroups\n",
    "pheatmap(coef_matrix, cluster_cols = TRUE, cluster_rows = FALSE, \n",
    "         col = heat.colors(10), display_numbers = FALSE,\n",
    "         main = \"NMF Coefficient Matrix (Group Membership)\",\n",
    "         cellwidth = 15, cellheight = 15,\n",
    "         angle_col = 45)  # This rotates the column labels by 45 degrees\n",
    "\n",
    "# Access the cluster assignments for each sample\n",
    "cluster_assignments <- max.col(coef_matrix)\n",
    "\n",
    "# Add the cluster assignments to the original data\n",
    "lymphoma_data_with_clusters <- cbind(X_train_filtered_lymphoma, Cluster = cluster_assignments)\n",
    "\n",
    "# View the result\n",
    "head(lymphoma_data_with_clusters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df605bde-4bd3-4279-bfd9-e8bbb5d3f99b",
   "metadata": {},
   "outputs": [],
   "source": [
    "- This seems promising, explore this further.\n",
    "- Try NMF on both hypermethylated and hypomethylated feature sets."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69e34839-bd59-4837-811a-7fa5af7cadf3",
   "metadata": {},
   "source": [
    "# 17-11-2023 Correlation of modeled features with TMTV and ichorCNA from paired patients"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88fe4035-4597-4a36-bb83-e5b4af4b5c25",
   "metadata": {},
   "source": [
    "## import MTV and ichorCNA data, and append to pre-existing data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92909b51-71c9-458b-9f07-bc364d79d220",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import MTV data; data generated by the nuclear medicine team at UHN\n",
    "setwd(\"/cluster/projects/scottgroup/people/steven/human_plasma_cfChIP-Seq/liberate_plasma_cfchip/other_profiles/tmtv_data_lymphoma_cases\")\n",
    "tmtv_lymphoma <- readxl::read_xlsx(\"17-11-2023_PET_scan_data_lymphoma_cases.xlsx\")\n",
    "\n",
    "# import ichorCNA output and match to X_train_attributes in the same way you incorporated TMTV\n",
    "setwd(\"/cluster/projects/scottgroup/people/steven/human_plasma_cfChIP-Seq/liberate_plasma_cfchip/other_profiles/ichorCNA_data_shallow_WGS_lymphoma\")\n",
    "ichorCNA_lymphoma <- read.delim(\"Bratman.ichorCNA.txt\", header = TRUE, sep = \"\\t\")\n",
    "\n",
    "# merge MTV data frame with X_train_attributes\n",
    "X_train_attributes_all_data_types <- merge(X_train_attributes, tmtv_lymphoma, by.x = \"PATIENT_ID\", by.y = \"LIBERATE ID\", all.x = TRUE)\n",
    "\n",
    "# merge ichorCNA data frame with X_train_attributes; this step is where you're gaining duplicate rows!!\n",
    "X_train_attributes_all_data_types <- merge(X_train_attributes_all_data_types, ichorCNA_lymphoma, by.x = \"PATIENT_ID\", by.y = \"PATIENT_ID\", all.x = TRUE)\n",
    "\n",
    "# remove duplicate ichorCNA entries due to accidental duplicate sequencing of some shallow WGS profiles. Selected single representative entry for duplicates.\n",
    "X_train_attributes_all_data_types <- X_train_attributes_all_data_types %>%\n",
    "  distinct(PATIENT_ID, .keep_all = TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf58c67e-978a-4a0b-875b-b1f3ea31c90d",
   "metadata": {},
   "source": [
    "- Fix acquisition of duplicate rows here.\n",
    "- SOLUTION; THERE ARE DUPLICATE MEASUREMENTS FROM ichorCNA.\n",
    "- Work with \"X_train_attributes_all_data_types\" for all univariate and multivariate survival analyses."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34112a88-1fb1-42b9-8ef6-935fc77f8142",
   "metadata": {},
   "source": [
    "## correlation with MTV data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2de68fab-1290-493e-b8ff-6a5bc5397109",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# import MTV data; data generated by the nuclear medicine team at UHN\n",
    "setwd(\"/cluster/projects/scottgroup/people/steven/human_plasma_cfChIP-Seq/liberate_plasma_cfchip/other_profiles/tmtv_data_lymphoma_cases\")\n",
    "tmtv_lymphoma <- readxl::read_xlsx(\"17-11-2023_PET_scan_data_lymphoma_cases.xlsx\")\n",
    "\n",
    "# merge MTV data frame with X_train_attributes\n",
    "X_train_attributes_tmtv <- merge(X_train_attributes, tmtv_lymphoma, by.x = \"PATIENT_ID\", by.y = \"LIBERATE ID\", all.x = TRUE)\n",
    "\n",
    "# visualize correlation between median k27me3 expression and MTV data as scatter plot\n",
    "p <- X_train_attributes_all_data_types %>% \n",
    "  ggplot(aes(x = `STLG (SUV*mL/Kg)`, y = median_X_train_filtered)) +\n",
    "  # ggplot(aes(x = `STLG (SUV*mL/Kg)`, y = mean_VAF)) + # for correlation with CAPP-Seq\n",
    "  geom_point(size = 5) + \n",
    "  stat_smooth(method = \"lm\", formula = y ~ x, color = \"red\", se = TRUE) + \n",
    "  theme_minimal(base_size = 20) +\n",
    "  labs(#x = \"MTV (mL)\",\n",
    "       #x = \"TLG (SUV*mL)\",\n",
    "       #x = \"SMTV (mL/Kg)\",\n",
    "       x = \"STLG (SUV*mL/Kg)\",\n",
    "       y = \"Median of plasma H3K27me3 Hypermethylated Features\"\n",
    "       # y = \"Mean VAF from CAPP-Seq\"\n",
    "      ) #+\n",
    "  # scale_x_log10() +\n",
    "  # scale_y_log10()\n",
    "print(p)\n",
    "\n",
    "# lm stats\n",
    "summary(lm(median_X_train_filtered ~ `MTV (mL)`, data = X_train_attributes_all_data_types))\n",
    "summary(lm(median_X_train_filtered ~ `TLG (SUV*mL)`, data = X_train_attributes_all_data_types))\n",
    "summary(lm(median_X_train_filtered ~ `SMTV (mL/Kg)`, data = X_train_attributes_all_data_types))\n",
    "summary(lm(median_X_train_filtered ~ `STLG (SUV*mL/Kg)`, data = X_train_attributes_all_data_types))\n",
    "summary(lm(mean_VAF ~ `STLG (SUV*mL/Kg)`, data = X_train_attributes_all_data_types))\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a8a837c-c7a1-49a4-aa3b-b9ec54d4c9bf",
   "metadata": {},
   "source": [
    "Standardized uptake value, SUV, (also referred to as the dose uptake ratio, DUR) is a widely used, robust PET quantifier, calculated simply as a ratio of tissue radioactivity concentration (for example in units [kBq/mL]) at time T, CPET(T), and administered dose (for example in units [MBq]) at the time of injection divided by body weight (usually in units [kg]). (http://www.turkupetcentre.net/petanalysis/model_suv.html)\n",
    "\n",
    "TLG is the Total Lesion Glycolysis. Its unit is SUV*mL;\n",
    "• sTLG, the Standardized Total Lesion Glycolysis, corresponds to TLG/PatientWeight.\n",
    "It is measured in SUV*mL/kg;\n",
    "• MTV is the protocol main value, the total Metabolic Tumor Volume. Its unit\n",
    "is mL (or cm3\n",
    ");\n",
    "• sMTV is Standardized Metabolic Tumor Volume. More specifically, it is measured as MTV/PatientWeight and is in mL/kg;\n",
    "• Dmax is the distance between the two lesions that are the furthest apart (with\n",
    "ROIs center of mass as origin). It is one way to measure tumor dissemination;\n",
    "• wDmax is the distance between the two lesions that are the furthest apart\n",
    "calculated from a weigthed center of mass;\n",
    "• DmaxVox is the distance between the two lesions that are the furthest apart\n",
    "(with the outermost voxel).\n",
    "(https://www.lifexsoft.org/images/phocagallery/documentation/ProtocolMTV/UserGuide/MTVUserGuide.pdf)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc208cb4-92e8-498b-9a3f-4376d6b2a387",
   "metadata": {},
   "source": [
    "## 08-12-2023 Correlation of modeled features with tumor burden from ichorCNA from shallow WGS for matched cases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97b51e32-1dd4-49c2-acc6-1c795a77e07a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ADDITIONAL STEP; remove cases where tumor burden from ichorCNA is <5%.\n",
    "X_train_attributes_all_data_types_filtered <- X_train_attributes_all_data_types[X_train_attributes_all_data_types$`Tumor.Fraction` < 0.05,]\n",
    "\n",
    "# visualize correlation between median k27me3 expression and tumor burden from ichorCNA\n",
    "X_train_attributes_all_data_types %>% \n",
    "# X_train_attributes_all_data_types_filtered %>% \n",
    "  ggplot(aes(x = `Tumor.Fraction`, y = median_X_train_filtered)) +\n",
    "  # ggplot(aes(x = `Tumor.Fraction`, y = mean_VAF)) + # for correlation with CAPP-Seq\n",
    "  geom_point(size = 5) + \n",
    "  stat_smooth(method = \"lm\", formula = y ~ x, color = \"red\", se = TRUE) + \n",
    "  theme_minimal(base_size = 20) +\n",
    "  labs(x = \"ichorCNA Tumor Fraction\",\n",
    "       y = \"Median of plasma H3K27me3 Hypermethylated Features\"\n",
    "       # y = \"Mean VAF from CAPP-Seq\"\n",
    "      ) #+\n",
    "  # scale_x_log10() +\n",
    "  # scale_y_log10()\n",
    "\n",
    "# lm stats\n",
    "summary(lm(median_X_train_filtered ~ `Tumor.Fraction`, data = X_train_attributes_all_data_types))\n",
    "# summary(lm(median_X_train_filtered ~ `Tumor.Fraction`, data = X_train_attributes_all_data_types_filtered))\n",
    "summary(lm(mean_VAF ~ `Tumor.Fraction`, data = X_train_attributes_all_data_types))\n",
    "# summary(lm(mean_VAF ~ `Tumor.Fraction`, data = X_train_attributes_all_data_types_filtered))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96f2e88f-3c9a-40b9-9260-aca6b541a02d",
   "metadata": {},
   "source": [
    "# 27-11-2023 UMAP of all filtered features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d23b6118-c6b3-4657-a23b-4f913c6afa2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# apply UMAP to all filtered features in 10Kb bins\n",
    "#Visualize training data as UMAP; see if data clustering is different than PCA, as non-linear combinations of variables are a consideration.\n",
    "#Or consider using all data for this.\n",
    "library(umap)\n",
    "library(Rtsne)\n",
    "\n",
    "# Compute the UMAP embedding\n",
    "umap_all_features <- umap(cpm_norm)\n",
    "\n",
    "# Create a data frame with the UMAP coordinates\n",
    "umap_data <- as.data.frame(umap_all_features$layout)\n",
    "colnames(umap_data) <- c(\"UMAP1\", \"UMAP2\")\n",
    "\n",
    "#cbind metadata to the UMAP embeddings.\n",
    "umap_data <- cbind(umap_data,X_train_attributes)\n",
    "\n",
    "# Plot the UMAP\n",
    "umap_data %>% \n",
    "  ggplot(aes(x = UMAP1, y = UMAP2)) +\n",
    "  # ggplot(aes(x = UMAP1, y = UMAP2, colour = timepoint)) +\n",
    "  # ggplot(aes(x = UMAP1, y = UMAP2, colour = ANN_ARBOR_STAGE)) +\n",
    "  # ggplot(aes(x = UMAP1, y = UMAP2, colour = DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY.DIAGNOSIS_SPECIFIC_DLBCL)) +\n",
    "  geom_point() +\n",
    "  theme_minimal(base_size=20)\n",
    "\n",
    "\n",
    "# Plot the tSNE alone\n",
    "tsne_df <- data.frame(tsne_all_filtered_features_counts$Y)\n",
    "tsne_df <- cbind(tsne_df,all_filtered_features$samples)\n",
    "\n",
    "p <- tsne_df %>% \n",
    "  #ggplot(aes(x = X1, y = X2, colour = timepoint)) +\n",
    "  ggplot(aes(x = X1, y = X2, colour = ANN_ARBOR_STAGE)) +\n",
    "  #ggplot(aes(x = X1, y = X2, colour = DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY.DIAGNOSIS_SPECIFIC_DLBCL)) +\n",
    "  geom_point(size = 5) +\n",
    "  theme_minimal(base_size=15)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bec2b1c8-8412-4ce1-b977-659b98e562b1",
   "metadata": {},
   "source": [
    "- Re-think this analysis; in its current state, it's visualizing all the features over UMAP 1 and 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83f99230-f463-43dc-8212-9171b6f38b6c",
   "metadata": {},
   "source": [
    "# 28-11-2023 hypermethylated vs hypomethylated feature analyses"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d95735cc-8c6b-4817-87d9-1bbfe20ea3ec",
   "metadata": {},
   "source": [
    "## PCA - hyper vs hypo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08a51103-5ee8-48fc-bcee-40367227d127",
   "metadata": {},
   "outputs": [],
   "source": [
    "# compare significant hypermethylated and hypomethylated features via PCA. Parallel to the comparison using NMF, which revealed differences.\n",
    "pca_data <- t(X_train_filtered) # for feature subset; exchange feature set where the boxplot visualizations were above.\n",
    "\n",
    "# perform PCA\n",
    "pca_res <- prcomp(pca_data, scale = FALSE) #Perform PCA.\n",
    "\n",
    "# Plot the first two principal components\n",
    "pca_res$x %>% \n",
    "  as.data.frame %>%\n",
    "  #rownames_to_column(\"cancer_status\") %>%\n",
    "  #separate(cancer_status,c(\"subtype\")) %>%\n",
    "  ggplot(aes(x=PC1,y=PC2)) +\n",
    "  #geom_point(aes(color=list$samples$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY),size=5) +\n",
    "  #geom_point(aes(color=list$samples$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY.DIAGNOSIS_SPECIFIC_DLBCL),size=5) +\n",
    "  #geom_point(aes(color=list$samples$timepoint),size=5) +\n",
    "  geom_point(aes(color=y_train),size=5) +\n",
    "  #geom_point(aes(color=list$samples$batch),size=5) +\n",
    "  # geom_label(label = colnames(X_train_filtered)) + # to label the data points and figure out which two samples were the outliers\n",
    "  theme_minimal(base_size=20) + \n",
    "  labs(colour = \"Group\") +\n",
    "  xlab(paste0(\"PC1 (\", round(pca_res$sdev[1]^2*100/sum(pca_res$sdev^2), 1), \"%)\")) +\n",
    "  ylab(paste0(\"PC2 (\", round(pca_res$sdev[2]^2*100/sum(pca_res$sdev^2), 1), \"%)\")) +\n",
    "  theme(legend.position=\"right\") #+\n",
    "  # scale_x_continuous(limits = c(-20,20)) # zoom into the cluster of healthy and cancer samples +\n",
    "  # scale_y_continuous(limits = c(-35,35)) # zoom into the cluster of healthy and cancer samples\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f304b264-9d59-46c4-b767-da7d8d18b341",
   "metadata": {},
   "source": [
    "- Interestingly, no separation between lymphoma and non-lymphoma cases when hyper or hyopmethylated features are used. But the median expression over the features is informative for groups.\n",
    "- Two non-lymphoma cases which separate in the healthy cases are interesting.\n",
    "- Figure out where these features are located. Also, look at the modeled features for more info.\n",
    "- Side note; HUCON 35 and 36 are major outliers in the modeling and may be driving poor performance in the test set."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f883517d-1e62-41e3-ab7b-e80bcc171505",
   "metadata": {},
   "source": [
    "# 29-11-2023 Gene-specific plasma H3K27me3 expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eee30a21-fa13-4160-83ac-e35d3f1422eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# subset to prognostic genes in lymphoma, see if there are prominent difference across lymphoma cases.\n",
    "\n",
    "# import 10kb reference\n",
    "setwd(\"/cluster/projects/scottgroup/people/steven/reference_sets/other_references\")\n",
    "annotated_ref_no_X_Y_M_blacklist <- read.delim(\"10kb_bin_genome_wide_annotated_reference_with_regulatory_features_no_X_Y_M_ENCODEBlacklist.txt\", header = TRUE, sep = \"\\t\")\n",
    "\n",
    "# import gene reference\n",
    "setwd(\"/cluster/projects/scottgroup/people/steven/reference_sets/UCSC_reference\")\n",
    "gene_reference <- read.delim(\"UCSC_allGENCODE_V44_gene_annotations\", header = TRUE, sep = \"\\t\")\n",
    "\n",
    "#aside; import NFE2 binding sites\n",
    "setwd(\"/cluster/projects/scottgroup/people/steven/reference_sets/tfbs_reference/\")\n",
    "nfe2_reference <- read.delim(\"all_NFE2_tfbs.tsv\", header = FALSE, sep = \"\\t\")\n",
    "colnames(nfe2_reference) <- c(\"chrom\", \"start\", \"end\")\n",
    "\n",
    "# subset reference to only chr1-22\n",
    "gene_reference <- gene_reference[gene_reference$chrom %in% paste0(\"chr\",seq(1:22)),]\n",
    "\n",
    "# subset gene reference to genes of interest\n",
    "# NOTE; change code below to specify genes of interest\n",
    "# gene_reference_subset <- gene_reference[gene_reference$name2 %in% paste0(\"RNA5S\",seq(1:17)),] # subset to gene of interest\n",
    "gene_reference_subset <- gene_reference[gene_reference$name2 %in% paste0(c(\"STAT3\")),] # BCL2 and BCL6 are interesting; outliers\n",
    "\n",
    "# EXTRA CODE; IF YOU WISH TO SUBSET TO A PANEL OF GENES\n",
    "setwd(\"/cluster/projects/scottgroup/people/steven/reference_sets/other_references/gene_lists\")\n",
    "# gene_panel <- read.delim(\"Nanostring_Lymph2Cx_Scott_2014_genes_ALL.txt\", header = FALSE, sep = \"\\t\") # for all\n",
    "# gene_panel <- read.delim(\"epic_seq_Alizadeh_genes_ALL.txt\", header = FALSE, sep = \"\\t\") # for all\n",
    "# gene_panel <- read.delim(\"Schmitz_NEJM_2018_fig1_genes_ALL.txt\", header = FALSE, sep = \"\\t\") # for all\n",
    "# gene_panel <- read.delim(\"Schmitz_NEJM_2018_fig1_genes_GCB_ONLY.txt\", header = FALSE, sep = \"\\t\") # for GCB\n",
    "# gene_panel <- read.delim(\"Schmitz_NEJM_2018_fig1_genes_nonGCB_ONLY.txt\", header = FALSE, sep = \"\\t\") # for non-GCB\n",
    "# gene_reference_subset <- gene_reference[gene_reference$name2 %in% gene_panel$V1,] # for the panels, double check there aren't extra genes\n",
    "\n",
    "# convert gene reference subset and 10kb reference to GRanges objects and check the overlap. Keep bin ID for overlapping ranges\n",
    "gr_gene_reference_subset <- GRanges(gene_reference_subset[,c(3,5,6)])\n",
    "# gr_gene_reference_subset <- GRanges(nfe2_reference[,c(1:3)]) # for TFBS reference\n",
    "gr_annotated_ref_no_X_Y_M_blacklist <- GRanges(annotated_ref_no_X_Y_M_blacklist)\n",
    "\n",
    "# find overlaps in ranges\n",
    "overlap_regions <- findOverlaps(gr_gene_reference_subset, gr_annotated_ref_no_X_Y_M_blacklist)\n",
    "\n",
    "# subset to overlap queries and save bin IDs for subsetting original data matrix (plasma H3K27me3)\n",
    "gr_annotated_ref_no_X_Y_M_blacklist_filtered_to_gene_subset <- gr_annotated_ref_no_X_Y_M_blacklist[subjectHits(overlap_regions)]\n",
    "gene_bins <- unique(gr_annotated_ref_no_X_Y_M_blacklist_filtered_to_gene_subset$bin_id)\n",
    "\n",
    "# subset normalized matrix to bin IDs\n",
    "cpm_norm_genes <- cpm_norm_log2[rownames(cpm_norm_log2) %in% gene_bins,]\n",
    "\n",
    "# take median across bins per sample. ACTIVE IF GENE SPANS MORE THAN ONE 10KB BIN.\n",
    "cpm_norm_genes_median <- as.data.frame(apply(cpm_norm_genes, 2, median))\n",
    "colnames(cpm_norm_genes_median) <- c(\"cpm_norm_genes_median\")\n",
    "\n",
    "# append cpm_norm_genes_median to sample metadata\n",
    "X_train_attributes_update <- cbind(X_train_attributes,cpm_norm_genes_median)\n",
    "\n",
    "# if gene fits in one 10kb bin\n",
    "# X_train_attributes_update <- cbind(X_train_attributes,cpm_norm_genes)\n",
    "\n",
    "# remove FL cases for SCHMIDT et al\n",
    "# X_train_attributes_update <- X_train_attributes_update[!(X_train_attributes_update$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY == \"FL\"),]\n",
    "\n",
    "# incorporate other data types\n",
    "# linear regression with MTV and ichorCNA\n",
    "setwd(\"/cluster/projects/scottgroup/people/steven/human_plasma_cfChIP-Seq/liberate_plasma_cfchip/other_profiles/tmtv_data_lymphoma_cases\")\n",
    "tmtv_lymphoma <- readxl::read_xlsx(\"17-11-2023_PET_scan_data_lymphoma_cases.xlsx\")\n",
    "\n",
    "# import ichorCNA output and match to X_train_attributes in the same way you incorporated TMTV\n",
    "setwd(\"/cluster/projects/scottgroup/people/steven/human_plasma_cfChIP-Seq/liberate_plasma_cfchip/other_profiles/ichorCNA_data_shallow_WGS_lymphoma\")\n",
    "ichorCNA_lymphoma <- read.delim(\"Bratman.ichorCNA.txt\", header = TRUE, sep = \"\\t\")\n",
    "\n",
    "# merge MTV data frame with X_train_attributes_update\n",
    "X_train_attributes_update <- merge(X_train_attributes_update, tmtv_lymphoma, by.x = \"PATIENT_ID\", by.y = \"LIBERATE ID\", all.x = TRUE)\n",
    "\n",
    "# merge ichorCNA data frame with X_train_attributes; this step is where you're gaining duplicate rows!!\n",
    "X_train_attributes_update <- merge(X_train_attributes_update, ichorCNA_lymphoma, by.x = \"PATIENT_ID\", by.y = \"PATIENT_ID\", all.x = TRUE)\n",
    "\n",
    "# remove duplicate ichorCNA entries due to accidental duplicate sequencing of some shallow WGS profiles. Selected single representative entry for duplicates.\n",
    "X_train_attributes_update <- X_train_attributes_update %>%\n",
    "  distinct(PATIENT_ID, .keep_all = TRUE)\n",
    "\n",
    "# visualize across healthy and cancer conditions\n",
    "X_train_attributes_update %>% \n",
    "  # ggplot(aes(x = `DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY+DIAGNOSIS_SPECIFIC_DLBCL`, y = cpm_norm_genes_median)) +#change to sum or median, depending on what you're visualizing\n",
    "  ggplot(aes(x = Stage_Group, y = cpm_norm_genes_median)) + # when gene spans >1 bin\n",
    "  # ggplot(aes(x = Stage_Group, y = cpm_norm_genes)) + # when gene spans one bin\n",
    "  # geom_boxplot(aes(fill = `DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY+DIAGNOSIS_SPECIFIC_DLBCL`), alpha = 0.3) +\n",
    "  geom_boxplot(aes(fill = Stage_Group), alpha = 0.3) +\n",
    "  geom_point(aes(colour = mean_VAF), position = position_jitterdodge(0.2), size = 7, alpha = 0.6) + \n",
    "  theme_minimal(base_size = 18) +\n",
    "  #scale_color_gradient(low = \"#185a9d\", high = \"#43cea2\") +\n",
    "  scale_color_gradient(low = \"blue\", high = \"red\") +\n",
    "  labs(x = \"Subtype\",\n",
    "       #x = \"IPI\",\n",
    "       # y = \"Median of Hypermethylated Feature Expression\",\n",
    "       # y = \"Median of ALL Hypermethylated Feature Expression\",\n",
    "       y = \"Median of H3K27me3 expression over gene\",\n",
    "       colour = \"mean VAF\", \n",
    "       # fill = \"Stage\"\n",
    "       fill = \"Stage\"\n",
    "       #colour = \"IPI\"\n",
    "       )\n",
    "  #ggtitle(\"Median Hypermethylated Counts of Differential Features\")\n",
    "  #ggtitle(\"Median of Counts over Features\")\n",
    "  #ggtitle(\"Sum of Counts over Modeled Features\")\n",
    "  ggtitle(\"Median of Counts over Genes\") # change gene name if required\n",
    "\n",
    "# linear regression with CAPP-Seq\n",
    "X_train_attributes_update %>% \n",
    "  ggplot(aes(x = mean_VAF, y = cpm_norm_genes_median)) +\n",
    "  # ggplot(aes(x = mean_VAF, y = cpm_norm_genes)) +\n",
    "  # ggplot(aes(x = `STLG (SUV*mL/Kg)`, y = cpm_norm_genes_median)) +\n",
    "  # ggplot(aes(x = `Tumor.Fraction`, y = cpm_norm_genes_median)) +\n",
    "  geom_point(size = 5) + \n",
    "  stat_smooth(method = \"lm\", formula = y ~ x, color = \"red\", se = TRUE) + \n",
    "  theme_minimal(base_size = 20) +\n",
    "  labs(x = \"Mean VAF\",\n",
    "       y = \"Median of Hypermethylated Feature Expression\"\n",
    "      )\n",
    "\n",
    "summary(lm(cpm_norm_genes_median ~ mean_VAF, X_train_attributes_update))\n",
    "# summary(lm(cpm_norm_genes ~ mean_VAF, X_train_attributes_update)) # for 1 bin gene\n",
    "# summary(lm(cpm_norm_genes_median ~ `STLG (SUV*mL/Kg)`, data = X_train_attributes_update_tmtv))\n",
    "# summary(lm(cpm_norm_genes_median ~ `Tumor.Fraction`, data = X_train_attributes_update_tmtv))\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "##### SURVIVAL #####\n",
    "# remove healthy cases from X_train_attributes\n",
    "X_train_attributes_update <- X_train_attributes_update[!(X_train_attributes_update$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY == \"HEALTHY\"),]\n",
    "\n",
    "# convert PFS and OS to numeric\n",
    "X_train_attributes_update$PFS <- as.numeric(X_train_attributes_update$PFS)\n",
    "X_train_attributes_update$OS <- as.numeric(X_train_attributes_update$OS)\n",
    "X_train_attributes_update$CODE_PFS <- as.numeric(X_train_attributes_update$CODE_PFS)\n",
    "X_train_attributes_update$CODE_OS <- as.numeric(X_train_attributes_update$CODE_OS)\n",
    "\n",
    "# survival based off thresholds of gene K27 expression\n",
    "# median_threshold <- median(X_train_attributes_update$cpm_norm_genes_median[!(X_train_attributes_update$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY == \"HEALTHY\")]) # for individual genes\n",
    "mean_threshold <- median(X_train_attributes_update$cpm_norm_genes_median[!(X_train_attributes_update$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY == \"HEALTHY\")]) # for individual genes\n",
    "\n",
    "library(survival)\n",
    "library(survminer)\n",
    "\n",
    "# determine optimal cutoff using survminer surv_cutpoint\n",
    "# cutpoint <- surv_cutpoint(data = X_train_attributes_lymphoma, \n",
    "#                           time = \"PFS\", \n",
    "#                           event = \"CODE_PFS\",\n",
    "#                           variables = \"median_X_train_filtered\")\n",
    "\n",
    "# cutpoint <- cutpoint$cutpoint$cutpoint # this is the stupidest functional code I've ever written\n",
    "\n",
    "# create survival object\n",
    "survival_object <- Surv(time = X_train_attributes_update$OS, event = X_train_attributes_update$CODE_OS) # for overall survival\n",
    "# survival_object <- Surv(time = X_train_attributes_lymphoma$PFS, event = X_train_attributes_lymphoma$CODE_PFS) # for progression free survival\n",
    "\n",
    "# might have to remove NAs (aka healthy samples)\n",
    "# high = above threshold, low = below threshold\n",
    "# X_train_attributes_update$threshold <- ifelse(X_train_attributes_update$cpm_norm_genes_median > median_threshold, \"High\", \"Low\")\n",
    "X_train_attributes_update$threshold <- ifelse(X_train_attributes_update$cpm_norm_genes_median > mean_threshold, \"High\", \"Low\")\n",
    "\n",
    "# \n",
    "km_fit <- survfit(survival_object ~ threshold, data = X_train_attributes_update)\n",
    "\n",
    "# plot Kaplan Meier. Healthy samples were checked for exclusion from this plot.\n",
    "survplot <- ggsurvplot(km_fit, \n",
    "                data = X_train_attributes_update, \n",
    "                pval = TRUE,\n",
    "                conf.int = TRUE,\n",
    "                ggtheme = theme_minimal(),\n",
    "                risk.table = TRUE,\n",
    "                xlab = \"Time (Years)\",\n",
    "                #ylab = \"Overall Survival\",\n",
    "                ylab = \"Progression Free Survival (PFS)\"\n",
    "                )\n",
    "survplot\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01cc7c5f-f6d8-4030-a652-4142aae7c7f5",
   "metadata": {},
   "source": [
    "- first, check 5S ribosomal RNAs as a positive control region bw cancer and healthy\n",
    "- RNA5S1-17; chr1 228610268 - 228646158; from Ensembl\n",
    "- Similar median across the groups, but so variable within groups.\n",
    "- Interesting finding though, when we map the mean VAF onto the data points, high burden samples have the lowest expression of bins converging with RNA5S. Is this because K27me3 is generally higher over other regions so less reads in this region?\n",
    "- See if this is the case for other genes which we'd expect to be different in the cancer cases\n",
    "- Consider looking at entire Schmitz et al gene panel to see if they segregate the subtypes.\n",
    "- Also consider taking the sum, but this will be influenced by gene size.\n",
    "- You need to split the Schmitz list into GCB vs non-GCB enriched. Then test.\n",
    "\n",
    "- 15-12-2023; look into TFBS for NFE2, Sasha recommended the database the Griffin paper used. She downloaded the dataset to the cluster, I can subset to all sites targeted by NFE2. These sites have chromosomal locations, so I can subset to these regions in 10kb K27me3 bins and look at their joint expression.\n",
    "- Follow up on current NFE2 targets; lots of sites, but will likely be compiled into overlapping 10kb bins. UPDATE; 149235 BINS IN TOTAL WITH NFE2 TFBS; this is too many.\n",
    "- Stick with the discovery platform discussed at the lymphoma meeting."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38fcf431-5a0f-4431-9602-057d1734355e",
   "metadata": {},
   "source": [
    "## Stats - Welch's t-test between groups for different genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc278502-d105-4515-a5e3-c23110912f01",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Welch's t-test\n",
    "\n",
    "# for stage group\n",
    "stats_healthy <- X_train_attributes_update[X_train_attributes_update$Stage_Group == \"non-lymphoma\",]\n",
    "stats_early <- X_train_attributes_update[X_train_attributes_update$Stage_Group == \"Early Stage\",]\n",
    "stats_late <- X_train_attributes_update[X_train_attributes_update$Stage_Group == \"Late Stage\",]\n",
    "\n",
    "t.test(stats_healthy$cpm_norm_genes_median, stats_early$cpm_norm_genes_median, var.equal = FALSE)\n",
    "t.test(stats_healthy$cpm_norm_genes_median, stats_late$cpm_norm_genes_median, var.equal = FALSE)\n",
    "t.test(stats_early$cpm_norm_genes_median, stats_late$cpm_norm_genes_median, var.equal = FALSE)\n",
    "\n",
    "# for GCB vs non-GCB\n",
    "stats_gcb <- X_train_attributes_update[X_train_attributes_update$`DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY+DIAGNOSIS_SPECIFIC_DLBCL` == \"GCB\",]\n",
    "stats_non_gcb <- X_train_attributes_update[X_train_attributes_update$`DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY+DIAGNOSIS_SPECIFIC_DLBCL` == \"non-GCB\",]\n",
    "stats_fl <- X_train_attributes_update[X_train_attributes_update$`DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY+DIAGNOSIS_SPECIFIC_DLBCL` == \"FL\",]\n",
    "\n",
    "t.test(stats_gcb$cpm_norm_genes_median, stats_non_gcb$cpm_norm_genes_median, var.equal = FALSE)\n",
    "t.test(stats_fl$cpm_norm_genes_median, stats_non_gcb$cpm_norm_genes_median, var.equal = FALSE)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbbac542-fba9-4945-9002-40a9498f7e8c",
   "metadata": {},
   "source": [
    "RNA5S genes: p = 0.2469\n",
    "BCL6: p = 0.1236\n",
    "BCL2: p = 0.0791\n",
    "\n",
    "GCB VS NON-GCB:\n",
    "NOTCH2: P = 0.07998\n",
    "STAT6: P = 0.05235 ; consider combining a couple of these genes? Or all genes from the Schmitz NEJM paper.\n",
    "EZH2: P = 0.3845; even though the effect size looks larger, maybe median isn't appropriate here?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fcbce8ef-5c57-4901-a717-838e2fded9fb",
   "metadata": {},
   "source": [
    "## DO NOT RUN; code to generate 10kb reference with gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84dd48fa-e6d4-40b2-872e-8de0963a881e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# libraries\n",
    "library(annotatr)\n",
    "\n",
    "# import 10kb/100kb reference\n",
    "setwd(\"~/OneDrive - University of Toronto/1 MBP/1 Bratman Lab/1 Experiments/11 Sequencing/Reference/other_reference\")\n",
    "ref_10kb <- read.delim(\"10kb_bin_genome_wide_genomic_ranges_reference_raw.txt\", header = TRUE, sep = \"\\t\")\n",
    "#ref_100kb <- read.delim(\"100kb_bin_genome_wide_genomic_ranges_reference_raw.txt\", header = TRUE, sep = \"\\t\")\n",
    "\n",
    "# annotate using annotatR\n",
    "# subset to only chrom, start, end\n",
    "bin_annotations_regions <- ref_10kb[,1:3] # change, depending on which reference you use\n",
    "#bin_annotations_regions <- ref_100kb[,1:3]\n",
    "bin_annotations_regions <- setNames(bin_annotations_regions, c(\"seqnames\",\"start\",\"end\"))\n",
    "bin_annotations_regions_gr <- GRanges(bin_annotations_regions)\n",
    "\n",
    "#Create three separate annotation sets as columns; CpG annotations, noncoding regulatory, coding regulatory.\n",
    "annotations_hg38_basicgenes <- c(\"hg38_basicgenes\")\n",
    "annotations_hg38_genic <- c(\"hg38_genes_1to5kb\",\n",
    "                            \"hg38_genes_promoters\", \n",
    "                            \"hg38_genes_5UTRs\", \n",
    "                            \"hg38_genes_exons\",\n",
    "                            \"hg38_genes_firstexons\",\n",
    "                            \"hg38_genes_introns\",\n",
    "                            \"hg38_genes_intronexonboundaries\",\n",
    "                            \"hg38_genes_exonintronboundaries\",\n",
    "                            \"hg38_genes_3UTRs\",\n",
    "                            \"hg38_genes_intergenic\")\n",
    "annotations_hg38_enhancers <- c(\"hg38_enhancers_fantom\")\n",
    "\n",
    "annotation_build_hg38_cpg <- build_annotations(genome = \"hg38\", annotations = \"hg38_cpgs\") #basic annotation preset; CpG features\n",
    "annotation_build_hg38_genic <- build_annotations(genome = \"hg38\", annotations = annotations_hg38_genic) #basic annotation preset; genic features\n",
    "annotation_build_hg38_basicgenes <- build_annotations(genome = \"hg38\", annotations = annotations_hg38_basicgenes) #basic annotation preset; genic features\n",
    "annotation_build_hg38_enhancers <- build_annotations(genome = \"hg38\", annotations = annotations_hg38_enhancers) #annotate enhancers (FANTOM5 w liftover)\n",
    "\n",
    "\n",
    "#annotate regions\n",
    "annotate_300bp_bins_hg38_cpg <- annotate_regions(regions = bin_annotations_regions_gr,\n",
    "                                                 annotations = annotation_build_hg38_cpg,\n",
    "                                                 quiet = FALSE)\n",
    "unique_annotate_300bp_bins_hg38_cpg <- unique(annotate_300bp_bins_hg38_cpg) #remove duplicate annotations\n",
    "\n",
    "annotate_300bp_bins_hg38_genic <- annotate_regions(regions = bin_annotations_regions_gr,\n",
    "                                                 annotations = annotation_build_hg38_genic,\n",
    "                                                 quiet = FALSE)\n",
    "unique_annotate_300bp_bins_hg38_genic <- unique(annotate_300bp_bins_hg38_genic) #remove duplicate annotations\n",
    "\n",
    "annotate_300bp_bins_hg38_enhancers <- annotate_regions(regions = bin_annotations_regions_gr,\n",
    "                                                 annotations = annotation_build_hg38_enhancers,\n",
    "                                                 quiet = FALSE)\n",
    "unique_annotate_300bp_bins_hg38_enhancers <- unique(annotate_300bp_bins_hg38_enhancers) #remove duplicate annotations\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# gene annotations require additional code, due to duplicate genes in bins\n",
    "# \n",
    "annotate_300bp_bins_hg38_basicgenes <- annotate_regions(regions = bin_annotations_regions_gr,\n",
    "                                                 annotations = annotation_build_hg38_basicgenes,\n",
    "                                                 quiet = FALSE)\n",
    "\n",
    "# convert to data frame\n",
    "annotate_300bp_bins_hg38_basicgenes <- as.data.frame(annotate_300bp_bins_hg38_basicgenes) # the 10kb reference now has bin IDs from start\n",
    "\n",
    "# remove rows where annot.symbol == \"NA\"\n",
    "annotate_300bp_bins_hg38_basicgenes <- annotate_300bp_bins_hg38_basicgenes[!(is.na(annotate_300bp_bins_hg38_basicgenes$annot.symbol)),]\n",
    "\n",
    "# use dplyr for this, select by bin_ID, and apply some function to collapse and remove duplicates of annot.symbol\n",
    "gene_names_per_bin <- annotate_300bp_bins_hg38_basicgenes %>% \n",
    "  group_by(bin_id) %>% \n",
    "  summarize(gene_names = paste(unique(annot.symbol), collapse = \", \")) # separates genes in a single column with a comma\n",
    "\n",
    "gene_names_per_bin <- as.data.frame(gene_names_per_bin)\n",
    "\n",
    "\n",
    "#Convert one of the GRanges objects back to a data frame. Then do this for the remainder.\n",
    "df_unique_annotate_300bp_bins_hg38_cpg <- as.data.frame(unique_annotate_300bp_bins_hg38_cpg)\n",
    "df_unique_annotate_300bp_bins_hg38_genic <- as.data.frame(unique_annotate_300bp_bins_hg38_genic)\n",
    "df_unique_annotate_300bp_bins_hg38_enhancers <- as.data.frame(unique_annotate_300bp_bins_hg38_enhancers)\n",
    "\n",
    "\n",
    "#Match the annotated data frame to \n",
    "df_merged_cpg <- merge(x = bin_annotations_regions, y = df_unique_annotate_300bp_bins_hg38_cpg, by = c(\"seqnames\", \"start\", \"end\"), all = TRUE)\n",
    "df_merged_genic <- merge(x = bin_annotations_regions, y = df_unique_annotate_300bp_bins_hg38_genic, by = c(\"seqnames\", \"start\", \"end\"), all = TRUE)\n",
    "df_merged_enhancers <- merge(x = bin_annotations_regions, y = df_unique_annotate_300bp_bins_hg38_enhancers, \n",
    "                             by = c(\"seqnames\", \"start\", \"end\"), all = TRUE)\n",
    "\n",
    "\n",
    "#Bind new annotated features to original data frame of annotated bins.\n",
    "bin_annotations_final <- cbind(ref_10kb,df_merged_cpg$annot.type,df_merged_genic$annot.type,df_merged_enhancers$annot.type)\n",
    "#bin_annotations_final <- cbind(ref_100kb,df_merged_cpg$annot.type,df_merged_genic$annot.type,df_merged_enhancers$annot.type)\n",
    "\n",
    "# merge bin_annotations_final and gene_names_per_bin, as there are many bins missing from gene_names_per_bin, since not all bins have genes.\n",
    "bin_annotations_final <- merge(bin_annotations_final, gene_names_per_bin, by = \"bin_id\", all.x = TRUE)\n",
    "\n",
    "bin_annotations_final <- data.frame(lapply(bin_annotations_final, function(x) ifelse(is.na(x), \"other\", x))) #make all NAs in the data frame \"other\".\n",
    "\n",
    "bin_annotations_final <- setNames(bin_annotations_final,\n",
    "                                  c(\"chrom\",\"start\",\"end\",\"width\",\"strand\",\n",
    "                                    \"bin_id\",\"cpg_annotations\",\"genic_annotations\",\"enhancer_annotations\", \"gene_name\"))\n",
    "\n",
    "#export annotation set at text file\n",
    "setwd(\"~/Library/CloudStorage/OneDrive-UniversityofToronto/1 MBP/1 Bratman Lab/1 Experiments/11 Sequencing/Reference/other_reference\")\n",
    "#write.table(bin_annotations_final,\n",
    "#            file = \"10kb_bin_genome_wide_annotated_reference_with_regulatory_features.txt\",\n",
    "#            sep = \"\\t\",\n",
    "#            row.names = FALSE,\n",
    "#            quote = FALSE)\n",
    "\n",
    "write.table(bin_annotations_final,\n",
    "            file = \"10kb_bin_genome_wide_annotated_reference_with_regulatory_features_and_gene_names.txt\",\n",
    "            sep = \"\\t\",\n",
    "            row.names = FALSE,\n",
    "            quote = FALSE)\n",
    "\n",
    "\n",
    "## ADDITIONAL CODE; create gene reference vector, with no duplicate gene names\n",
    "unique_genes <- unique(annotate_300bp_bins_hg38_basicgenes$annot.symbol)\n",
    "\n",
    "# save unique_genes as a reference set\n",
    "setwd(\"~/Library/CloudStorage/OneDrive-UniversityofToronto/1 MBP/1 Bratman Lab/1 Experiments/11 Sequencing/Reference/other_reference\")\n",
    "write(unique_genes, \"unique_gene_names_annotatR.txt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1508be3-bcc0-4447-8f07-057618bd0039",
   "metadata": {},
   "source": [
    "- Accidentally shifted columns of the reference to the right; fix before proceeding."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41643bee-0807-4a5d-8fc5-f6c644afe8a8",
   "metadata": {},
   "source": [
    "## Check expression of plasma K27me3 over all genes, sort by strongest correlation with mean VAF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96fb9fc1-b69e-4824-8a75-c59aefc1e3bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# goal; to identify a minimal set of genes who's plasma H3K27me3 expression is significantly correlated with tumor burden. These targets\n",
    "# may be prognostic.\n",
    "\n",
    "# import 10kb reference with gene names per bin. Reference was created using annotatR; see chunk for creating annotation reference in RMarkdown file\n",
    "# Note; creating the reference requires internet, so code is not included here.\n",
    "# Note; ENCODE Blacklist regions were not removed from the bins, since they will be matched to genes for the analysis.\n",
    "setwd(\"/cluster/projects/scottgroup/people/steven/reference_sets/other_references\")\n",
    "annotated_ref_w_gene_names <- read.delim(\"10kb_bin_genome_wide_annotated_reference_with_regulatory_features_and_gene_names.txt\", header = TRUE, sep = \"\\t\")\n",
    "\n",
    "# import unique gene list as character vector, with each gene represented only once in the vector.\n",
    "# import gene reference\n",
    "unique_genes <- read.delim(\"unique_gene_names_annotatR.txt\", header = FALSE, sep = \"\\t\")\n",
    "colnames(unique_genes) <- c(\"gene_name\")\n",
    "\n",
    "# loop; for each gene in unique_genes, subset annotated_ref_w_gene_names to rows with this gene, and quantify the median expression\n",
    "# of all normalized plasma K27me3 bins with those genes. Append median expression with gene name to empty data frame (median_gene_k27_expression).\n",
    "median_gene_k27_expression <- data.frame()\n",
    "\n",
    "\n",
    "##### LEFT OFF HERE #####\n",
    "for (i in 1:nrow(unique_genes)){\n",
    "    # Extract gene names for the current row\n",
    "    current_gene <- strsplit(unique_genes$gene_name[i], \", \")[[1]]\n",
    "\n",
    "    # Subset annotated reference to bins with gene\n",
    "    temp <- annotated_ref_w_gene_names[unlist(sapply(current_gene, function(gene) str_detect(annotated_ref_w_gene_names$gene_name, gene))),]\n",
    "                                                     \n",
    "    # subset normalized matrix to annotated reference subset, by bin_id\n",
    "    matrix_genes <- cpm_norm_log2[rownames(cpm_norm_log2) %in% temp$bin_id,]\n",
    "                                                     \n",
    "    # take median expression of plasma K27me3 per patient, over the gene region\n",
    "    median_expression <- t(as.data.frame(apply(matrix_genes, \n",
    "                               MARGIN = 2, \n",
    "                               FUN = median)))\n",
    "\n",
    "    # make rowname of median_expression, the gene which the 10kb matrix was subset by\n",
    "    rownames(median_expression) <- current_gene\n",
    "                                                     \n",
    "    # bind median expression to median_gene_k27_expression\n",
    "    median_gene_k27_expression <- rbind(median_gene_k27_expression, median_expression)      \n",
    "} # NEW ERROR ; MAY BE BECAUSE SOME OF THE GENES ARE NOT PRESENT IN THE NORMALIZED MATRIX, OR SOME ROWS HAVE NO GENES; LOOK INTO THE ERROR THROWN BY THE LOOP\n",
    "\n",
    "# for each row of the new data frame, find the correlation with mean VAF per patient, and append the R2 and p-val to the data frame\n",
    "\n",
    "# sort data frame by desc R2 (also, try descending p-val)\n",
    "\n",
    "# select top 5% of correlations and explore further\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# CHATGPT SOLUTION\n",
    "# Initialize an empty data frame to store the results\n",
    "correlation_results <- data.frame(gene_name = character(0), R2 = numeric(0), p_value = numeric(0))\n",
    "\n",
    "# Iterate over unique genes\n",
    "for (i in 1:nrow(unique_genes)) {\n",
    "    # Extract gene names for the current row\n",
    "    current_genes <- strsplit(unique_genes$gene_name[i], \", \")[[1]]\n",
    "\n",
    "    # Subset annotated reference to bins with gene\n",
    "    temp <- annotated_ref_w_gene_names[unlist(sapply(current_genes, function(gene) str_detect(annotated_ref_w_gene_names$gene_name, gene))),]\n",
    "\n",
    "    # If you want to further process 'temp', you can add your code here\n",
    "\n",
    "    # For each row of 'temp', find the correlation with mean VAF per patient\n",
    "    for (j in 1:nrow(temp)) {\n",
    "        # Assuming 'normalized_matrix' is your normalized matrix\n",
    "        current_row <- normalized_matrix[temp$bin_id[j], ]\n",
    "\n",
    "        # Calculate correlation\n",
    "        cor_result <- cor.test(current_row, your_mean_VAF_vector)\n",
    "\n",
    "        # Append the results to the data frame\n",
    "        correlation_results <- rbind(correlation_results, data.frame(\n",
    "            gene_name = temp$gene_name[j],\n",
    "            R2 = cor_result$estimate^2,\n",
    "            p_value = cor_result$p.value\n",
    "        ))\n",
    "    }\n",
    "}\n",
    "\n",
    "# Sort data frame by descending R2\n",
    "correlation_results <- correlation_results[order(-correlation_results$R2),]\n",
    "\n",
    "# Select top 5% of correlations\n",
    "top_5_percent <- head(correlation_results, nrow(correlation_results) * 0.05)\n",
    "\n",
    "# Explore further with 'top_5_percent'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e014fac9-7afa-48ab-b6d1-e7bf23c86232",
   "metadata": {},
   "source": [
    "- Continue this; use comments for guidance.\n",
    "- Biases may be introduced by this approach, for instance, in gene clusters, there will be a lot of overlapping genes in bins. Measurements of gene K27 expression may be homogeneous over these clusters, so that genes in close proximity may receive the same measurement, therefore same correlation with mean VAF.\n",
    "- An additional filter may be required; 1) for genes in close proximity, figure out which is most representative of the expression? Or 2) select the larger gene in the bin, as this would me more likely represented by K27 expression, which is broadly distributed?\n",
    "\n",
    "- 03-01-2024; Side note; blacklist regions were included in the gene names reference. And apparently, some genes are in blacklist regions; consider filtering out blacklist regions from the reference, or at the beginning of this analysis (once the code is sorted).\n",
    "- Also, need to remove healthy samples from the normalized matrix, cpm_norm_log2."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc977dcf-6ae0-4b5b-a982-fdf0ba0c18d9",
   "metadata": {},
   "source": [
    "# 07-12-2023 Incorporation of test data for validation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6de21c11-6314-4472-b9b2-da7d8323ddf6",
   "metadata": {},
   "source": [
    "## Bar plots over stage and IPI, with paired random feature selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bcc73ad-5422-41b9-94ee-fb9be51f9e39",
   "metadata": {},
   "outputs": [],
   "source": [
    "# start by subsetting the test matrix to the hypermethylated features of interest from Random Forest\n",
    "cpm_X_test <- cpm(X_test, normalized.lib.sizes = TRUE, log = FALSE)\n",
    "\n",
    "# log normalize and subset features\n",
    "cpm_X_test_log2 <- log2(cpm_X_test + 1)\n",
    "X_test_filtered <- cpm_X_test_log2[(rownames(cpm_X_test_log2)) %in% predictors_rf_hyper,]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# subset sample metadata to training samples.\n",
    "test_sample_metadata <- sample_list[sample_list$sample_name %in% colnames(X_test_filtered),]\n",
    "\n",
    "# re-import clinical metadata for test set\n",
    "setwd(\"/cluster/home/sdemichi/steven_scottgroup.txt/human_plasma_cfChIP-Seq/liberate_plasma_cfchip/temp/\")\n",
    "clinical_metadata <- as.data.frame(readxl::read_xlsx(\"25_07_2023_cohort_metadata_labeling_UPDATED_WITH_CLINICAL_ANNOTATIONS_plus_library_size_cfdna_concn.xlsx\"))\n",
    "\n",
    "# replace spaces in \"Donor ID (rep)\" so that you can match directly to mean VAF burden$Sample_id\n",
    "clinical_metadata$`Donor ID (rep)` <- gsub(\" \", \"_\", clinical_metadata$`Donor ID (rep)`)\n",
    "\n",
    "# match mean VAF information to clinical metadata\n",
    "clinical_metadata$mean_VAF <- mean_VAF_burden$mean_VAF[match(clinical_metadata$'Donor ID (rep)', mean_VAF_burden$Sample_id)]\n",
    "\n",
    "# filter clinical metadata by append cfDNA concentration to test, from clinical metadata.\n",
    "clinical_metadata_test <- clinical_metadata[clinical_metadata$PATIENT_ID %in% test_sample_metadata$PATIENT_ID,]\n",
    "\n",
    "# remove all T3 samples from the metadata\n",
    "clinical_metadata_test <- subset(clinical_metadata_test, !grepl(\"T3|T5\", clinical_metadata_test$`Donor ID (rep)`))\n",
    "\n",
    "# add new column to clinical_metadata_test which matches the colnames of X_test_attributes\n",
    "clinical_metadata_test <- clinical_metadata_test %>%\n",
    "  mutate(Bioinformatic_name_update = paste0(\"liberate_\",`Bioinformatic Name`,\"_K27me3\"))\n",
    "\n",
    "# sort clinical_metadata in order of colnames of X_train_attributes$PATIENT_ID\n",
    "clinical_metadata_test <- clinical_metadata_test[match(colnames(X_test_filtered), clinical_metadata_test$Bioinformatic_name_update), ] # NEED TO MAKE SURE THEY'RE IN ORDER; CORT BY X_TEST COLNAMES.\n",
    "\n",
    "# overwrite X_train_attributes with clinical_metadata\n",
    "X_test_attributes <- clinical_metadata_test\n",
    "\n",
    "# save new matrix as a DGEobject, since this links the sample attributes and the counts\n",
    "X_test_filtered_features_plus_attributes <- DGEList(counts = X_test_filtered, samples = X_test_attributes, remove.zeros = TRUE)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#calculate the median of each column (aka counts over top 1500 differential features, for each sample)\n",
    "median_X_test_filtered <- apply(X_test_filtered_features_plus_attributes$counts, \n",
    "                   MARGIN = 2, \n",
    "                   FUN = median)\n",
    "\n",
    "#bind the newly calculated medians and sumsto the pre-existing metadata\n",
    "X_test_attributes <- as.data.frame(cbind(X_test_attributes, median_X_test_filtered)) #Subset this data frame for plotting.\n",
    "\n",
    "##### METADATA-SPECIFIC CHANGES BEFORE PLOTTING #####\n",
    "X_test_attributes$ANN_ARBOR_STAGE <- ifelse(X_test_attributes$ANN_ARBOR_STAGE == \"NA\", \"non-lymphoma\", X_test_attributes$ANN_ARBOR_STAGE) # convert NA characters to \"non-lymphoma\"\n",
    "X_test_attributes$IPI <- as.character(X_test_attributes$IPI) #convert IPI from int to char\n",
    "\n",
    "X_test_attributes$mean_VAF <- as.numeric(X_test_attributes$mean_VAF) #convert IPI from int to char\n",
    "\n",
    "\n",
    "#For tumor diameter, create bins (<5cm, 5-10cm, 10-15cm, >15cm), then assign samples to those bins based off of tumor diameter; separate column\n",
    "X_test_attributes$TUMOR_MASS_MAX_CM <- as.numeric(X_test_attributes$TUMOR_MASS_MAX_CM)\n",
    "X_test_attributes <- X_test_attributes %>% \n",
    "  mutate(tumor_diameter_binned = cut(TUMOR_MASS_MAX_CM, breaks = c(0, 5, 10, 15, 30), labels = c(\"<5\", \"5-10\", \"10-15\", \">15\")))\n",
    "\n",
    "# combine Stage I/II and III/IV samples into early and late stage, respectively, for visualization\n",
    "X_test_attributes <- X_test_attributes %>%\n",
    "  mutate(Stage_Group = case_when(\n",
    "    ANN_ARBOR_STAGE %in% c(1, 2) ~ \"Early Stage\",\n",
    "    ANN_ARBOR_STAGE %in% c(3, 4) ~ \"Late Stage\",\n",
    "    TRUE ~ as.character(ANN_ARBOR_STAGE)  # Keep other values as is\n",
    "  ))\n",
    "\n",
    "# set factor level for Stage_Group\n",
    "X_test_attributes$Stage_Group <- factor(X_test_attributes$Stage_Group, levels = c(\"non-lymphoma\", \"Early Stage\", \"Late Stage\"))\n",
    "\n",
    "# 24-10-2023; create new variable where IPI is grouped 1+2 and 4+5, as this is clinically relevant\n",
    "X_test_attributes <- X_test_attributes %>%\n",
    "  mutate(IPI_grouped = case_when(\n",
    "    IPI %in% c(0, 1) ~ \"0+1\",\n",
    "    IPI %in% c(4, 5) ~ \"4+5\",\n",
    "    TRUE ~ as.character(IPI)  # Keep other values as is\n",
    "  ))\n",
    "\n",
    "# 24-10-2023; subset X_train_attributes to FL cases, and visualize FLIPI\n",
    "X_test_attributes_FL <- X_test_attributes[X_test_attributes$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY == \"FL\", ]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "##### PLOT #####\n",
    "\n",
    "# modify visualization to incorporate mean VAF as the colour; KEEP WORKING ON THIS! The points are not striking here, maybe resort to the heat map for this visualization.\n",
    "# p <- X_train_attributes_FL %>% \n",
    "p <- X_test_attributes %>% \n",
    "  ggplot(aes(x = IPI_grouped, y = median_X_test_filtered)) +#change to sum or median, depending on what you're visualizing\n",
    "  # ggplot(aes(x = Stage_Group, y = median_X_test_filtered)) +#change to sum or median, depending on what you're visualizing\n",
    "  #ggplot(aes(x = as.character(NB_EN_SITES), y = sum_X_train_filtered, colour = as.character(NB_EN_SITES))) +\n",
    "  geom_boxplot(aes(fill = IPI_grouped), alpha = 0.3) +\n",
    "  # geom_boxplot(aes(fill = Stage_Group), alpha = 0.3) +\n",
    "  geom_point(aes(colour = mean_VAF), position = position_jitterdodge(0.2), size = 7, alpha = 0.6) + \n",
    "  theme_minimal(base_size = 18) +\n",
    "  #scale_color_gradient(low = \"#185a9d\", high = \"#43cea2\") +\n",
    "  scale_color_gradient(low = \"blue\", high = \"red\") +\n",
    "  labs(x = \"Stage_Group\",\n",
    "       #x = \"IPI\",\n",
    "       # y = \"Median of Hypermethylated Feature Expression\",\n",
    "       # y = \"Median of ALL Hypermethylated Feature Expression\",\n",
    "        y = \"Median of MODELED Hypermethylated Feature Expression\",\n",
    "        colour = \"Stage_Group\", \n",
    "       #colour = \"IPI\"\n",
    "       )\n",
    "  #ggtitle(\"Median Hypermethylated Counts of Differential Features\")\n",
    "  #ggtitle(\"Median of Counts over Features\")\n",
    "  #ggtitle(\"Sum of Counts over Modeled Features\")\n",
    "  ggtitle(\"Median of Counts over Modeled Features\")\n",
    "print(p)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6d03e10-b894-4694-9d14-e12e6f24be28",
   "metadata": {},
   "source": [
    "- Trends in test set are generally holding up. Less of a difference between early stage and healthy, but that's okay, it's a very small test set."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddec149d-8625-4e66-9a58-896ff1eb18c1",
   "metadata": {},
   "source": [
    "## 13-12-2023 Survival incorporating test set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9915753-bbf4-410b-8f25-a5aea7015127",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e16e6586-8522-4f4a-b3e3-4f1bcffb8f91",
   "metadata": {},
   "source": [
    "# 08-12-2023 Combination of variables into a single analysis (K27me3, TMTV, mean VAF, and ichorCNA tumor burden)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1f92c73-c697-4e67-9ce2-1882250c71da",
   "metadata": {},
   "outputs": [],
   "source": [
    "# subset matrix to only lymphoma cases\n",
    "compared_burden_variables <- X_train_attributes_all_data_types[!(X_train_attributes_all_data_types$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY == \"HEALTHY\"),]\n",
    "\n",
    "compared_burden_variables <- compared_burden_variables[,c(\"PATIENT_ID\",\n",
    "                                                          \"median_X_train_filtered\",# K27me3 modeled feature expression\n",
    "                                                          \"mean_VAF\",# capp-seq\n",
    "                                                          \"STLG (SUV*mL/Kg)\",# MTV\n",
    "                                                          \"Tumor.Fraction\")]# ichorCNA\n",
    "\n",
    "# for this analysis, make NA values 0 for comparisons\n",
    "compared_burden_variables[is.na(compared_burden_variables)] <- 0\n",
    "\n",
    "scaled_compared_burden_variables <- as.data.frame(scale(compared_burden_variables[,2:5]))\n",
    "\n",
    "# re-append the patient IDs from the original matrix\n",
    "scaled_compared_burden_variables <- cbind(scaled_compared_burden_variables, PATIENT_ID = compared_burden_variables$PATIENT_ID)\n",
    "\n",
    "# order the data frame by decreasing median_X_train_filtered\n",
    "# test <- order(scaled_compared_burden_variables$median_X_train_filtered, decreasing = TRUE)\n",
    "\n",
    "# convert from wide to long format for visualization\n",
    "scaled_compared_burden_variables <- gather(scaled_compared_burden_variables, key = \"Measurement\", value = \"Value\", -PATIENT_ID)\n",
    "\n",
    "# visualize scaled data as box plot\n",
    "scaled_compared_burden_variables %>% \n",
    "  ggplot(aes(x = PATIENT_ID, y = Measurement, fill = Value)) +#change to sum or median, depending on what you're visualizing\n",
    "  geom_tile() +\n",
    "  theme_minimal(base_size = 18) +\n",
    "  scale_color_gradient(low = \"blue\", high = \"red\") #+\n",
    "  # labs(x = \"Stage_Group\",\n",
    "  #      #x = \"IPI\",\n",
    "  #      # y = \"Median of Hypermethylated Feature Expression\",\n",
    "  #      # y = \"Median of ALL Hypermethylated Feature Expression\",\n",
    "  #       y = \"Median of MODELED Hypermethylated Feature Expression\",\n",
    "  #       colour = \"Stage_Group\", \n",
    "  #      #colour = \"IPI\"\n",
    "  #      )\n",
    "  # #ggtitle(\"Median Hypermethylated Counts of Differential Features\")\n",
    "  # #ggtitle(\"Median of Counts over Features\")\n",
    "  # #ggtitle(\"Sum of Counts over Modeled Features\")\n",
    "  # ggtitle(\"Median of Counts over Modeled Features\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d7906fb-1324-4967-9122-0e900ea1fe2a",
   "metadata": {},
   "source": [
    "- Figure out what's going on here. Why are there cases that have high tumor burden by ichorCNA or TMTV, which do not have detectable mutations by CAPP-Seq.\n",
    "- Continue working on this grouped comparison. Scaling was performed, need to re-name the variable with patient IDs, convert to long format, and visualize.\n",
    "- issue; _C_ cases are present; need to remove HUCON and _C_ from patient ID column.\n",
    "- I just realized, there are duplicate rows in this data frame with different measurements for ichorCNA; follow up. Probably the reason for more rows present.\n",
    "- Duplicate rows removed; keep working on this. Maybe sort the heatmap or go for some other visualization? Also, some sort of concordance index between measurements? Or statistical test?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46d4d7b3-5e5f-4ce4-8def-8d9d7b960516",
   "metadata": {},
   "source": [
    "## Grid of linear correlations across data types; MTV, K27me3, CAPP-Seq, and ichorCNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e10d719-d3d1-487e-81ad-326d47b28d9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# code from above chunk\n",
    "compared_burden_variables <- X_train_attributes_all_data_types[!(X_train_attributes_all_data_types$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY == \"HEALTHY\"),]\n",
    "\n",
    "compared_burden_variables <- compared_burden_variables[,c(\"PATIENT_ID\",\n",
    "                                                          \"median_X_train_filtered\",# K27me3 modeled feature expression\n",
    "                                                          \"mean_VAF\",# capp-seq\n",
    "                                                          \"STLG (SUV*mL/Kg)\",# MTV\n",
    "                                                          # \"MTV (mL)\",# MTV; less correlated with burden measurements\n",
    "                                                          \"Tumor.Fraction\")]# ichorCNA\n",
    "\n",
    "# for this analysis, make NA values 0 for comparisons\n",
    "compared_burden_variables[is.na(compared_burden_variables)] <- 0\n",
    "\n",
    "# CHANGE IF REQUIRED; remove low tumor burden samples\n",
    "\n",
    "# tumor fraction (ichorCNA) < 0.03 (3%)\n",
    "# compared_burden_variables <- compared_burden_variables[!(compared_burden_variables$Tumor.Fraction < 0.03),]\n",
    "\n",
    "# mean VAF < 0.03 (3%)\n",
    "# compared_burden_variables <- compared_burden_variables[!(compared_burden_variables$mean_VAF < 0.03),]\n",
    "\n",
    "# both tumor fraction (ichorCNA) < 0.05 (5%) and mean VAF < 0.03 (3%)\n",
    "# need to be more stringent\n",
    "# compared_burden_variables <- compared_burden_variables[!(compared_burden_variables$mean_VAF < 0.03 & compared_burden_variables$Tumor.Fraction < 0.05),]\n",
    "\n",
    "\n",
    "# combine the 4 measurements into a single grid of correlations.\n",
    "# Customize upper panel\n",
    "upper.panel<-function(x, y){\n",
    "  points(x,y, pch=10)\n",
    "  r <- round(cor(x, y), digits=2)\n",
    "  p <- round(cor.test(x, y)$p.value, digits=6)\n",
    "  txt <- paste0(\"R = \", r, \"\\n\", \"p = \", p)\n",
    "  usr <- par(\"usr\")\n",
    "  on.exit(par(usr))\n",
    "  par(usr = c(0, 1, 0, 1))\n",
    "  text(0.5, 0.9, txt, cex = 1.5)\n",
    "}\n",
    "pairs(compared_burden_variables[,2:5], lower.panel = NULL, \n",
    "      upper.panel = upper.panel)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e2512f2-202e-412a-aa3c-3323889f1d6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "- Code from http://www.sthda.com/english/wiki/scatter-plot-matrices-r-base-graphs "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4755f34-98e1-4bf8-a115-904bce246aa2",
   "metadata": {},
   "source": [
    "# 15-01-2024 heatmap of clinical metadata, entire cohort (Figure 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef85f0aa-ee5f-43c6-ba3f-8e00af860604",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load library for heatmap visualization\n",
    "library(pheatmap)\n",
    "\n",
    "# import clinical metadata\n",
    "setwd(\"/cluster/home/sdemichi/steven_scottgroup.txt/human_plasma_cfChIP-Seq/liberate_plasma_cfchip/temp/\")\n",
    "clinical_metadata <- as.data.frame(readxl::read_xlsx(\"25_07_2023_cohort_metadata_labeling_UPDATED_WITH_CLINICAL_ANNOTATIONS_plus_library_size_cfdna_concn.xlsx\"))\n",
    "\n",
    "# import mean VAF information\n",
    "setwd(\"/cluster/projects/scottgroup/people/steven/human_plasma_cfChIP-Seq/liberate_plasma_cfchip/other_profiles/cappseq_data_lymphoma/updated_batch_2_all_cases/\")\n",
    "mean_VAF_burden <- read.delim(\"Bratman.dcs.uniq.burden.txt\", header = TRUE, sep = \"\\t\") # for dcs.unique\n",
    "\n",
    "# replace spaces in \"Donor ID (rep)\" so that you can match directly to mean VAF burden$Sample_id\n",
    "clinical_metadata$`Donor ID (rep)` <- gsub(\" \", \"_\", clinical_metadata$`Donor ID (rep)`)\n",
    "\n",
    "# match mean VAF information to clinical metadata\n",
    "clinical_metadata$mean_VAF <- mean_VAF_burden$mean_VAF[match(clinical_metadata$'Donor ID (rep)', mean_VAF_burden$Sample_id)]\n",
    "\n",
    "# remove healthy cases for this visualization\n",
    "clinical_metadata_lymphoma <- clinical_metadata[!(clinical_metadata$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY) == \"HEALTHY\",]\n",
    "\n",
    "# remove PBMC data\n",
    "clinical_metadata_lymphoma <- clinical_metadata_lymphoma[!(clinical_metadata_lymphoma$DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY) == \"PBMC\",]\n",
    "\n",
    "# mutate additional column, which represents whether baseline samples have follow-up plasma\n",
    "clinical_metadata_lymphoma <- clinical_metadata_lymphoma %>%\n",
    "mutate(follow_up_plasma = duplicated(substr(`Bioinformatic Name`, 1, 4), fromLast = TRUE) | duplicated(substr(`Bioinformatic Name`, 1, 4)))\n",
    "\n",
    "# remove all T3 and T5 samples\n",
    "clinical_metadata_lymphoma <- clinical_metadata_lymphoma[!str_detect(clinical_metadata_lymphoma$`Bioinformatic Name`, \"T3|T5\"),]\n",
    "\n",
    "# incorporate ichorCNA and MTV\n",
    "# import MTV data; data generated by the nuclear medicine team at UHN\n",
    "setwd(\"/cluster/projects/scottgroup/people/steven/human_plasma_cfChIP-Seq/liberate_plasma_cfchip/other_profiles/tmtv_data_lymphoma_cases\")\n",
    "tmtv_lymphoma <- readxl::read_xlsx(\"17-11-2023_PET_scan_data_lymphoma_cases.xlsx\")\n",
    "\n",
    "# import ichorCNA output and match to X_train_attributes in the same way you incorporated TMTV\n",
    "setwd(\"/cluster/projects/scottgroup/people/steven/human_plasma_cfChIP-Seq/liberate_plasma_cfchip/other_profiles/ichorCNA_data_shallow_WGS_lymphoma\")\n",
    "ichorCNA_lymphoma <- read.delim(\"Bratman.ichorCNA.txt\", header = TRUE, sep = \"\\t\")\n",
    "\n",
    "# merge MTV data frame with X_train_attributes\n",
    "clinical_metadata_lymphoma <- merge(clinical_metadata_lymphoma, tmtv_lymphoma, by.x = \"PATIENT_ID\", by.y = \"LIBERATE ID\", all.x = TRUE)\n",
    "\n",
    "# merge ichorCNA data frame with X_train_attributes; this step is where you're gaining duplicate rows!!\n",
    "clinical_metadata_lymphoma <- merge(clinical_metadata_lymphoma, ichorCNA_lymphoma, by.x = \"PATIENT_ID\", by.y = \"PATIENT_ID\", all.x = TRUE)\n",
    "\n",
    "# remove duplicate ichorCNA entries due to accidental duplicate sequencing of some shallow WGS profiles. Selected single representative entry for duplicates.\n",
    "clinical_metadata_lymphoma <- clinical_metadata_lymphoma %>%\n",
    "  distinct(PATIENT_ID, .keep_all = TRUE)\n",
    "\n",
    "# mutate an additional column onto the data frame, named by the data type itself, which indicates whether the data is present.\n",
    "clinical_metadata_lymphoma <- clinical_metadata_lymphoma %>%\n",
    "  mutate(\n",
    "      # CAPP-Seq\n",
    "      CAPP_Seq = case_when(!is.na(mean_VAF) ~ TRUE, is.na(mean_VAF) ~ FALSE),\n",
    "      \n",
    "      # shallow WGS\n",
    "      shallow_WGS = case_when(!is.na(`Tumor.Fraction`) ~ TRUE, is.na(`Tumor.Fraction`) ~ FALSE),\n",
    "      \n",
    "      # MTV\n",
    "      metabolic_tumor_volume = case_when(!is.na(`STLG (SUV*mL/Kg)`) ~ TRUE, is.na(`STLG (SUV*mL/Kg)`) ~ FALSE))\n",
    "\n",
    "# subset to only essential columns of the metadata for visualization\n",
    "heatmap_metadata <- clinical_metadata_lymphoma[,c(9,\n",
    "                                                  17,\n",
    "                                                  23,\n",
    "                                                  24,\n",
    "                                                  25,\n",
    "                                                  27:31,\n",
    "                                                  34:35,\n",
    "                                                  47,\n",
    "                                                  74:76\n",
    "                                                 )]\n",
    "\n",
    "# convert columns w numeric values to numeric\n",
    "heatmap_metadata <- heatmap_metadata %>%\n",
    "  mutate_at(vars(AGE_CONSENT,\n",
    "                 TUMOR_MASS_MAX_CM,\n",
    "                 LDH,\n",
    "                 NB_EN_SITES), as.numeric)\n",
    "\n",
    "# to see indexed column names\n",
    "# print(paste0(\"Column names and indices:\\n\", paste0(seq_along(colnames(clinical_metadata_lymphoma)), \". \", colnames(clinical_metadata_lymphoma), collapse = \"|\")))\n",
    "\n",
    "# generate heatmap, once data is compiled\n",
    "# Set the row names to donor IDs\n",
    "# rownames(heatmap_metadata) <- heatmap_metadata$`Donor ID (rep)`\n",
    "\n",
    "\n",
    "\n",
    "# CONTINUE FROM HERE\n",
    "\n",
    "# # Generate the heatmap\n",
    "# heatmap(heatmap_metadata#, \n",
    "#          # cluster_rows=FALSE, \n",
    "#          # cluster_cols=FALSE#, \n",
    "#          # color=colorRampPalette(c(\"white\", \"blue\"))(100), \n",
    "#          # fontsize_row=5, fontsize_col=8, \n",
    "#          # labels_col=colnames(clinical_metadata), \n",
    "#          # labels_row=rownames(clinical_metadata), \n",
    "#          # main=\"Clinical Metadata Heatmap\"\n",
    "#         )\n",
    "\n",
    "# for all metadata\n",
    "heatmap_metadata %>%\n",
    "  ggplot() +\n",
    "  geom_tile(aes(x = `Donor ID (rep)`, y = \"DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY\", fill = DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY), show.legend = FALSE, width = 0.9, height = 0.9) +\n",
    "  geom_tile(aes(x = `Donor ID (rep)`, y = \"TREATMENT\", fill = TREATMENT), show.legend = FALSE, width = 0.9, height = 0.9) +\n",
    "  geom_tile(aes(x = `Donor ID (rep)`, y = \"SEX\", fill = SEX), show.legend = FALSE, width = 0.9, height = 0.9) +\n",
    "  # # geom_tile(aes(x = `Donor ID (rep)`, y = \"AGE_CONSENT\", fill = AGE_CONSENT), show.legend = FALSE, width = 0.9, height = 0.9) +\n",
    "  geom_tile(aes(x = `Donor ID (rep)`, y = \"ANN_ARBOR_STAGE\", fill = ANN_ARBOR_STAGE), show.legend = FALSE, width = 0.9, height = 0.9) +\n",
    "  # # geom_tile(aes(x = `Donor ID (rep)`, y = \"TUMOR_MASS_MAX_CM\", fill = TUMOR_MASS_MAX_CM), show.legend = FALSE, width = 0.9, height = 0.9) +\n",
    "  # # geom_tile(aes(x = `Donor ID (rep)`, y = \"ECOG\", fill = ECOG), show.legend = FALSE, width = 0.9, height = 0.9) +\n",
    "  # # geom_tile(aes(x = `Donor ID (rep)`, y = \"LDH\", fill = LDH), show.legend = FALSE, width = 0.9, height = 0.9) +\n",
    "  # # geom_tile(aes(x = `Donor ID (rep)`, y = \"NB_EN_SITES\", fill = NB_EN_SITES), show.legend = FALSE, width = 0.9, height = 0.9) +\n",
    "  geom_tile(aes(x = `Donor ID (rep)`, y = \"IPI\", fill = IPI), show.legend = FALSE, width = 0.9, height = 0.9) +\n",
    "  geom_tile(aes(x = `Donor ID (rep)`, y = \"FLIPI\", fill = FLIPI), show.legend = FALSE, width = 0.9, height = 0.9) +\n",
    "  geom_tile(aes(x = `Donor ID (rep)`, y = \"follow_up_plasma\", fill = follow_up_plasma), show.legend = FALSE, width = 0.9, height = 0.9) +\n",
    "  geom_tile(aes(x = `Donor ID (rep)`, y = \"CAPP_Seq\", fill = CAPP_Seq), show.legend = FALSE, width = 0.9, height = 0.9) +\n",
    "  geom_tile(aes(x = `Donor ID (rep)`, y = \"shallow_WGS\", fill = shallow_WGS), show.legend = FALSE, width = 0.9, height = 0.9) +\n",
    "  geom_tile(aes(x = `Donor ID (rep)`, y = \"metabolic_tumor_volume\", fill = metabolic_tumor_volume), show.legend = FALSE, width = 0.9, height = 0.9) +\n",
    "  theme_minimal(base_size = 16) +\n",
    "  theme(axis.text.x = element_text(angle = 45, hjust=1)) +\n",
    "  coord_equal()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# for clinical metadata only\n",
    "heatmap_metadata %>%\n",
    "  ggplot() +\n",
    "  geom_tile(aes(x = `Donor ID (rep)`, y = \"DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY\", fill = DIAGNOSIS_CLASS_SUMMARIZE_HEALTHY), show.legend = FALSE, width = 0.9, height = 0.9) +\n",
    "  geom_tile(aes(x = `Donor ID (rep)`, y = \"TREATMENT\", fill = TREATMENT), show.legend = FALSE, width = 0.9, height = 0.9) +\n",
    "  geom_tile(aes(x = `Donor ID (rep)`, y = \"SEX\", fill = SEX), show.legend = FALSE, width = 0.9, height = 0.9) +\n",
    "  # geom_tile(aes(x = `Donor ID (rep)`, y = \"AGE_CONSENT\", fill = AGE_CONSENT), show.legend = FALSE, width = 0.9, height = 0.9) +\n",
    "  geom_tile(aes(x = `Donor ID (rep)`, y = \"ANN_ARBOR_STAGE\", fill = ANN_ARBOR_STAGE), show.legend = FALSE, width = 0.9, height = 0.9) +\n",
    "  # # geom_tile(aes(x = `Donor ID (rep)`, y = \"TUMOR_MASS_MAX_CM\", fill = TUMOR_MASS_MAX_CM), show.legend = FALSE, width = 0.9, height = 0.9) +\n",
    "  # # geom_tile(aes(x = `Donor ID (rep)`, y = \"ECOG\", fill = ECOG), show.legend = FALSE, width = 0.9, height = 0.9) +\n",
    "  # # geom_tile(aes(x = `Donor ID (rep)`, y = \"LDH\", fill = LDH), show.legend = FALSE, width = 0.9, height = 0.9) +\n",
    "  # # geom_tile(aes(x = `Donor ID (rep)`, y = \"NB_EN_SITES\", fill = NB_EN_SITES), show.legend = FALSE, width = 0.9, height = 0.9) +\n",
    "  geom_tile(aes(x = `Donor ID (rep)`, y = \"IPI\", fill = IPI), show.legend = FALSE, width = 0.9, height = 0.9) +\n",
    "  geom_tile(aes(x = `Donor ID (rep)`, y = \"FLIPI\", fill = FLIPI), show.legend = FALSE, width = 0.9, height = 0.9) +\n",
    "  geom_tile(aes(x = `Donor ID (rep)`, y = \"follow_up_plasma\", fill = follow_up_plasma), show.legend = FALSE, width = 0.9, height = 0.9) +\n",
    "  theme_minimal(base_size = 16) +\n",
    "  theme(axis.text.x = element_text(angle = 45, hjust=1)) +\n",
    "  coord_equal()\n",
    "\n",
    "# for sequencing data only\n",
    "heatmap_metadata %>%\n",
    "  ggplot() +\n",
    "  geom_tile(aes(x = `Donor ID (rep)`, y = \"CAPP_Seq\", fill = CAPP_Seq), width = 0.9, height = 0.9) +\n",
    "  geom_tile(aes(x = `Donor ID (rep)`, y = \"shallow_WGS\", fill = shallow_WGS), width = 0.9, height = 0.9) +\n",
    "  geom_tile(aes(x = `Donor ID (rep)`, y = \"metabolic_tumor_volume\", fill = metabolic_tumor_volume), width = 0.9, height = 0.9) +\n",
    "  theme_minimal(base_size = 16) +\n",
    "  theme(axis.text.x = element_text(angle = 45, hjust=1)) +\n",
    "  coord_equal()\n",
    "\n",
    "# for numeric (STILL GETTING THE SAME ERROR when you add multiple gradients)\n",
    "heatmap_metadata %>%\n",
    "  ggplot() +\n",
    "  geom_tile(aes(x = `Donor ID (rep)`, y = \"AGE_CONSENT\", fill = AGE_CONSENT), width = 0.9, height = 0.9) +\n",
    "  geom_tile(aes(x = `Donor ID (rep)`, y = \"TUMOR_MASS_MAX_CM\", fill = TUMOR_MASS_MAX_CM), width = 0.9, height = 0.9) +\n",
    "  # geom_tile(aes(x = `Donor ID (rep)`, y = \"ECOG\", fill = ECOG), width = 0.9, height = 0.9) +\n",
    "  geom_tile(aes(x = `Donor ID (rep)`, y = \"LDH\", fill = LDH), width = 0.9, height = 0.9) + \n",
    "  geom_tile(aes(x = `Donor ID (rep)`, y = \"NB_EN_SITES\", fill = NB_EN_SITES), show.legend = FALSE, width = 0.9, height = 0.9) +\n",
    "  theme_minimal(base_size = 16) +\n",
    "  theme(axis.text.x = element_text(angle = 45, hjust=1)) +\n",
    "  scale_fill_gradient() +\n",
    "  coord_equal()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Convert the data frame from wide to long format (THIS IS TECHNICALLY THE WAY IT SHOULD BE DONE)\n",
    "heatmap_metadata_long <- pivot_longer(heatmap_metadata, cols = everything(), names_to = \"variable\", values_to = \"value\")\n",
    "\n",
    "# Create a simplified heatmap\n",
    "ggplot(heatmap_metadata_long, aes(x = `Donor ID (rep)`, y = variable, fill = value)) +\n",
    "  geom_tile() +\n",
    "  theme_minimal(base_size = 18) +\n",
    "  scale_fill_identity()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
